{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating SLiM exome recombination maps\n",
    "\n",
    "This notebook generates a recombination map for SLiM that implements a realistic exome structure and includes the positions of sites from the archaic admixture array. It also saves the positions of array sites and coordinates of exons separately."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pybedtools import BedTool\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "plt.style.use('ggplot')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "RECOMB_RATE = 1e-8 # crossovers per bp per generation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exon coordinates processing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First download the GTF annotations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = pd.read_table('ftp://ftp.ensembl.org/pub/release-75/gtf/homo_sapiens/Homo_sapiens.GRCh37.75.gtf.gz',\n",
    "                    header=None, sep='\\t', skipinitialspace=True, skiprows=5, compression='gzip',\n",
    "                    names=['chrom', 'source', 'feature', 'start', 'end',\n",
    "                           'score', 'strand', 'frame', 'attribute'], low_memory=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What does the data look like?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14412</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972\"; transcript_id \"ENST...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972\"; transcript_id \"ENST...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972\"; transcript_id \"ENST...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972\"; transcript_id \"ENST...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom                source     feature  start    end score strand frame  \\\n",
       "0     1            pseudogene        gene  11869  14412     .      +     .   \n",
       "1     1  processed_transcript  transcript  11869  14409     .      +     .   \n",
       "2     1  processed_transcript        exon  11869  12227     .      +     .   \n",
       "3     1  processed_transcript        exon  12613  12721     .      +     .   \n",
       "4     1  processed_transcript        exon  13221  14409     .      +     .   \n",
       "\n",
       "                                           attribute  \n",
       "0  gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"...  \n",
       "1  gene_id \"ENSG00000223972\"; transcript_id \"ENST...  \n",
       "2  gene_id \"ENSG00000223972\"; transcript_id \"ENST...  \n",
       "3  gene_id \"ENSG00000223972\"; transcript_id \"ENST...  \n",
       "4  gene_id \"ENSG00000223972\"; transcript_id \"ENST...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "protein_coding                        2001054\n",
       "nonsense_mediated_decay                293471\n",
       "processed_transcript                   173401\n",
       "retained_intron                        150034\n",
       "lincRNA                                 55928\n",
       "antisense                               45811\n",
       "processed_pseudogene                    24517\n",
       "pseudogene                              17830\n",
       "unprocessed_pseudogene                  13763\n",
       "miRNA                                   10209\n",
       "transcribed_unprocessed_pseudogene       7931\n",
       "misc_RNA                                 6554\n",
       "snRNA                                    6215\n",
       "snoRNA                                   4791\n",
       "sense_intronic                           3241\n",
       "Name: source, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.source.value_counts().head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "exon              1306656\n",
       "CDS                791856\n",
       "UTR                304070\n",
       "transcript         215170\n",
       "stop_codon          73411\n",
       "start_codon         73358\n",
       "gene                63677\n",
       "Selenocysteine        114\n",
       "Name: feature, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.feature.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Take only autosomal protein-coding regions, since we won't be simulating sex chromosomes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "AUTOSOMES = [str(i + 1) for i in range(22)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "exons = gtf[gtf.chrom.isin(AUTOSOMES) &\n",
    "           (gtf.source == \"protein_coding\") &\n",
    "           (gtf.feature == \"exon\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cds = gtf[gtf.chrom.isin(AUTOSOMES) &\n",
    "           (gtf.source == \"protein_coding\") &\n",
    "           (gtf.feature == \"CDS\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "72334698"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BedTool.from_dataframe(exons).sort().merge().total_coverage()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33514076"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BedTool.from_dataframe(cds).sort().merge().total_coverage()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**TODO** It's clear that I should've used the CDS sequence, not just the exonic sequence (which includes UTRs). This leads to a 50% reduction in the total amount of sequence to simulate! Crazy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "exons = gtf[gtf.chrom.isin(AUTOSOMES) &\n",
    "           (gtf.source == \"protein_coding\") &\n",
    "           (gtf.feature == \"exon\")]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge the overlapping exons:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "exons = BedTool.from_dataframe(exons).sort().merge().to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "exons['type'] = 'exon'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert chromosome IDs to integers and sort coordinates:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "exons.chrom = exons.chrom.astype(int)\n",
    "exons.sort_values(by=['chrom', 'start'], inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What is the distribution of exon lengths?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "lengths = exons.end - exons.start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    214193.000000\n",
       "mean        337.708039\n",
       "std         776.506070\n",
       "min           1.000000\n",
       "25%          96.000000\n",
       "50%         138.000000\n",
       "75%         217.000000\n",
       "max       24409.000000\n",
       "dtype: float64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lengths.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Specify the recombination rate within exons:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "exons.ix[exons.type == 'exon', 'recomb_rate'] = RECOMB_RATE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The modified dataframe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>recomb_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>134900</td>\n",
       "      <td>135802</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>137620</td>\n",
       "      <td>139379</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>367639</td>\n",
       "      <td>368634</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>621058</td>\n",
       "      <td>622053</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end  type   recomb_rate\n",
       "0      1   69090   70008  exon  1.000000e-08\n",
       "1      1  134900  135802  exon  1.000000e-08\n",
       "2      1  137620  139379  exon  1.000000e-08\n",
       "3      1  367639  368634  exon  1.000000e-08\n",
       "4      1  621058  622053  exon  1.000000e-08"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process positions from the archaic admixture array"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the coordinates of sites from the archaic admixture array:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "array_sites = pd.read_table('../raw_data/ice_age/filter.gz', names=['chrom', 'end', 'set_1', 'set_2'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Subset to sites in the so called \"Set 1\" (see `raw_data/ice_age/README.md` for more information):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "array_sites = array_sites[array_sites.set_1 == 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "array_sites['start'] = array_sites.end - 1\n",
    "array_sites = array_sites[['chrom', 'start', 'end']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "array_sites['type'] = 'snp'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "DataFrame after filtering:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>847982</td>\n",
       "      <td>847983</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>853088</td>\n",
       "      <td>853089</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>853595</td>\n",
       "      <td>853596</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>854792</td>\n",
       "      <td>854793</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>867150</td>\n",
       "      <td>867151</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end type\n",
       "0      1  847982  847983  snp\n",
       "1      1  853088  853089  snp\n",
       "2      1  853595  853596  snp\n",
       "3      1  854792  854793  snp\n",
       "5      1  867150  867151  snp"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "array_sites.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a union of exon regions with admixture array informative positions\n",
    "\n",
    "Inter-exon informative sites will be simulated individually, with an appropriate recombination rate set between them and adjacent exons or other informative sites. Intra-exon sites will be ignored when building the recombination map (they will be recombining as a part of the exons)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Take the subset of array sites that lie outside exons:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "exome = BedTool.from_dataframe(exons)\n",
    "sites_outside_exons = BedTool.from_dataframe(array_sites).intersect(exome, v=True).to_dataframe().rename(columns={'name': 'type'})\n",
    "\n",
    "# we can't simulate all admixture array sites, unfortunately,\n",
    "# so let's focus on a subset\n",
    "sites_outside_exons = sites_outside_exons.sample(40000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "exons_and_sites = pd.concat([sites_outside_exons,\n",
    "                             exons]).sort_values(by=['chrom', 'start']).reset_index(drop=True)\n",
    "exons_and_sites = exons_and_sites[['chrom', 'start', 'end', 'type', 'recomb_rate']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>recomb_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>134900</td>\n",
       "      <td>135802</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>137620</td>\n",
       "      <td>139379</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>367639</td>\n",
       "      <td>368634</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>621058</td>\n",
       "      <td>622053</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end  type   recomb_rate\n",
       "0      1   69090   70008  exon  1.000000e-08\n",
       "1      1  134900  135802  exon  1.000000e-08\n",
       "2      1  137620  139379  exon  1.000000e-08\n",
       "3      1  367639  368634  exon  1.000000e-08\n",
       "4      1  621058  622053  exon  1.000000e-08"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons_and_sites.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "254193"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(exons_and_sites)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generating the recombination map of exons and archaic admixture array sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_recombination_gaps(regions):\n",
    "    \"\"\"Add 1 bp \"recombination gap\" between each record in\n",
    "    a given DataFrame of exon and snp coordinates.\n",
    "    \"\"\"\n",
    "    # create a new DataFrame with coordinates of 1 bp gaps\n",
    "    gaps = pd.DataFrame({'chrom'       : regions.chrom.values,\n",
    "                         'start'       : regions.end.values,\n",
    "                         'end'         : regions.end.values + 1,\n",
    "                         'recomb_rate' : list(RECOMB_RATE * (regions.start[1:].values -  # between exons/sites\n",
    "                                                             regions.end[:-1].values)) +\n",
    "                                         [0.5], # between chromosomes\n",
    "                         'type'        : 'gap'},\n",
    "                        columns=['chrom', 'start', 'end', 'recomb_rate', 'type'])\n",
    "\n",
    "    # merge the dataframes of regions and gap coordinates\n",
    "    regions_and_gaps = pd.concat([regions, gaps]).sort_values(by=['chrom', 'start']).reset_index(drop=True)\n",
    "\n",
    "    return regions_and_gaps[['chrom', 'start', 'end', 'type', 'recomb_rate']]\n",
    "\n",
    "\n",
    "def concatenate_regions(regions):\n",
    "    \"\"\"Concatenate regions/sites on all chromosome as if they were\n",
    "    directly adjacent on a single chromosome, changing the coordinates\n",
    "    accordingly.\n",
    "    \"\"\"\n",
    "    concat_regions = regions.copy()\n",
    "\n",
    "    concat_regions['width'] = regions.end - regions.start\n",
    "    concat_regions['slim_start'] = pd.Series([0] + list(concat_regions.width[:-1])).cumsum().values\n",
    "    concat_regions['slim_end'] = concat_regions.width.cumsum() - 1\n",
    "    \n",
    "    return concat_regions\n",
    "\n",
    "\n",
    "def create_recomb_map(regions):\n",
    "    \"\"\"Create recombination map for all regions/sites on all chromosomes.\"\"\"\n",
    "    recomb_map = []\n",
    "\n",
    "    for chrom in AUTOSOMES:\n",
    "        # add recombination gaps between regions on this chromosome\n",
    "        recomb_map.append(add_recombination_gaps(regions.query('chrom == {}'.format(chrom))))\n",
    "        \n",
    "    # concatenate exome recombination maps of all chromosomes\n",
    "    recomb_map = pd.concat(recomb_map, ignore_index=True).sort_values(by=['chrom', 'start']).reset_index(drop=True)\n",
    "\n",
    "    # remove the very last base of the recombination map\n",
    "    # (it has a 0.5 recombination rate anyway and there's no other chromosome\n",
    "    # after it)\n",
    "    recomb_map = recomb_map[:-1]\n",
    "    \n",
    "    return recomb_map"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Recombination rate between exons and/or informative positions is implemented by inserting a 1 bp \"gap\" between each adjacent region/site and setting the recombination rate at these positions to $L \\times 1\\cdot10^{-8}$ crossovers per generation ($L$ is the distance between exons/sites).\n",
    "\n",
    "The recombination rate of the \"gap\" between the last feature on one chromosome and the first feature on another chromosome will be 0.5."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a recombination map of exons and informative sites:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "recomb_map = create_recomb_map(exons_and_sites)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>recomb_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>70008</td>\n",
       "      <td>70009</td>\n",
       "      <td>gap</td>\n",
       "      <td>6.489200e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>134900</td>\n",
       "      <td>135802</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>135802</td>\n",
       "      <td>135803</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.818000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>137620</td>\n",
       "      <td>139379</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end  type   recomb_rate\n",
       "0      1   69090   70008  exon  1.000000e-08\n",
       "1      1   70008   70009   gap  6.489200e-04\n",
       "2      1  134900  135802  exon  1.000000e-08\n",
       "3      1  135802  135803   gap  1.818000e-05\n",
       "4      1  137620  139379  exon  1.000000e-08"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "recomb_map.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter out \"gaps\" between directly adjacent features:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "recomb_map = recomb_map.query('(type == \"snp\") | (recomb_rate > 0)')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert the coordinates of all regions/sites/gaps into SLiM's 0-based single segment coordinate system:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "concat_map = concatenate_regions(recomb_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>recomb_rate</th>\n",
       "      <th>width</th>\n",
       "      <th>slim_start</th>\n",
       "      <th>slim_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>918</td>\n",
       "      <td>0</td>\n",
       "      <td>917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>70008</td>\n",
       "      <td>70009</td>\n",
       "      <td>gap</td>\n",
       "      <td>6.489200e-04</td>\n",
       "      <td>1</td>\n",
       "      <td>918</td>\n",
       "      <td>918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>134900</td>\n",
       "      <td>135802</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>902</td>\n",
       "      <td>919</td>\n",
       "      <td>1820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>135802</td>\n",
       "      <td>135803</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.818000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>1821</td>\n",
       "      <td>1821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>137620</td>\n",
       "      <td>139379</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>1759</td>\n",
       "      <td>1822</td>\n",
       "      <td>3580</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end  type   recomb_rate  width  slim_start  slim_end\n",
       "0      1   69090   70008  exon  1.000000e-08    918           0       917\n",
       "1      1   70008   70009   gap  6.489200e-04      1         918       918\n",
       "2      1  134900  135802  exon  1.000000e-08    902         919      1820\n",
       "3      1  135802  135803   gap  1.818000e-05      1        1821      1821\n",
       "4      1  137620  139379  exon  1.000000e-08   1759        1822      3580"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concat_map.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>recomb_rate</th>\n",
       "      <th>width</th>\n",
       "      <th>slim_start</th>\n",
       "      <th>slim_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>508380</th>\n",
       "      <td>22</td>\n",
       "      <td>51220615</td>\n",
       "      <td>51220779</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>164</td>\n",
       "      <td>72628011</td>\n",
       "      <td>72628174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508381</th>\n",
       "      <td>22</td>\n",
       "      <td>51220779</td>\n",
       "      <td>51220780</td>\n",
       "      <td>gap</td>\n",
       "      <td>4.170000e-06</td>\n",
       "      <td>1</td>\n",
       "      <td>72628175</td>\n",
       "      <td>72628175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508382</th>\n",
       "      <td>22</td>\n",
       "      <td>51221196</td>\n",
       "      <td>51221714</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>518</td>\n",
       "      <td>72628176</td>\n",
       "      <td>72628693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508383</th>\n",
       "      <td>22</td>\n",
       "      <td>51221714</td>\n",
       "      <td>51221715</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.140000e-06</td>\n",
       "      <td>1</td>\n",
       "      <td>72628694</td>\n",
       "      <td>72628694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508384</th>\n",
       "      <td>22</td>\n",
       "      <td>51221928</td>\n",
       "      <td>51222091</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>163</td>\n",
       "      <td>72628695</td>\n",
       "      <td>72628857</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom     start       end  type   recomb_rate  width  slim_start  \\\n",
       "508380     22  51220615  51220779  exon  1.000000e-08    164    72628011   \n",
       "508381     22  51220779  51220780   gap  4.170000e-06      1    72628175   \n",
       "508382     22  51221196  51221714  exon  1.000000e-08    518    72628176   \n",
       "508383     22  51221714  51221715   gap  2.140000e-06      1    72628694   \n",
       "508384     22  51221928  51222091  exon  1.000000e-08    163    72628695   \n",
       "\n",
       "        slim_end  \n",
       "508380  72628174  \n",
       "508381  72628175  \n",
       "508382  72628693  \n",
       "508383  72628694  \n",
       "508384  72628857  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concat_map.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save the recombination map in a SLiM-friendly format (end-positions of exons and gaps, without the positions of SNPs since they don't have recombination rates themselves):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_map.query('type != \"snp\"')[['slim_end', 'recomb_rate']].to_csv('../clean_data/exome_and_sites_recombination_map.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generating recombination map of exons only (ignoring admixture array sites)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "recomb_map_exons_only = create_recomb_map(exons_and_sites.query('type != \"snp\"'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "concat_map_exons_only = concatenate_regions(recomb_map_exons_only)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "concat_map_exons_only[['slim_end', 'recomb_rate']].to_csv('../clean_data/exome_only_recombination_map.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>recomb_rate</th>\n",
       "      <th>width</th>\n",
       "      <th>slim_start</th>\n",
       "      <th>slim_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>428380</th>\n",
       "      <td>22</td>\n",
       "      <td>51220615</td>\n",
       "      <td>51220779</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>164</td>\n",
       "      <td>72548043</td>\n",
       "      <td>72548206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428381</th>\n",
       "      <td>22</td>\n",
       "      <td>51220779</td>\n",
       "      <td>51220780</td>\n",
       "      <td>gap</td>\n",
       "      <td>4.170000e-06</td>\n",
       "      <td>1</td>\n",
       "      <td>72548207</td>\n",
       "      <td>72548207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428382</th>\n",
       "      <td>22</td>\n",
       "      <td>51221196</td>\n",
       "      <td>51221714</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>518</td>\n",
       "      <td>72548208</td>\n",
       "      <td>72548725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428383</th>\n",
       "      <td>22</td>\n",
       "      <td>51221714</td>\n",
       "      <td>51221715</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.140000e-06</td>\n",
       "      <td>1</td>\n",
       "      <td>72548726</td>\n",
       "      <td>72548726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428384</th>\n",
       "      <td>22</td>\n",
       "      <td>51221928</td>\n",
       "      <td>51222091</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>163</td>\n",
       "      <td>72548727</td>\n",
       "      <td>72548889</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom     start       end  type   recomb_rate  width  slim_start  \\\n",
       "428380     22  51220615  51220779  exon  1.000000e-08    164    72548043   \n",
       "428381     22  51220779  51220780   gap  4.170000e-06      1    72548207   \n",
       "428382     22  51221196  51221714  exon  1.000000e-08    518    72548208   \n",
       "428383     22  51221714  51221715   gap  2.140000e-06      1    72548726   \n",
       "428384     22  51221928  51222091  exon  1.000000e-08    163    72548727   \n",
       "\n",
       "        slim_end  \n",
       "428380  72548206  \n",
       "428381  72548207  \n",
       "428382  72548725  \n",
       "428383  72548726  \n",
       "428384  72548889  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concat_map_exons_only.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save SLiM coordinates of exons only (ignoring admixture array sites)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "concat_map_exons_only.query('type == \"exon\"')[['slim_start', 'slim_end']].to_csv('../clean_data/exome_only_exon_coordinates.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br><br><br>\n",
    "## Save SLiM coordinates of all sites from the archaic admixture array\n",
    "\n",
    "Recombination map includes only positions of sites that fall outside of exonic regions. However, in order to simulate the sites from the archaic admixture array, we have to know the positions of sites _within_ exons too.\n",
    "\n",
    "SLiM simulates all exons and individual sites as one continuous segment of concatenated regions, with coordinates of exons and sites shifted appropriately. To obtain the coordinates of array sites within exons, we need to find out, for each site, which exon does it fall in and calculate its position relative to the start of that exon."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Take the subset of array sites that lie inside exons:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sites_within_exons = BedTool.from_dataframe(array_sites).intersect(exome).to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "sites_within_exons = sites_within_exons.rename(columns={'name': 'type'})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get a DataFrame of the coordinates of exons that contain a site from the admixture array (will contain multiple copies of one exon if more than one site falls within that exon):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "exons_with_sites = BedTool.from_dataframe(concat_map.query('type == \"exon\"')).intersect(BedTool.from_dataframe(sites_within_exons), wa=True).to_dataframe()\n",
    "\n",
    "\n",
    "# rename and subset columns\n",
    "exons_with_sites = exons_with_sites.rename(columns={'name': 'type',\n",
    "                                                    'score': 'recomb_rate',\n",
    "                                                    'strand': 'width',\n",
    "                                                    'thickStart': 'slim_start',\n",
    "                                                    'thickEnd': 'slim_end'})[['chrom', 'start', 'end', 'type', 'recomb_rate', 'width', 'slim_start', 'slim_end']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate the position of each site relative to the start of \"its\" exon and convert this position into a SLiM single-segment coordinate (i.e. relative the the position 0 of the simulated segment):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "sites_within_exons['width'] = 1\n",
    "sites_within_exons['slim_start'] = sites_within_exons.start - exons_with_sites.start + exons_with_sites.slim_start\n",
    "sites_within_exons['slim_end'] = sites_within_exons.slim_start"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Output the coordinates of subsampled sites:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "subsampled_exonic_sites = sites_within_exons.sample(10000).sort_values(by=['slim_start']).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "subsampled_exonic_sites['slim_start'].to_csv('../clean_data/admixture_array_coordinates_exonic.txt', index=False)\n",
    "concat_map.query('type == \"snp\"')['slim_start'].to_csv('../clean_data/admixture_array_coordinates_nonexonic.txt', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save SLiM coordinates of exonic regions only\n",
    "\n",
    "This is required for specification for the `initializeGenomicElement` function in SLiM."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_map.query('type == \"exon\"')[['slim_start', 'slim_end']].to_csv('../clean_data/exome_and_sites_exon_coordinates.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br><br><br><br><br>\n",
    "# Testing of the recombination map building functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sample 5 exons/sites for each chromosome to make a testing data set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_data = exons_and_sites.groupby('chrom') \\\n",
    "                           .apply(pd.DataFrame.head) \\\n",
    "                           .sort_values(by=['chrom', 'start']) \\\n",
    "                           .reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>recomb_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>134900</td>\n",
       "      <td>135802</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>137620</td>\n",
       "      <td>139379</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>367639</td>\n",
       "      <td>368634</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>621058</td>\n",
       "      <td>622053</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end  type   recomb_rate\n",
       "0      1   69090   70008  exon  1.000000e-08\n",
       "1      1  134900  135802  exon  1.000000e-08\n",
       "2      1  137620  139379  exon  1.000000e-08\n",
       "3      1  367639  368634  exon  1.000000e-08\n",
       "4      1  621058  622053  exon  1.000000e-08"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Test `add_recombination_gaps`\n",
    "\n",
    "The function should add a 1bp gap between each exon or site (with a recombination rate at this gap proportional to the distance between the adjacent features) and it should also insert a 0.5 recombination rate between chromosomes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>recomb_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>70008</td>\n",
       "      <td>70009</td>\n",
       "      <td>gap</td>\n",
       "      <td>6.489200e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>134900</td>\n",
       "      <td>135802</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>135802</td>\n",
       "      <td>135803</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.818000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>137620</td>\n",
       "      <td>139379</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>139379</td>\n",
       "      <td>139380</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.282600e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>367639</td>\n",
       "      <td>368634</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>368634</td>\n",
       "      <td>368635</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.524240e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>621058</td>\n",
       "      <td>622053</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>622053</td>\n",
       "      <td>622054</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end  type   recomb_rate\n",
       "0      1   69090   70008  exon  1.000000e-08\n",
       "1      1   70008   70009   gap  6.489200e-04\n",
       "2      1  134900  135802  exon  1.000000e-08\n",
       "3      1  135802  135803   gap  1.818000e-05\n",
       "4      1  137620  139379  exon  1.000000e-08\n",
       "5      1  139379  139380   gap  2.282600e-03\n",
       "6      1  367639  368634  exon  1.000000e-08\n",
       "7      1  368634  368635   gap  2.524240e-03\n",
       "8      1  621058  622053  exon  1.000000e-08\n",
       "9      1  622053  622054   gap  5.000000e-01"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "add_recombination_gaps(test_data.query('chrom == 1'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Test `concatenate_regions`\n",
    "\n",
    "The function should concatenate all regions into a single continuous segment. This segment should start at 0 and its last position should be equal to the total length of all regions - 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>recomb_rate</th>\n",
       "      <th>width</th>\n",
       "      <th>slim_start</th>\n",
       "      <th>slim_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>918</td>\n",
       "      <td>0</td>\n",
       "      <td>917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>134900</td>\n",
       "      <td>135802</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>902</td>\n",
       "      <td>918</td>\n",
       "      <td>1819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>137620</td>\n",
       "      <td>139379</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>1759</td>\n",
       "      <td>1820</td>\n",
       "      <td>3578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>367639</td>\n",
       "      <td>368634</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>995</td>\n",
       "      <td>3579</td>\n",
       "      <td>4573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>621058</td>\n",
       "      <td>622053</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>995</td>\n",
       "      <td>4574</td>\n",
       "      <td>5568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>38813</td>\n",
       "      <td>41627</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>2814</td>\n",
       "      <td>5569</td>\n",
       "      <td>8382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>45439</td>\n",
       "      <td>46385</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>946</td>\n",
       "      <td>8383</td>\n",
       "      <td>9328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>52317</td>\n",
       "      <td>52318</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>9329</td>\n",
       "      <td>9329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2</td>\n",
       "      <td>74220</td>\n",
       "      <td>74221</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>9330</td>\n",
       "      <td>9330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2</td>\n",
       "      <td>128193</td>\n",
       "      <td>128194</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>9331</td>\n",
       "      <td>9331</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end  type   recomb_rate  width  slim_start  slim_end\n",
       "0      1   69090   70008  exon  1.000000e-08    918           0       917\n",
       "1      1  134900  135802  exon  1.000000e-08    902         918      1819\n",
       "2      1  137620  139379  exon  1.000000e-08   1759        1820      3578\n",
       "3      1  367639  368634  exon  1.000000e-08    995        3579      4573\n",
       "4      1  621058  622053  exon  1.000000e-08    995        4574      5568\n",
       "5      2   38813   41627  exon  1.000000e-08   2814        5569      8382\n",
       "6      2   45439   46385  exon  1.000000e-08    946        8383      9328\n",
       "7      2   52317   52318   snp           NaN      1        9329      9329\n",
       "8      2   74220   74221   snp           NaN      1        9330      9330\n",
       "9      2  128193  128194   snp           NaN      1        9331      9331"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concatenate_regions(test_data.query('chrom == 1 | chrom == 2'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What is the correct end of the single continuous segment?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9331"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum((test_data.query('chrom == 1 | chrom == 2')).end -\n",
    "    (test_data.query('chrom == 1 | chrom == 2')).start) - 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Test `create_recomb_map`\n",
    "\n",
    "The function should create a recombination map of all exons and sites in the given DataFrame. A 1 bp gap should be inserted between all regions, and a similar gap with a recombination rate of 0.5 should be inserted between each \"adjacent\" pair of chromosomes.\n",
    "\n",
    "SNPs have no recombination rate and they are not going to be a part of the recombination map output file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>recomb_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>70008</td>\n",
       "      <td>70009</td>\n",
       "      <td>gap</td>\n",
       "      <td>6.489200e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>134900</td>\n",
       "      <td>135802</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>135802</td>\n",
       "      <td>135803</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.818000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>137620</td>\n",
       "      <td>139379</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>139379</td>\n",
       "      <td>139380</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.282600e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>367639</td>\n",
       "      <td>368634</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>368634</td>\n",
       "      <td>368635</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.524240e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>621058</td>\n",
       "      <td>622053</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>622053</td>\n",
       "      <td>622054</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2</td>\n",
       "      <td>38813</td>\n",
       "      <td>41627</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2</td>\n",
       "      <td>41627</td>\n",
       "      <td>41628</td>\n",
       "      <td>gap</td>\n",
       "      <td>3.812000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2</td>\n",
       "      <td>45439</td>\n",
       "      <td>46385</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2</td>\n",
       "      <td>46385</td>\n",
       "      <td>46386</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.932000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2</td>\n",
       "      <td>52317</td>\n",
       "      <td>52318</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2</td>\n",
       "      <td>52318</td>\n",
       "      <td>52319</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.190200e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>2</td>\n",
       "      <td>74220</td>\n",
       "      <td>74221</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>2</td>\n",
       "      <td>74221</td>\n",
       "      <td>74222</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.397200e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>2</td>\n",
       "      <td>128193</td>\n",
       "      <td>128194</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>2</td>\n",
       "      <td>128194</td>\n",
       "      <td>128195</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>3</td>\n",
       "      <td>124408</td>\n",
       "      <td>124409</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>3</td>\n",
       "      <td>124409</td>\n",
       "      <td>124410</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.138690e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>3</td>\n",
       "      <td>238278</td>\n",
       "      <td>238746</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>3</td>\n",
       "      <td>238746</td>\n",
       "      <td>238747</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.260000e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>3</td>\n",
       "      <td>238972</td>\n",
       "      <td>239051</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>3</td>\n",
       "      <td>239051</td>\n",
       "      <td>239052</td>\n",
       "      <td>gap</td>\n",
       "      <td>6.380000e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>3</td>\n",
       "      <td>239689</td>\n",
       "      <td>239775</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>3</td>\n",
       "      <td>239775</td>\n",
       "      <td>239776</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.019700e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>3</td>\n",
       "      <td>249972</td>\n",
       "      <td>249973</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>3</td>\n",
       "      <td>249973</td>\n",
       "      <td>249974</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>19</td>\n",
       "      <td>282809</td>\n",
       "      <td>282810</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>20</td>\n",
       "      <td>68350</td>\n",
       "      <td>68408</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>20</td>\n",
       "      <td>68408</td>\n",
       "      <td>68409</td>\n",
       "      <td>gap</td>\n",
       "      <td>8.237000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>20</td>\n",
       "      <td>76645</td>\n",
       "      <td>77174</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>20</td>\n",
       "      <td>77174</td>\n",
       "      <td>77175</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.586300e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>20</td>\n",
       "      <td>103037</td>\n",
       "      <td>103038</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>20</td>\n",
       "      <td>103038</td>\n",
       "      <td>103039</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.997100e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>20</td>\n",
       "      <td>123009</td>\n",
       "      <td>123327</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>20</td>\n",
       "      <td>123327</td>\n",
       "      <td>123328</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.728000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>20</td>\n",
       "      <td>126055</td>\n",
       "      <td>126392</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>20</td>\n",
       "      <td>126392</td>\n",
       "      <td>126393</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>21</td>\n",
       "      <td>10906200</td>\n",
       "      <td>10907040</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>21</td>\n",
       "      <td>10907040</td>\n",
       "      <td>10907041</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.784000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>21</td>\n",
       "      <td>10908824</td>\n",
       "      <td>10908895</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>21</td>\n",
       "      <td>10908895</td>\n",
       "      <td>10908896</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.411000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>21</td>\n",
       "      <td>10910306</td>\n",
       "      <td>10910399</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>21</td>\n",
       "      <td>10910399</td>\n",
       "      <td>10910400</td>\n",
       "      <td>gap</td>\n",
       "      <td>3.963000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>21</td>\n",
       "      <td>10914362</td>\n",
       "      <td>10914442</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>21</td>\n",
       "      <td>10914442</td>\n",
       "      <td>10914443</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.927000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>21</td>\n",
       "      <td>10916369</td>\n",
       "      <td>10916475</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>21</td>\n",
       "      <td>10916475</td>\n",
       "      <td>10916476</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>22</td>\n",
       "      <td>16256440</td>\n",
       "      <td>16256677</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>22</td>\n",
       "      <td>16256677</td>\n",
       "      <td>16256678</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.507000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>22</td>\n",
       "      <td>16258184</td>\n",
       "      <td>16258303</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>22</td>\n",
       "      <td>16258303</td>\n",
       "      <td>16258304</td>\n",
       "      <td>gap</td>\n",
       "      <td>8.625000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>22</td>\n",
       "      <td>16266928</td>\n",
       "      <td>16267095</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>22</td>\n",
       "      <td>16267095</td>\n",
       "      <td>16267096</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.041000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>22</td>\n",
       "      <td>16268136</td>\n",
       "      <td>16268181</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>22</td>\n",
       "      <td>16268181</td>\n",
       "      <td>16268182</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.691000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>22</td>\n",
       "      <td>16269872</td>\n",
       "      <td>16269943</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>219 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom     start       end  type   recomb_rate\n",
       "0        1     69090     70008  exon  1.000000e-08\n",
       "1        1     70008     70009   gap  6.489200e-04\n",
       "2        1    134900    135802  exon  1.000000e-08\n",
       "3        1    135802    135803   gap  1.818000e-05\n",
       "4        1    137620    139379  exon  1.000000e-08\n",
       "5        1    139379    139380   gap  2.282600e-03\n",
       "6        1    367639    368634  exon  1.000000e-08\n",
       "7        1    368634    368635   gap  2.524240e-03\n",
       "8        1    621058    622053  exon  1.000000e-08\n",
       "9        1    622053    622054   gap  5.000000e-01\n",
       "10       2     38813     41627  exon  1.000000e-08\n",
       "11       2     41627     41628   gap  3.812000e-05\n",
       "12       2     45439     46385  exon  1.000000e-08\n",
       "13       2     46385     46386   gap  5.932000e-05\n",
       "14       2     52317     52318   snp           NaN\n",
       "15       2     52318     52319   gap  2.190200e-04\n",
       "16       2     74220     74221   snp           NaN\n",
       "17       2     74221     74222   gap  5.397200e-04\n",
       "18       2    128193    128194   snp           NaN\n",
       "19       2    128194    128195   gap  5.000000e-01\n",
       "20       3    124408    124409   snp           NaN\n",
       "21       3    124409    124410   gap  1.138690e-03\n",
       "22       3    238278    238746  exon  1.000000e-08\n",
       "23       3    238746    238747   gap  2.260000e-06\n",
       "24       3    238972    239051  exon  1.000000e-08\n",
       "25       3    239051    239052   gap  6.380000e-06\n",
       "26       3    239689    239775  exon  1.000000e-08\n",
       "27       3    239775    239776   gap  1.019700e-04\n",
       "28       3    249972    249973   snp           NaN\n",
       "29       3    249973    249974   gap  5.000000e-01\n",
       "..     ...       ...       ...   ...           ...\n",
       "189     19    282809    282810   gap  5.000000e-01\n",
       "190     20     68350     68408  exon  1.000000e-08\n",
       "191     20     68408     68409   gap  8.237000e-05\n",
       "192     20     76645     77174  exon  1.000000e-08\n",
       "193     20     77174     77175   gap  2.586300e-04\n",
       "194     20    103037    103038   snp           NaN\n",
       "195     20    103038    103039   gap  1.997100e-04\n",
       "196     20    123009    123327  exon  1.000000e-08\n",
       "197     20    123327    123328   gap  2.728000e-05\n",
       "198     20    126055    126392  exon  1.000000e-08\n",
       "199     20    126392    126393   gap  5.000000e-01\n",
       "200     21  10906200  10907040  exon  1.000000e-08\n",
       "201     21  10907040  10907041   gap  1.784000e-05\n",
       "202     21  10908824  10908895  exon  1.000000e-08\n",
       "203     21  10908895  10908896   gap  1.411000e-05\n",
       "204     21  10910306  10910399  exon  1.000000e-08\n",
       "205     21  10910399  10910400   gap  3.963000e-05\n",
       "206     21  10914362  10914442  exon  1.000000e-08\n",
       "207     21  10914442  10914443   gap  1.927000e-05\n",
       "208     21  10916369  10916475  exon  1.000000e-08\n",
       "209     21  10916475  10916476   gap  5.000000e-01\n",
       "210     22  16256440  16256677  exon  1.000000e-08\n",
       "211     22  16256677  16256678   gap  1.507000e-05\n",
       "212     22  16258184  16258303  exon  1.000000e-08\n",
       "213     22  16258303  16258304   gap  8.625000e-05\n",
       "214     22  16266928  16267095  exon  1.000000e-08\n",
       "215     22  16267095  16267096   gap  1.041000e-05\n",
       "216     22  16268136  16268181  exon  1.000000e-08\n",
       "217     22  16268181  16268182   gap  1.691000e-05\n",
       "218     22  16269872  16269943  exon  1.000000e-08\n",
       "\n",
       "[219 rows x 5 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "create_recomb_map(test_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Test `create_recomb_map` and `concatenate_regoins` together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>recomb_rate</th>\n",
       "      <th>width</th>\n",
       "      <th>slim_start</th>\n",
       "      <th>slim_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>918</td>\n",
       "      <td>0</td>\n",
       "      <td>917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>70008</td>\n",
       "      <td>70009</td>\n",
       "      <td>gap</td>\n",
       "      <td>6.489200e-04</td>\n",
       "      <td>1</td>\n",
       "      <td>918</td>\n",
       "      <td>918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>134900</td>\n",
       "      <td>135802</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>902</td>\n",
       "      <td>919</td>\n",
       "      <td>1820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>135802</td>\n",
       "      <td>135803</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.818000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>1821</td>\n",
       "      <td>1821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>137620</td>\n",
       "      <td>139379</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>1759</td>\n",
       "      <td>1822</td>\n",
       "      <td>3580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>139379</td>\n",
       "      <td>139380</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.282600e-03</td>\n",
       "      <td>1</td>\n",
       "      <td>3581</td>\n",
       "      <td>3581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>367639</td>\n",
       "      <td>368634</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>995</td>\n",
       "      <td>3582</td>\n",
       "      <td>4576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>368634</td>\n",
       "      <td>368635</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.524240e-03</td>\n",
       "      <td>1</td>\n",
       "      <td>4577</td>\n",
       "      <td>4577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>621058</td>\n",
       "      <td>622053</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>995</td>\n",
       "      <td>4578</td>\n",
       "      <td>5572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>622053</td>\n",
       "      <td>622054</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>5573</td>\n",
       "      <td>5573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2</td>\n",
       "      <td>38813</td>\n",
       "      <td>41627</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>2814</td>\n",
       "      <td>5574</td>\n",
       "      <td>8387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2</td>\n",
       "      <td>41627</td>\n",
       "      <td>41628</td>\n",
       "      <td>gap</td>\n",
       "      <td>3.812000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>8388</td>\n",
       "      <td>8388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2</td>\n",
       "      <td>45439</td>\n",
       "      <td>46385</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>946</td>\n",
       "      <td>8389</td>\n",
       "      <td>9334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2</td>\n",
       "      <td>46385</td>\n",
       "      <td>46386</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.932000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>9335</td>\n",
       "      <td>9335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2</td>\n",
       "      <td>52317</td>\n",
       "      <td>52318</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>9336</td>\n",
       "      <td>9336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2</td>\n",
       "      <td>52318</td>\n",
       "      <td>52319</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.190200e-04</td>\n",
       "      <td>1</td>\n",
       "      <td>9337</td>\n",
       "      <td>9337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>2</td>\n",
       "      <td>74220</td>\n",
       "      <td>74221</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>9338</td>\n",
       "      <td>9338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>2</td>\n",
       "      <td>74221</td>\n",
       "      <td>74222</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.397200e-04</td>\n",
       "      <td>1</td>\n",
       "      <td>9339</td>\n",
       "      <td>9339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>2</td>\n",
       "      <td>128193</td>\n",
       "      <td>128194</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>9340</td>\n",
       "      <td>9340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>2</td>\n",
       "      <td>128194</td>\n",
       "      <td>128195</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>9341</td>\n",
       "      <td>9341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>3</td>\n",
       "      <td>124408</td>\n",
       "      <td>124409</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>9342</td>\n",
       "      <td>9342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>3</td>\n",
       "      <td>124409</td>\n",
       "      <td>124410</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.138690e-03</td>\n",
       "      <td>1</td>\n",
       "      <td>9343</td>\n",
       "      <td>9343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>3</td>\n",
       "      <td>238278</td>\n",
       "      <td>238746</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>468</td>\n",
       "      <td>9344</td>\n",
       "      <td>9811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>3</td>\n",
       "      <td>238746</td>\n",
       "      <td>238747</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.260000e-06</td>\n",
       "      <td>1</td>\n",
       "      <td>9812</td>\n",
       "      <td>9812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>3</td>\n",
       "      <td>238972</td>\n",
       "      <td>239051</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>79</td>\n",
       "      <td>9813</td>\n",
       "      <td>9891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>3</td>\n",
       "      <td>239051</td>\n",
       "      <td>239052</td>\n",
       "      <td>gap</td>\n",
       "      <td>6.380000e-06</td>\n",
       "      <td>1</td>\n",
       "      <td>9892</td>\n",
       "      <td>9892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>3</td>\n",
       "      <td>239689</td>\n",
       "      <td>239775</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>86</td>\n",
       "      <td>9893</td>\n",
       "      <td>9978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>3</td>\n",
       "      <td>239775</td>\n",
       "      <td>239776</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.019700e-04</td>\n",
       "      <td>1</td>\n",
       "      <td>9979</td>\n",
       "      <td>9979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>3</td>\n",
       "      <td>249972</td>\n",
       "      <td>249973</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>9980</td>\n",
       "      <td>9980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>3</td>\n",
       "      <td>249973</td>\n",
       "      <td>249974</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>9981</td>\n",
       "      <td>9981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>19</td>\n",
       "      <td>282809</td>\n",
       "      <td>282810</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>38254</td>\n",
       "      <td>38254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>20</td>\n",
       "      <td>68350</td>\n",
       "      <td>68408</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>58</td>\n",
       "      <td>38255</td>\n",
       "      <td>38312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>20</td>\n",
       "      <td>68408</td>\n",
       "      <td>68409</td>\n",
       "      <td>gap</td>\n",
       "      <td>8.237000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>38313</td>\n",
       "      <td>38313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>20</td>\n",
       "      <td>76645</td>\n",
       "      <td>77174</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>529</td>\n",
       "      <td>38314</td>\n",
       "      <td>38842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>20</td>\n",
       "      <td>77174</td>\n",
       "      <td>77175</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.586300e-04</td>\n",
       "      <td>1</td>\n",
       "      <td>38843</td>\n",
       "      <td>38843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>20</td>\n",
       "      <td>103037</td>\n",
       "      <td>103038</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>38844</td>\n",
       "      <td>38844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>20</td>\n",
       "      <td>103038</td>\n",
       "      <td>103039</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.997100e-04</td>\n",
       "      <td>1</td>\n",
       "      <td>38845</td>\n",
       "      <td>38845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>20</td>\n",
       "      <td>123009</td>\n",
       "      <td>123327</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>318</td>\n",
       "      <td>38846</td>\n",
       "      <td>39163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>20</td>\n",
       "      <td>123327</td>\n",
       "      <td>123328</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.728000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>39164</td>\n",
       "      <td>39164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>20</td>\n",
       "      <td>126055</td>\n",
       "      <td>126392</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>337</td>\n",
       "      <td>39165</td>\n",
       "      <td>39501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>20</td>\n",
       "      <td>126392</td>\n",
       "      <td>126393</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>39502</td>\n",
       "      <td>39502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>21</td>\n",
       "      <td>10906200</td>\n",
       "      <td>10907040</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>840</td>\n",
       "      <td>39503</td>\n",
       "      <td>40342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>21</td>\n",
       "      <td>10907040</td>\n",
       "      <td>10907041</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.784000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>40343</td>\n",
       "      <td>40343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>21</td>\n",
       "      <td>10908824</td>\n",
       "      <td>10908895</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>71</td>\n",
       "      <td>40344</td>\n",
       "      <td>40414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>21</td>\n",
       "      <td>10908895</td>\n",
       "      <td>10908896</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.411000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>40415</td>\n",
       "      <td>40415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>21</td>\n",
       "      <td>10910306</td>\n",
       "      <td>10910399</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>93</td>\n",
       "      <td>40416</td>\n",
       "      <td>40508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>21</td>\n",
       "      <td>10910399</td>\n",
       "      <td>10910400</td>\n",
       "      <td>gap</td>\n",
       "      <td>3.963000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>40509</td>\n",
       "      <td>40509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>21</td>\n",
       "      <td>10914362</td>\n",
       "      <td>10914442</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>80</td>\n",
       "      <td>40510</td>\n",
       "      <td>40589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>21</td>\n",
       "      <td>10914442</td>\n",
       "      <td>10914443</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.927000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>40590</td>\n",
       "      <td>40590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>21</td>\n",
       "      <td>10916369</td>\n",
       "      <td>10916475</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>106</td>\n",
       "      <td>40591</td>\n",
       "      <td>40696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>21</td>\n",
       "      <td>10916475</td>\n",
       "      <td>10916476</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>40697</td>\n",
       "      <td>40697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>22</td>\n",
       "      <td>16256440</td>\n",
       "      <td>16256677</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>237</td>\n",
       "      <td>40698</td>\n",
       "      <td>40934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>22</td>\n",
       "      <td>16256677</td>\n",
       "      <td>16256678</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.507000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>40935</td>\n",
       "      <td>40935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>22</td>\n",
       "      <td>16258184</td>\n",
       "      <td>16258303</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>119</td>\n",
       "      <td>40936</td>\n",
       "      <td>41054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>22</td>\n",
       "      <td>16258303</td>\n",
       "      <td>16258304</td>\n",
       "      <td>gap</td>\n",
       "      <td>8.625000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>41055</td>\n",
       "      <td>41055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>22</td>\n",
       "      <td>16266928</td>\n",
       "      <td>16267095</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>167</td>\n",
       "      <td>41056</td>\n",
       "      <td>41222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>22</td>\n",
       "      <td>16267095</td>\n",
       "      <td>16267096</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.041000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>41223</td>\n",
       "      <td>41223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>22</td>\n",
       "      <td>16268136</td>\n",
       "      <td>16268181</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>45</td>\n",
       "      <td>41224</td>\n",
       "      <td>41268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>22</td>\n",
       "      <td>16268181</td>\n",
       "      <td>16268182</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.691000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>41269</td>\n",
       "      <td>41269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>22</td>\n",
       "      <td>16269872</td>\n",
       "      <td>16269943</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>71</td>\n",
       "      <td>41270</td>\n",
       "      <td>41340</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>219 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom     start       end  type   recomb_rate  width  slim_start  \\\n",
       "0        1     69090     70008  exon  1.000000e-08    918           0   \n",
       "1        1     70008     70009   gap  6.489200e-04      1         918   \n",
       "2        1    134900    135802  exon  1.000000e-08    902         919   \n",
       "3        1    135802    135803   gap  1.818000e-05      1        1821   \n",
       "4        1    137620    139379  exon  1.000000e-08   1759        1822   \n",
       "5        1    139379    139380   gap  2.282600e-03      1        3581   \n",
       "6        1    367639    368634  exon  1.000000e-08    995        3582   \n",
       "7        1    368634    368635   gap  2.524240e-03      1        4577   \n",
       "8        1    621058    622053  exon  1.000000e-08    995        4578   \n",
       "9        1    622053    622054   gap  5.000000e-01      1        5573   \n",
       "10       2     38813     41627  exon  1.000000e-08   2814        5574   \n",
       "11       2     41627     41628   gap  3.812000e-05      1        8388   \n",
       "12       2     45439     46385  exon  1.000000e-08    946        8389   \n",
       "13       2     46385     46386   gap  5.932000e-05      1        9335   \n",
       "14       2     52317     52318   snp           NaN      1        9336   \n",
       "15       2     52318     52319   gap  2.190200e-04      1        9337   \n",
       "16       2     74220     74221   snp           NaN      1        9338   \n",
       "17       2     74221     74222   gap  5.397200e-04      1        9339   \n",
       "18       2    128193    128194   snp           NaN      1        9340   \n",
       "19       2    128194    128195   gap  5.000000e-01      1        9341   \n",
       "20       3    124408    124409   snp           NaN      1        9342   \n",
       "21       3    124409    124410   gap  1.138690e-03      1        9343   \n",
       "22       3    238278    238746  exon  1.000000e-08    468        9344   \n",
       "23       3    238746    238747   gap  2.260000e-06      1        9812   \n",
       "24       3    238972    239051  exon  1.000000e-08     79        9813   \n",
       "25       3    239051    239052   gap  6.380000e-06      1        9892   \n",
       "26       3    239689    239775  exon  1.000000e-08     86        9893   \n",
       "27       3    239775    239776   gap  1.019700e-04      1        9979   \n",
       "28       3    249972    249973   snp           NaN      1        9980   \n",
       "29       3    249973    249974   gap  5.000000e-01      1        9981   \n",
       "..     ...       ...       ...   ...           ...    ...         ...   \n",
       "189     19    282809    282810   gap  5.000000e-01      1       38254   \n",
       "190     20     68350     68408  exon  1.000000e-08     58       38255   \n",
       "191     20     68408     68409   gap  8.237000e-05      1       38313   \n",
       "192     20     76645     77174  exon  1.000000e-08    529       38314   \n",
       "193     20     77174     77175   gap  2.586300e-04      1       38843   \n",
       "194     20    103037    103038   snp           NaN      1       38844   \n",
       "195     20    103038    103039   gap  1.997100e-04      1       38845   \n",
       "196     20    123009    123327  exon  1.000000e-08    318       38846   \n",
       "197     20    123327    123328   gap  2.728000e-05      1       39164   \n",
       "198     20    126055    126392  exon  1.000000e-08    337       39165   \n",
       "199     20    126392    126393   gap  5.000000e-01      1       39502   \n",
       "200     21  10906200  10907040  exon  1.000000e-08    840       39503   \n",
       "201     21  10907040  10907041   gap  1.784000e-05      1       40343   \n",
       "202     21  10908824  10908895  exon  1.000000e-08     71       40344   \n",
       "203     21  10908895  10908896   gap  1.411000e-05      1       40415   \n",
       "204     21  10910306  10910399  exon  1.000000e-08     93       40416   \n",
       "205     21  10910399  10910400   gap  3.963000e-05      1       40509   \n",
       "206     21  10914362  10914442  exon  1.000000e-08     80       40510   \n",
       "207     21  10914442  10914443   gap  1.927000e-05      1       40590   \n",
       "208     21  10916369  10916475  exon  1.000000e-08    106       40591   \n",
       "209     21  10916475  10916476   gap  5.000000e-01      1       40697   \n",
       "210     22  16256440  16256677  exon  1.000000e-08    237       40698   \n",
       "211     22  16256677  16256678   gap  1.507000e-05      1       40935   \n",
       "212     22  16258184  16258303  exon  1.000000e-08    119       40936   \n",
       "213     22  16258303  16258304   gap  8.625000e-05      1       41055   \n",
       "214     22  16266928  16267095  exon  1.000000e-08    167       41056   \n",
       "215     22  16267095  16267096   gap  1.041000e-05      1       41223   \n",
       "216     22  16268136  16268181  exon  1.000000e-08     45       41224   \n",
       "217     22  16268181  16268182   gap  1.691000e-05      1       41269   \n",
       "218     22  16269872  16269943  exon  1.000000e-08     71       41270   \n",
       "\n",
       "     slim_end  \n",
       "0         917  \n",
       "1         918  \n",
       "2        1820  \n",
       "3        1821  \n",
       "4        3580  \n",
       "5        3581  \n",
       "6        4576  \n",
       "7        4577  \n",
       "8        5572  \n",
       "9        5573  \n",
       "10       8387  \n",
       "11       8388  \n",
       "12       9334  \n",
       "13       9335  \n",
       "14       9336  \n",
       "15       9337  \n",
       "16       9338  \n",
       "17       9339  \n",
       "18       9340  \n",
       "19       9341  \n",
       "20       9342  \n",
       "21       9343  \n",
       "22       9811  \n",
       "23       9812  \n",
       "24       9891  \n",
       "25       9892  \n",
       "26       9978  \n",
       "27       9979  \n",
       "28       9980  \n",
       "29       9981  \n",
       "..        ...  \n",
       "189     38254  \n",
       "190     38312  \n",
       "191     38313  \n",
       "192     38842  \n",
       "193     38843  \n",
       "194     38844  \n",
       "195     38845  \n",
       "196     39163  \n",
       "197     39164  \n",
       "198     39501  \n",
       "199     39502  \n",
       "200     40342  \n",
       "201     40343  \n",
       "202     40414  \n",
       "203     40415  \n",
       "204     40508  \n",
       "205     40509  \n",
       "206     40589  \n",
       "207     40590  \n",
       "208     40696  \n",
       "209     40697  \n",
       "210     40934  \n",
       "211     40935  \n",
       "212     41054  \n",
       "213     41055  \n",
       "214     41222  \n",
       "215     41223  \n",
       "216     41268  \n",
       "217     41269  \n",
       "218     41340  \n",
       "\n",
       "[219 rows x 8 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concatenate_regions(create_recomb_map(test_data))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
