{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating SLiM exome recombination maps\n",
    "\n",
    "This notebook generates a recombination map for SLiM that implements a realistic exome structure and includes the positions of sites from the archaic admixture array. It also saves the positions of array sites and coordinates of exons separately."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from pybedtools import BedTool\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "plt.style.use('ggplot')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "RECOMB_RATE = 1e-8 # crossovers per bp per generation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exon coordinates processing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First download the GTF annotations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gtf = pd.read_table('ftp://ftp.ensembl.org/pub/release-75/gtf/homo_sapiens/Homo_sapiens.GRCh37.75.gtf.gz',\n",
    "                    header=None, sep='\\t', skipinitialspace=True, skiprows=5, compression='gzip',\n",
    "                    names=['chrom', 'source', 'feature', 'start', 'end',\n",
    "                           'score', 'strand', 'frame', 'attribute'], low_memory=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What does the data look like?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14412</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972\"; transcript_id \"ENST...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972\"; transcript_id \"ENST...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972\"; transcript_id \"ENST...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972\"; transcript_id \"ENST...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom                source     feature  start    end score strand frame  \\\n",
       "0     1            pseudogene        gene  11869  14412     .      +     .   \n",
       "1     1  processed_transcript  transcript  11869  14409     .      +     .   \n",
       "2     1  processed_transcript        exon  11869  12227     .      +     .   \n",
       "3     1  processed_transcript        exon  12613  12721     .      +     .   \n",
       "4     1  processed_transcript        exon  13221  14409     .      +     .   \n",
       "\n",
       "                                           attribute  \n",
       "0  gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"...  \n",
       "1  gene_id \"ENSG00000223972\"; transcript_id \"ENST...  \n",
       "2  gene_id \"ENSG00000223972\"; transcript_id \"ENST...  \n",
       "3  gene_id \"ENSG00000223972\"; transcript_id \"ENST...  \n",
       "4  gene_id \"ENSG00000223972\"; transcript_id \"ENST...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "protein_coding                        2001054\n",
       "nonsense_mediated_decay                293471\n",
       "processed_transcript                   173401\n",
       "retained_intron                        150034\n",
       "lincRNA                                 55928\n",
       "antisense                               45811\n",
       "processed_pseudogene                    24517\n",
       "pseudogene                              17830\n",
       "unprocessed_pseudogene                  13763\n",
       "miRNA                                   10209\n",
       "transcribed_unprocessed_pseudogene       7931\n",
       "misc_RNA                                 6554\n",
       "snRNA                                    6215\n",
       "snoRNA                                   4791\n",
       "sense_intronic                           3241\n",
       "polymorphic_pseudogene                   1906\n",
       "rRNA                                     1706\n",
       "unitary_pseudogene                       1467\n",
       "sense_overlapping                        1463\n",
       "IG_V_gene                                1277\n",
       "transcribed_processed_pseudogene         1182\n",
       "non_stop_decay                           1132\n",
       "TR_V_gene                                1019\n",
       "IG_V_pseudogene                           912\n",
       "TR_J_gene                                 328\n",
       "IG_C_gene                                 313\n",
       "IG_D_gene                                 260\n",
       "TR_V_pseudogene                           147\n",
       "3prime_overlapping_ncrna                  116\n",
       "IG_J_gene                                 102\n",
       "Mt_tRNA                                    66\n",
       "TR_C_gene                                  64\n",
       "IG_C_pseudogene                            51\n",
       "IG_J_pseudogene                            18\n",
       "TR_J_pseudogene                            12\n",
       "TR_D_gene                                  12\n",
       "Mt_rRNA                                     6\n",
       "translated_processed_pseudogene             3\n",
       "Name: source, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.source.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "exon              1306656\n",
       "CDS                791856\n",
       "UTR                304070\n",
       "transcript         215170\n",
       "stop_codon          73411\n",
       "start_codon         73358\n",
       "gene                63677\n",
       "Selenocysteine        114\n",
       "Name: feature, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.feature.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Take only autosomal protein-coding regions, since we won't be simulating sex chromosomes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "AUTOSOMES = [str(i + 1) for i in range(22)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "exons = gtf[gtf.chrom.isin(AUTOSOMES) &\n",
    "           (gtf.source == \"protein_coding\") &\n",
    "           (gtf.feature == \"exon\")]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge the overlapping exons:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "exons = BedTool.from_dataframe(exons).sort().merge().to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "exons['type'] = 'exon'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert chromosome IDs to integers and sort coordinates:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "exons.chrom = exons.chrom.astype(int)\n",
    "exons.sort_values(by=['chrom', 'start'], inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What is the distribution of exon lengths?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "lengths = exons.end - exons.start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    214193.000000\n",
       "mean        337.708039\n",
       "std         776.506070\n",
       "min           1.000000\n",
       "25%          96.000000\n",
       "50%         138.000000\n",
       "75%         217.000000\n",
       "max       24409.000000\n",
       "dtype: float64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lengths.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZsAAAEECAYAAAD+qJluAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3X9sW+d97/E3SUVNKZ6IOZTZSfE8t6Y931KRnYXaHG+R\nf6TAkCgDjK4QlgZZZdjIddPOHrG5qa77AxdN7bvGiUz9zFAF/TENC9Q4NFbsRwtEkp3WwiDVUarQ\nzTpWcRAtVmWRlSpWdG2KvH9oOrQiJ44lHcuSPy/AEM9DnuPn+YLyh89zjnkcuVwuh4iIiI2cS90B\nERFZ+RQ2IiJiO4WNiIjYTmEjIiK2U9iIiIjtFDYiImK7gmu9oLW1lTNnzlBcXMzRo0cBiMfjPP/8\n80xNTeFyudi7dy/r1q0DIBqN0tXVhcvlora2lk2bNgEwODhIS0sLly9f5p577qG2thaATCZDU1MT\ng4ODGIZBOBympKQEgO7ubqLRKACf/OQn2bZt26IXQERE7HfNmc2OHTs4dOjQrLZ//Md/5C/+4i/4\nxje+QU1NDe3t7QAMDQ3R09NDfX09dXV1tLW1MfPfeNra2ti3bx+RSITz58/T398PQGdnJx6Ph4aG\nBqqrq61jpVIpjh8/zpEjRzh8+DAvvvgik5OTH2hQsVjsg1dghVMt8lSLPNUiT7XIs7MW1wybjRs3\nUlRUNKvN6/Va//D/5je/4c477wSgr6+PrVu34nK58Pv9lJaWEo/HGRsbI51OEwgEAKiqqqK3txeA\n3t5ea8ayZcsWXn/9dQBee+01KioqcLvdFBUVUVFRYQXUtejNk6da5KkWeapFnmqRZ2ctrrmMdjWP\nPvooX/7yl/nud78LwNe+9jUAkskkGzZssF5nmibJZBKXy4XP57PafT4fyWTS2mfmOafTidvtJpVK\nzWq/8lgiIrL8zOsCgdbWVnbv3k1rayuf+cxnaG1tXbQO6dtzRERWnnnNbOLxOF/+8peB6aWv5557\nDpiefYyOjlqvSyQSmKaJaZokEok57TP7zGxns1nS6TQejwfTNGdN6RKJBOXl5VftTywWm/Xampqa\n+QxrRVIt8lSLPNUiT7XIq6mpoaOjw9oOBoMEg8FFOfYHCptcLjdrxvE7v/M7nD17lo9//OMMDAxQ\nWloKQCgUoqGhgYcffphkMsnw8DCBQACHw4Hb7SYej7Nu3TpOnTrFgw8+aO1z8uRJ1q9fT09PjxUo\nmzZt4oUXXmBycpJsNsvAwACPPvroVft3tYK8884711+NFcgwDCYmJpa6GzcF1SJPtchTLfLKysps\nC1/Htb71ORKJcPbsWSYmJiguLqampobf+73fo62tjUwmw2233cbevXv56Ec/Ckxf+tzZ2UlBQcGc\nS5+bm5utS593794NwOXLl2lsbOTcuXMYhsGBAwfw+/3A9KXPL730Eg6H47ovfVbYTNMvUp5qkada\n5KkWeWVlZbYd+5phs1wpbKbpFylPtchTLfJUizw7w0bfICAiIrZT2IiIiO0UNiIiYjuFjYiI2E5h\nIyIitlPYiIiI7RQ2IiJiO4WNiIjYTmEjIiK2U9iIiIjt5vWtz8uB6+fTN2Fz+FaR8X1kiXsjInJr\nW7Fhc+np/wNA4eMHQWEjIrKktIwmIiK2U9iIiIjtFDYiImI7hY2IiNjumhcItLa2cubMGYqLizl6\n9KjV/m//9m/88Ic/xOl08gd/8AfWLZuj0ShdXV24XK45d+psaWmx7tRZW1sLQCaToampicHBQQzD\nIBwOU1JSAkzfqTMajQJc9506RUTk5nHNmc2OHTs4dOjQrLZYLMZPfvITjh49yjPPPMOf/dmfATA0\nNERPTw/19fXU1dXR1tbGzI1A29ra2LdvH5FIhPPnz9Pf3w9AZ2cnHo+HhoYGqquraW9vByCVSnH8\n+HGOHDnC4cOHefHFF5mcnFzUwYuIyI1xzbDZuHEjRUVFs9p++MMfsmvXLlwuFwB33HEHAH19fWzd\nuhWXy4Xf76e0tJR4PM7Y2BjpdJpAIABAVVUVvb29APT29lozli1btvD669P/P+a1116joqICt9tN\nUVERFRUVVkCJiMjyMq//Z3P+/HnOnj3LP/3TP1FYWMhjjz3Gxz72MZLJJBs2bLBeZ5omyWQSl8uF\nz+ez2n0+H8lkEoBkMmk953Q6cbvdpFKpWe1XHktERJafeYXN1NQUv/nNb/j6179OPB7n2Wefpamp\naVE6NLPsdj1isRixWMzarqmpsR47nU7chrEofVuOCgsLMW7h8V9JtchTLfJUi9k6Ojqsx8FgkGAw\nuCjHnVfYlJSU8Ed/9EcABAIBnE4nExMTmKbJ6Oio9bpEIoFpmpimSSKRmNMOWM+Zpkk2myWdTuPx\neDBNc1aAJBIJysvLr9qf9ytINptlYmJiPsNcEQzDuKXHfyXVIk+1yFMt8gzDmPVhfTF9oEufc7nc\nrBlHZWWldW7lnXfeIZPJYBgGoVCI06dPk8lkGBkZYXh4mEAggNfrxe12E4/HyeVynDp1isrKSgBC\noRAnT54EoKenxwqUTZs2MTAwwOTkJKlUioGBAevKNhERWV4cuWusW0UiEc6ePcvExATFxcXU1NRQ\nVVVFS0sL586d47bbbuMv//Iv+fjHPw5MX/rc2dlJQUHBnEufm5ubrUufd+/eDcDly5dpbGzk3Llz\nGIbBgQMH8Pv9wPSlzy+99BIOh+O6L31+uzoETH832lTl/ddfmRVCn9ryVIs81SJPtcgrKyuz7djX\nDJvlSmEzTb9IeapFnmqRp1rk2Rk2+gYBERGxncJGRERsp7ARERHbKWxERMR2ChsREbGdwkZERGyn\nsBEREdspbERExHYKGxERsZ3CRkREbKewERER2ylsRETEdgobERGxncJGRERsp7ARERHbKWxERMR2\nBdd6QWtrK2fOnKG4uJijR4/Oeu773/8+7e3tPP/883g8HmD6Tp1dXV24XK45d+psaWmx7tRZW1sL\nQCaToampicHBQQzDIBwOU1JSAkzfqTMajQJc9506RUTk5nHNmc2OHTs4dOjQnPZEIsFPf/pTKxgA\nhoaG6Onpob6+nrq6Otra2pi5EWhbWxv79u0jEolw/vx5+vv7Aejs7MTj8dDQ0EB1dTXt7e0ApFIp\njh8/zpEjRzh8+DAvvvgik5OTizJoERG5sa4ZNhs3bqSoqGhO+3e+8x0ee+yxWW19fX1s3boVl8uF\n3++ntLSUeDzO2NgY6XSaQCAAQFVVFb29vQD09vZaM5YtW7bw+uuvA/Daa69RUVGB2+2mqKiIiooK\nK6BERGR5mdc5m76+Pnw+H2vWrJnVnkwmZ810TNMkmUySTCbx+XxWu8/nI5lMWvvMPOd0OnG73aRS\nqTn7zBxLRESWn2ues3m3S5cuEY1G+dKXvmRHf6xlt+sRi8WIxWLWdk1NjfXY6XTiNoxF6dtyVFhY\niHELj/9KqkWeapGnWszW0dFhPQ4GgwSDwUU57nWHzfDwMCMjIxw8eJBcLkcymeTJJ5/k8OHDmKbJ\n6Oio9dpEIoFpmpimSSKRmNMOWM+Zpkk2myWdTuPxeDBNc1aAJBIJysvLr9qn9ytINptlYmLieoe5\nYhiGcUuP/0qqRZ5qkada5BmGMevD+mL6QMtouVzOmnGsWbOGb37zmzQ1NdHc3Ixpmvzd3/0dxcXF\nhEIhTp8+TSaTYWRkhOHhYQKBAF6vF7fbTTweJ5fLcerUKSorKwEIhUKcPHkSgJ6eHitQNm3axMDA\nAJOTk6RSKQYGBqwr20REZHm55swmEolw9uxZJiYm+OxnP0tNTQ07duywnnc4HNbj1atXc9999xEO\nhykoKGDv3r3W83v27KG5udm69Hnz5s0A7Ny5k8bGRvbv349hGBw4cAAAj8fDn//5n/PFL34Rh8PB\npz71qateqCAiIjc/R24+J0mWgberQwAUPn6Qqcr7l7g3S0dLBHmqRZ5qkada5JWVldl2bH2DgIiI\n2E5hIyIitlPYiIiI7RQ2IiJiO4WNiIjYTmEjIiK2U9iIiIjtFDYiImI7hY2IiNhOYSMiIrZT2IiI\niO0UNiIiYjuFjYiI2E5hIyIitlPYiIiI7RQ2IiJiu2veqbO1tZUzZ85QXFzM0aNHAWhvb+cnP/kJ\nBQUFfOQjH+GJJ57A7XYDEI1G6erqwuVyUVtba93KeXBwkJaWFutOnbW1tQBkMhmampoYHBzEMAzC\n4TAlJSUAdHd3E41GAfjkJz/Jtm3bFr0AIiJiv2vObHbs2MGhQ4dmtVVUVPDMM8/w9NNPU1payokT\nJwAYGhqip6eH+vp66urqaGtrY+ZGoG1tbezbt49IJML58+fp7+8HoLOzE4/HQ0NDA9XV1bS3twOQ\nSqU4fvw4R44c4fDhw7z44otMTk4u6uBFROTGuGbYbNy4kaKiolltFRUVOJ3Tu65fv55EIgFAX18f\nW7duxeVy4ff7KS0tJR6PMzY2RjqdJhAIAFBVVUVvby8Avb291oxly5YtvP766wC89tprVFRU4Ha7\nKSoqoqKiwgooERFZXhZ8zqarq4t77rkHgGQyaS2BAZimSTKZJJlM4vP5rHafz0cymbT2mXnO6XTi\ndrtJpVJz9pk5loiILD/XPGfzfl566SVcLhd/8id/slj9sZbdrkcsFiMWi1nbNTU11mOn04nbMBal\nb8tRYWEhxi08/iupFnmqRZ5qMVtHR4f1OBgMEgwGF+W48w6b7u5uXn31Vb7yla9YbaZpMjo6am0n\nEglM08Q0TWup7cr2mX1mtrPZLOl0Go/Hg2maswIkkUhQXl5+1b68X0Gy2SwTExPzHeayZxjGLT3+\nK6kWeapFnmqRZxjGrA/ri+kDLaPlcrlZM47+/n7++Z//mS984QvcdtttVnsoFOL06dNkMhlGRkYY\nHh4mEAjg9Xpxu93E43FyuRynTp2isrLS2ufkyZMA9PT0WIGyadMmBgYGmJycJJVKMTAwYF3ZJiIi\ny4sjd411q0gkwtmzZ5mYmKC4uJiamhqi0SiZTMaaeq5fv569e/cC05c+d3Z2UlBQMOfS5+bmZuvS\n5927dwNw+fJlGhsbOXfuHIZhcODAAfx+PzA9e3rppZdwOBzXfenz29UhAAofP8hU5f3XWZaVQ5/a\n8lSLPNUiT7XIKysrs+3Y1wyb5UphM02/SHmqRZ5qkada5NkZNvoGARERsZ3CRkREbKewERER2yls\nRETEdgobERGxncJGRERsp7ARERHbKWxERMR2ChsREbGdwkZERGynsBEREdspbERExHYKGxERsZ3C\nRkREbKewERER213zttCtra2cOXOG4uJijh49CkAqleLYsWNcuHABv99POBzG7XYD0zdP6+rqwuVy\nzbl5WktLi3XztNraWgAymQxNTU0MDg5iGAbhcJiSkhJg+uZp0WgU4LpvniYiIjePa85sduzYwaFD\nh2a1nThxgrvvvptIJEIwGLQCYWhoiJ6eHurr66mrq6Otrc26nXRbWxv79u0jEolw/vx5+vv7Aejs\n7MTj8dDQ0EB1dTXt7e3AdKAdP36cI0eOcPjwYV588UUmJycXdfAiInJjXDNsNm7cSFFR0ay2vr4+\na5axfft2ent7rfatW7ficrnw+/2UlpYSj8cZGxsjnU4TCAQAqKqqsvbp7e21jrVlyxZef/11AF57\n7TUqKipwu90UFRVRUVFhBZSIiCwv8zpnMz4+jtfrBcDr9TI+Pg5AMpm0lsAATNMkmUySTCbx+XxW\nu8/nI5lMWvvMPOd0OnG73aRSqTn7zBxLRESWn0W5QMDhcCzGYQCsZTcREVk5rnmBwNV4vV7Gxsas\nn8XFxcD07GN0dNR6XSKRwDRNTNMkkUjMaZ/ZZ2Y7m82STqfxeDyYpkksFpu1T3l5+VX7E4vFZr22\npqbGeux0OnEbxnyGuSIUFhZi3MLjv5Jqkada5KkWs3V0dFiPg8EgwWBwUY77gcIml8vNmnHce++9\ndHd3s2vXLrq7uwmFQgCEQiEaGhp4+OGHSSaTDA8PEwgEcDgcuN1u4vE469at49SpUzz44IPWPidP\nnmT9+vX09PRYgbJp0yZeeOEFJicnyWazDAwM8Oijj161f+9XkGw2y8TExAevyApjGMYtPf4rqRZ5\nqkWeapFnGMasD+uLyZG7xrpVJBLh7NmzTExMUFxcTE1NDZWVldTX1zM6OsqqVasIh8PWRQTRaJTO\nzk4KCgrmXPrc3NxsXfq8e/duAC5fvkxjYyPnzp3DMAwOHDiA3+8Hpi99fumll3A4HNd96fPb1dMB\nWPj4QaYq77/+yqwQ+kXKUy3yVIs81SKvrKzMtmNfM2yWK4XNNP0i5akWeapFnmqRZ2fY6BsERETE\ndgobERGxncJGRERsp7ARERHbKWxERMR2ChsREbGdwkZERGynsBEREdspbERExHYKGxERsZ3CRkRE\nbKewERER2ylsRETEdgobERGxncJGRERsp7ARERHbfaDbQr+XaDTKK6+8gtPpZM2aNTzxxBNcvHiR\nY8eOceHCBfx+P+FwGLfbbb2+q6sLl8s15y6eLS0t1l08a2trAchkMjQ1NTE4OIhhGITDYUpKShY2\nYhERueHmPbO5cOECL7/8Mt/4xjc4evQoU1NT/OhHP+LEiRPcfffdRCIRgsEg0WgUgKGhIXp6eqiv\nr6euro62tjZmbhLa1tbGvn37iEQinD9/nv7+fgA6OzvxeDw0NDRQXV1Ne3v7IgxZRERutHmHzYc/\n/GEKCgq4ePEiU1NTXLp0CdM06evrY9u2bQBs376d3t5eAPr6+ti6dSsulwu/309paSnxeJyxsTHS\n6TSBQACAqqoqa5/e3l7rWFu2bGFgYGBBgxURkaUx72U0j8fDww8/zBNPPMGHPvQhKioqqKioYHx8\nHK/XC4DX62V8fByAZDLJhg0brP1N0ySZTOJyufD5fFa7z+cjmUxa+8w853Q6KSoqIpVK4fF45ttt\nERFZAvOe2fzyl7/kX/7lX2hpaeHv//7v+e1vf8srr7wy53UOh2NBHbzSzLKbiIgsL/Oe2fziF7/g\n93//961Zxh/+4R/yn//5n3i9XsbGxqyfxcXFwPRMZnR01No/kUhgmiamaZJIJOa0z+wzs53NZkmn\n01ed1cRiMWKxmLVdU1NjPXY6nbgNY77DXPYKCwsxbuHxX0m1yFMt8lSL2To6OqzHwWCQYDC4KMed\nd9iUlZVx/PhxLl26xG233cbAwADr1q3j9ttvp7u7m127dtHd3U0oFAIgFArR0NDAww8/TDKZZHh4\nmEAggMPhwO12E4/HWbduHadOneLBBx+09jl58iTr16+np6eH8vLyq/bl/QqSzWaZmJiY7zCXPcMw\nbunxX0m1yFMt8lSLPMMwZn1YX0zzDpu1a9eybds2vvjFL+J0Olm7di2f+MQnuHjxIvX19XR1dbFq\n1SrC4TAAq1ev5r777iMcDlNQUMDevXutJbY9e/bQ3NxsXfq8efNmAHbu3EljYyP79+/HMAwOHDiw\nCEMWEZEbzZFboSdC3q6enlEVPn6Qqcr7l7g3S0ef2vJUizzVIk+1yCsrK7Pt2PoGARERsZ3CRkRE\nbKewERER2ylsRETEdgobERGxncJGRERsp7ARERHbKWxERMR2ChsREbGdwkZERGynsBEREdspbERE\nxHYKGxERsZ3CRkREbKewERER2ylsRETEdvO+UyfA5OQkzz33HG+//TYOh4PPfvazlJaWcuzYMS5c\nuIDf7yccDuN2uwGIRqN0dXXhcrmora1l06ZNAAwODtLS0mLdqbO2thaATCZDU1MTg4ODGIZBOBym\npKRkYSMWEZEbbkEzm29961vcc8891NfX8/TTT3PXXXdx4sQJ7r77biKRCMFgkGg0CsDQ0BA9PT3U\n19dTV1dHW1sbMzcJbWtrY9++fUQiEc6fP09/fz8AnZ2deDweGhoaqK6upr29fYHDFRGRpTDvsJmc\nnOSNN95gx44dALhcLtxuN319fWzbtg2A7du309vbC0BfXx9bt27F5XLh9/spLS0lHo8zNjZGOp0m\nEAgAUFVVZe3T29trHWvLli0MDAzMf6QiIrJk5r2MNjIygmEYtLS08NZbb/Gxj32M2tpaxsfH8Xq9\nAHi9XsbHxwFIJpNs2LDB2t80TZLJJC6XC5/PZ7X7fD6SyaS1z8xzTqeToqIiUqkUHo9nvt0WEZEl\nMO+wyWazvPnmm+zZs4d169bx7W9/mxMnTsx5ncPhWFAHrzSz7PZusViMWCxmbdfU1FiPnU4nbsNY\ntD4sN4WFhRi38PivpFrkqRZ5qsVsHR0d1uNgMEgwGFyU4847bEzTxOfzsW7dOmB6mevEiRN4vV7G\nxsasn8XFxdbrR0dHrf0TiQSmaWKaJolEYk77zD4z29lslnQ6fdVZzfsVJJvNMjExMd9hLnuGYdzS\n47+SapGnWuSpFnmGYcz6sL6Y5n3Oxuv14vP5eOeddwAYGBhg9erV3HvvvXR3dwPQ3d1NKBQCIBQK\ncfr0aTKZDCMjIwwPDxMIBPB6vbjdbuLxOLlcjlOnTlFZWWntc/LkSQB6enooLy9fyFhFRGSJLOjS\n5927d9PY2Egmk+EjH/kITzzxBNlslvr6erq6uli1ahXhcBiA1atXc9999xEOhykoKGDv3r3WEtue\nPXtobm62Ln3evHkzADt37qSxsZH9+/djGAYHDhxY4HBFRGQpOHLvdSJkmXu7enpGVfj4QaYq71/i\n3iwdLRHkqRZ5qkWeapFXVlZm27H1DQIiImI7hY2IiNhOYSMiIrZT2IiIiO0UNiIiYjuFjYiI2E5h\nIyIitlPYiIiI7RQ2IiJiO4WNiIjYTmEjIiK2U9iIiIjtFDYiImI7hY2IiNhOYSMiIrZT2IiIiO0W\ndKdOgGw2S11dHaZp8uSTT5JKpTh27BgXLlzA7/cTDodxu90ARKNRurq6cLlc1NbWsmnTJgAGBwdp\naWmx7tRZW1sLQCaToampicHBQQzDIBwOU1JSstAui4jIDbbgmc2//uu/ctddd1nbJ06c4O677yYS\niRAMBolGowAMDQ3R09NDfX09dXV1tLW1MXOT0La2Nvbt20ckEuH8+fP09/cD0NnZicfjoaGhgerq\natrb2xfaXRERWQILCptEIsGrr77KAw88YLX19fWxbds2ALZv305vb6/VvnXrVlwuF36/n9LSUuLx\nOGNjY6TTaQKBAABVVVXWPr29vdaxtmzZwsDAwEK6KyIiS2RBYfOd73yHxx57DIfDYbWNj4/j9XoB\n8Hq9jI+PA5BMJmctgZmmSTKZJJlM4vP5rHafz0cymbT2mXnO6XRSVFREKpVaSJdFRGQJzPuczZkz\nZyguLmbt2rXEYrH3fN2VQbRQM8tu7xaLxWb1oaamxnrsdDpxG8ai9WG5KSwsxLiFx38l1SJPtchT\nLWbr6OiwHgeDQYLB4KIcd95h88Ybb9DX18err77KpUuXSKfTNDY24vV6GRsbs34WFxcD0zOZ0dFR\na/9EIoFpmpimSSKRmNM+s8/MdjabJZ1O4/F45vTl/QqSzWaZmJiY7zCXPcMwbunxX0m1yFMt8lSL\nPMMwZn1YX0zzXkb79Kc/TWtrK01NTfz1X/815eXl/NVf/RX33nsv3d3dAHR3dxMKhQAIhUKcPn2a\nTCbDyMgIw8PDBAIBvF4vbrebeDxOLpfj1KlTVFZWWvucPHkSgJ6eHsrLyxc4XBERWQoLvvT53Xbt\n2kV9fT1dXV2sWrWKcDgMwOrVq7nvvvsIh8MUFBSwd+9ea4ltz549NDc3W5c+b968GYCdO3fS2NjI\n/v37MQyDAwcOLHZ3RUTkBnDk3utEyDL3dvX0jKrw8YNMVd6/xL1ZOloiyFMt8lSLPNUir6yszLZj\n6xsERETEdgobERGxncJGRERsp7ARERHbKWxERMR2ChsREbGdwkZERGynsBEREdspbERExHYKGxER\nsZ3CRkREbKewERER2ylsRETEdgobERGxncJGRERsp7ARERHbzftOnYlEgqamJsbHx3E4HDzwwAM8\n9NBDpFIpjh07xoULF/D7/YTDYdxuNwDRaJSuri5cLhe1tbVs2rQJgMHBQVpaWqw7ddbW1gKQyWRo\nampicHAQwzAIh8OUlJQsfNQiInJDzXtm43K5+MxnPsOzzz7L17/+dX7wgx/w3//935w4cYK7776b\nSCRCMBgkGo0CMDQ0RE9PD/X19dTV1dHW1sbMTULb2trYt28fkUiE8+fP09/fD0BnZycej4eGhgaq\nq6tpb29fhCGLiMiNNu+w8Xq9rF27FoDbb7+du+66i0QiQV9fH9u2bQNg+/bt9Pb2AtDX18fWrVtx\nuVz4/X5KS0uJx+OMjY2RTqcJBAIAVFVVWfv09vZax9qyZQsDAwPzHqiIiCydeS+jXWlkZIS33nqL\nDRs2MD4+jtfrBaYDaXx8HIBkMsmGDRusfUzTJJlM4nK58Pl8VrvP5yOZTFr7zDzndDopKioilUrh\n8Xiuq3+uX41C8sL//MWrmLpTS3EiIjfSgsPm4sWLPPvss9TW1nL77bfPed7hcCz0r7DMLLu9WywW\nIxaLWds1NTXWY6fTiWP8V6T/35MAfPjQM7jXfHTR+nSzKywsxDCMpe7GTUG1yFMt8lSL2To6OqzH\nwWCQYDC4KMddUNhMTU3xzDPPUFVVRWVlJTA9mxkbG7N+FhcXA9MzmdHRUWvfRCKBaZqYpkkikZjT\nPrPPzHY2myWdTl91VvN+BclmszCVuaLPGSYmJhYy7GXFMIxbarzvR7XIUy3yVIs8wzBmfVhfTAu6\n9Lm1tZXVq1fz0EMPWW333nsv3d3dAHR3dxMKhQAIhUKcPn2aTCbDyMgIw8PDBAIBvF4vbrebeDxO\nLpfj1KlTVnCFQiFOnjwJQE9PD+Xl5QvproiILJF5z2zeeOMNXnnlFdasWcMXvvAFHA4HjzzyCLt2\n7aK+vp6uri5WrVpFOBwGYPXq1dx3332Ew2EKCgrYu3evtcS2Z88empubrUufN2/eDMDOnTtpbGxk\n//79GIbBgQMHFmHIIiJyozly73UiZJl7u3p6RlX4+EEwS7j0P+dsCr/4d0yt+19L2bUbSksEeapF\nnmqRp1o+KxzEAAAHL0lEQVTklZWV2XZsfYOAiIjYTmEjIiK2U9iIiIjtFDYiImI7hY2IiNhOYSMi\nIrZT2IiIiO0UNiIiYjuFjYiI2E5hIyIitlPYiIiI7Rbl5mnLiaOgANcvfja9oRupiYjcELdc2DDx\nay5F/i8w/aWcKGxERGynZTQREbGdwkZERGy3LJbR+vv7+fa3v00ul2PHjh3s2rVrUY476/wN6ByO\niIhNbvqwyWazPP/883zlK1/hzjvvpK6ujsrKSu66666FH/yK8zcAH/rSM7iSF6Y3FDwiIovmpl9G\ni8fjlJaWsmrVKgoKCvjjP/5jent77fnLJn7Npf/35PRdPWdCR0REFuymn9kkk0l8Pp+1bZom8Xjc\n9r/3yiU2R5FB7jcTcx5Pd0gzIBGRa7npw2a+bvv0/wbAuTZA9tdj13+AK5bYPnTgq1d9DLOX3t4r\nlN4dUB/kde+3jwJORJYbRy6Xyy11J97Pz3/+c773ve9x6NAhAE6cOAEw6yKBWCxGLBaztmtqam5s\nJ0VEVoiOjg7rcTAYJBgMLspxb/pzNoFAgOHhYS5cuEAmk+HHP/4xoVBo1muCwSA1NTXWnyuLdatT\nLfJUizzVIk+1yOvo6Jj1b+liBQ0sg2U0p9PJnj17eOqpp8jlcuzcuZPVq1cvdbdEROQ63PRhA7B5\n82YikchSd0NERObppl9Gm4/FnPotd6pFnmqRp1rkqRZ5dtbipr9AQERElr8VObMREZGbi8JGRERs\ntywuELgedn1p583kc5/7HG63G4fDgcvl4siRI6RSKY4dO8aFCxfw+/2Ew2HcbjcA0WiUrq4uXC4X\ntbW1bNq0CYDBwUFaWlq4fPky99xzD7W1tUs4qg+utbWVM2fOUFxczNGjRwEWdfyZTIampiYGBwcx\nDINwOExJyc35n2ivVovvfe97vPzyyxQXFwPwyCOPsHnzZmDl1iKRSNDU1MT4+DgOh4MHHniAhx56\n6JZ8X7y7Fp/4xCd48MEHl/59kVtBpqamcp///OdzIyMjucuXL+f+9m//Njc0NLTU3Vp0n/vc53IT\nExOz2v7hH/4hd+LEiVwul8tFo9Fce3t7LpfL5d5+++3cwYMHc5lMJvfLX/4y9/nPfz6XzWZzuVwu\nV1dXl/uv//qvXC6Xyx0+fDj36quv3sBRzN/Pfvaz3Jtvvpn7m7/5G6ttMcf/gx/8IPfNb34zl8vl\ncj/+8Y9z9fX1N2xs1+tqtejo6Mh9//vfn/PalVyLX/3qV7k333wzl8vlcul0Ord///7c0NDQLfm+\neK9aLPX7YkUto93QL+1cQrlcjty7ruvo6+tj27ZtAGzfvt0ad19fH1u3bsXlcuH3+yktLSUejzM2\nNkY6nSYQCABQVVW1bGq1ceNGioqKZrUt5vh7e3utY23ZsoWBgYEbNbTrdrVaAHPeH7Cya+H1elm7\ndi0At99+O3fddReJROKWfF9crRbJZBJY2vfFigqbq31p50yRVxKHw8FTTz1FXV0dL7/8MgDj4+N4\nvV5g+s02Pj4OTNfkyuntTE3eXSufz7esa7WY47/yOafTSVFREalU6kYNZVH8+7//OwcPHuS5555j\ncnISuHVqMTIywltvvcWGDRtu+ffFTC3Wr18PLO37YsWds7kVfO1rX+POO+/k17/+NU899RRlZWVz\nXuNwOJagZzePxRz/1T4N3sz+9E//lE996lM4HA5eeOEFvvvd77Jv375FOfbNXouLFy/y7LPPUltb\ny+233z7n+VvpffHuWiz1+2JFzWxM02R0dNTaTiaTmKa5hD2yx5133gnAHXfcQWVlJfF4HK/Xy9jY\n9Ldbj42NWScB312TRCKBaZqYpkkikZjTvlwt5vivfC6bzZJOp/F4PDdqKAt2xx13WP+oPvDAA9Yt\nOVZ6LaampnjmmWeoqqqisrISuHXfF1erxVK/L1ZU2HyQL+1c7n77299y8eJFYPqTy09/+lPWrFnD\nvffeS3d3NwDd3d3WuEOhEKdPnyaTyTAyMsLw8DCBQACv14vb7SYej5PL5Th16pT1plwO3n3eajHH\nHwqFOHnyJAA9PT2Ul5ff2MFdp3fXYuYfV4D/+I//4Hd/93eBlV+L1tZWVq9ezUMPPWS13arvi6vV\nYqnfFyvuGwT6+/v51re+ZX1p50q79HlkZISnn34ah8PB1NQU999/P7t27SKVSlFfX8/o6CirVq0i\nHA5bJ46j0SidnZ0UFBTMuayxubnZuqxx9+7dSzm0DywSiXD27FkmJiYoLi6mpqaGysrKRRv/5cuX\naWxs5Ny5cxiGwYEDB/D7/Us23vdztVrEYjHOnTuHw+Fg1apVPP7449Z5i5VaizfeeIOvfvWrrFmz\nBofDgcPh4JFHHiEQCNxy74v3qsWPfvSjJX1frLiwERGRm8+KWkYTEZGbk8JGRERsp7ARERHbKWxE\nRMR2ChsREbGdwkZERGynsBEREdspbERExHb/H5OU7LCfXT0lAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f91fa25dcc0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure_ = plt.hist(lengths, bins=100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter out some very small exons:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "exons = exons[lengths > 10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Specify the recombination rate within exons:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "exons.ix[exons.type == 'exon', 'recomb_rate'] = RECOMB_RATE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The modified dataframe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>recomb_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>134900</td>\n",
       "      <td>135802</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>137620</td>\n",
       "      <td>139379</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>367639</td>\n",
       "      <td>368634</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>621058</td>\n",
       "      <td>622053</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end  type   recomb_rate\n",
       "0      1   69090   70008  exon  1.000000e-08\n",
       "1      1  134900  135802  exon  1.000000e-08\n",
       "2      1  137620  139379  exon  1.000000e-08\n",
       "3      1  367639  368634  exon  1.000000e-08\n",
       "4      1  621058  622053  exon  1.000000e-08"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process positions from the archaic admixture array"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the coordinates of sites from the archaic admixture array:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "array_sites = pd.read_table('../raw_data/ice_age/filter.gz', names=['chrom', 'end', 'set_1', 'set_2'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Subset to sites in the so called \"Set 1\" (see `raw_data/ice_age/README.md` for more information):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "array_sites = array_sites[array_sites.set_1 == 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "array_sites['start'] = array_sites.end - 1\n",
    "array_sites = array_sites[['chrom', 'start', 'end']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "array_sites['type'] = 'snp'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "DataFrame after filtering:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>847982</td>\n",
       "      <td>847983</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>853088</td>\n",
       "      <td>853089</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>853595</td>\n",
       "      <td>853596</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>854792</td>\n",
       "      <td>854793</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>867150</td>\n",
       "      <td>867151</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end type\n",
       "0      1  847982  847983  snp\n",
       "1      1  853088  853089  snp\n",
       "2      1  853595  853596  snp\n",
       "3      1  854792  854793  snp\n",
       "5      1  867150  867151  snp"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "array_sites.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a union of exon regions with admixture array informative positions\n",
    "\n",
    "Inter-exon informative sites will be simulated individually, with an appropriate recombination rate set between them and adjacent exons or other informative sites. Intra-exon sites will be ignored when building the recombination map (they will be recombining as a part of the exons)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Take the subset of array sites that lie outside exons:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "exome = BedTool.from_dataframe(exons)\n",
    "sites_outside_exons = BedTool.from_dataframe(array_sites).intersect(exome, v=True).to_dataframe().rename(columns={'name': 'type'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "exons_and_sites = pd.concat([sites_outside_exons,\n",
    "                             exons]).sort_values(by=['chrom', 'start']).reset_index(drop=True)\n",
    "exons_and_sites = exons_and_sites[['chrom', 'start', 'end', 'type', 'recomb_rate']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>recomb_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>134900</td>\n",
       "      <td>135802</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>137620</td>\n",
       "      <td>139379</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>367639</td>\n",
       "      <td>368634</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>621058</td>\n",
       "      <td>622053</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end  type   recomb_rate\n",
       "0      1   69090   70008  exon  1.000000e-08\n",
       "1      1  134900  135802  exon  1.000000e-08\n",
       "2      1  137620  139379  exon  1.000000e-08\n",
       "3      1  367639  368634  exon  1.000000e-08\n",
       "4      1  621058  622053  exon  1.000000e-08"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons_and_sites.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "978504"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(exons_and_sites)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generating the recombination map of exons and archaic admixture array sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def add_recombination_gaps(regions):\n",
    "    \"\"\"Add 1 bp \"recombination gap\" between each record in\n",
    "    a given DataFrame of exon and snp coordinates.\n",
    "    \"\"\"\n",
    "    # create a new DataFrame with coordinates of 1 bp gaps\n",
    "    gaps = pd.DataFrame({'chrom'       : regions.chrom.values,\n",
    "                         'start'       : regions.end.values,\n",
    "                         'end'         : regions.end.values + 1,\n",
    "                         'recomb_rate' : list(RECOMB_RATE * (regions.start[1:].values -  # between exons/sites\n",
    "                                                             regions.end[:-1].values)) +\n",
    "                                         [0.5], # between chromosomes\n",
    "                         'type'        : 'gap'},\n",
    "                        columns=['chrom', 'start', 'end', 'recomb_rate', 'type'])\n",
    "\n",
    "    # merge the dataframes of regions and gap coordinates\n",
    "    regions_and_gaps = pd.concat([regions, gaps]).sort_values(by=['chrom', 'start']).reset_index(drop=True)\n",
    "\n",
    "    return regions_and_gaps[['chrom', 'start', 'end', 'type', 'recomb_rate']]\n",
    "\n",
    "\n",
    "def concatenate_regions(regions):\n",
    "    \"\"\"Concatenate regions/sites on all chromosome as if they were\n",
    "    directly adjacent on a single chromosome, changing the coordinates\n",
    "    accordingly.\n",
    "    \"\"\"\n",
    "    concat_regions = regions.copy()\n",
    "\n",
    "    concat_regions['width'] = regions.end - regions.start\n",
    "    concat_regions['slim_start'] = pd.Series([0] + list(concat_regions.width[:-1])).cumsum()\n",
    "    concat_regions['slim_end'] = concat_regions.width.cumsum() - 1\n",
    "    \n",
    "    return concat_regions\n",
    "\n",
    "\n",
    "def create_recomb_map(regions):\n",
    "    \"\"\"Create recombination map for all regions/sites on all chromosomes.\"\"\"\n",
    "    recomb_map = []\n",
    "\n",
    "    for chrom in AUTOSOMES:\n",
    "        # add recombination gaps between regions on this chromosome\n",
    "        recomb_map.append(add_recombination_gaps(regions.query('chrom == {}'.format(chrom))))\n",
    "        \n",
    "    # concatenate exome recombination maps of all chromosomes\n",
    "    recomb_map = pd.concat(recomb_map, ignore_index=True).sort_values(by=['chrom', 'start']).reset_index(drop=True)\n",
    "\n",
    "    # remove the very last base of the recombination map\n",
    "    # (it has a 0.5 recombination rate anyway and there's no other chromosome\n",
    "    # after it)\n",
    "    recomb_map = recomb_map[:-1]\n",
    "    \n",
    "    return recomb_map"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Recombination rate between exons and/or informative positions is implemented by inserting a 1 bp \"gap\" between each adjacent region/site and setting the recombination rate at these positions to $L \\times 1\\cdot10^{-8}$ crossovers per generation ($L$ is the distance between exons/sites).\n",
    "\n",
    "The recombination rate of the \"gap\" between the last feature on one chromosome and the first feature on another chromosome will be 0.5."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a recombination map of exons and informative sites. Inserts 1bp recombination gaps between each region and site and between different chromosomes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "recomb_map = create_recomb_map(exons_and_sites)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>recomb_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>70008</td>\n",
       "      <td>70009</td>\n",
       "      <td>gap</td>\n",
       "      <td>6.489200e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>134900</td>\n",
       "      <td>135802</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>135802</td>\n",
       "      <td>135803</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.818000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>137620</td>\n",
       "      <td>139379</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end  type   recomb_rate\n",
       "0      1   69090   70008  exon  1.000000e-08\n",
       "1      1   70008   70009   gap  6.489200e-04\n",
       "2      1  134900  135802  exon  1.000000e-08\n",
       "3      1  135802  135803   gap  1.818000e-05\n",
       "4      1  137620  139379  exon  1.000000e-08"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "recomb_map.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert the coordinates of all regions/sites/gaps into SLiM's 0-based single segment coordinate system:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "concat_map = concatenate_regions(recomb_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>recomb_rate</th>\n",
       "      <th>width</th>\n",
       "      <th>slim_start</th>\n",
       "      <th>slim_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>918</td>\n",
       "      <td>0</td>\n",
       "      <td>917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>70008</td>\n",
       "      <td>70009</td>\n",
       "      <td>gap</td>\n",
       "      <td>6.489200e-04</td>\n",
       "      <td>1</td>\n",
       "      <td>918</td>\n",
       "      <td>918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>134900</td>\n",
       "      <td>135802</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>902</td>\n",
       "      <td>919</td>\n",
       "      <td>1820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>135802</td>\n",
       "      <td>135803</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.818000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>1821</td>\n",
       "      <td>1821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>137620</td>\n",
       "      <td>139379</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>1759</td>\n",
       "      <td>1822</td>\n",
       "      <td>3580</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end  type   recomb_rate  width  slim_start  slim_end\n",
       "0      1   69090   70008  exon  1.000000e-08    918           0       917\n",
       "1      1   70008   70009   gap  6.489200e-04      1         918       918\n",
       "2      1  134900  135802  exon  1.000000e-08    902         919      1820\n",
       "3      1  135802  135803   gap  1.818000e-05      1        1821      1821\n",
       "4      1  137620  139379  exon  1.000000e-08   1759        1822      3580"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concat_map.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save the recombination map in a SLiM-friendly format (end-positions of exons and gaps, without the positions of SNPs since they don't have recombination rates themselves):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "concat_map.query('type != \"snp\"')[['slim_end', 'recomb_rate']].to_csv('../slim/recombination_map.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save SLiM coordinates of all sites from the archaic admixture array\n",
    "\n",
    "Recombination map includes only positions of sites that fall outside of exonic regions. However, in order to simulate the sites from the archaic admixture array, we have to know the positions of sites _within_ exons too.\n",
    "\n",
    "SLiM simulates all exons and individual sites as one continuous segment of concatenated regions, with coordinates of exons and sites shifted appropriately. To obtain the coordinates of array sites within exons, we need to find out, for each site, which exon does it fall in and calculate its position relative to the start of that exon."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Take the subset of array sites that lie inside exons:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sites_within_exons = BedTool.from_dataframe(array_sites).intersect(exome).to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sites_within_exons = sites_within_exons.rename(columns={'name': 'type'})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get a DataFrame of the coordinates of exons that contain a site from the admixture array (will contain multiple copies of one exon if more than one site falls within that exon):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "exons_with_sites = BedTool.from_dataframe(concat_map.query('type == \"exon\"')).intersect(BedTool.from_dataframe(sites_within_exons), wa=True).to_dataframe()\n",
    "\n",
    "\n",
    "# rename and subset columns\n",
    "exons_with_sites = exons_with_sites.rename(columns={'name': 'type',\n",
    "                                                    'score': 'recomb_rate',\n",
    "                                                    'strand': 'width',\n",
    "                                                    'thickStart': 'slim_start',\n",
    "                                                    'thickEnd': 'slim_end'})[['chrom', 'start', 'end', 'type', 'recomb_rate', 'width', 'slim_start', 'slim_end']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate the position of each site relative to the start of \"its\" exon and convert this position into a SLiM single-segment coordinate (i.e. relative the the position 0 of the simulated segment):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sites_within_exons['width'] = 1\n",
    "sites_within_exons['slim_start'] = sites_within_exons.start - exons_with_sites.start + exons_with_sites.slim_start\n",
    "sites_within_exons['slim_end'] = sites_within_exons.slim_start"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge the coordinates of SNPs within and outside exons:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "all_sites = pd.concat([sites_within_exons, concat_map.query('type == \"snp\"')]).sort_values(by=['slim_start']).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Verify that we didn't loose any sites during all this coordinate acrobatics:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(array_sites) == len(all_sites)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Output the SLiM-based positions of informative sites:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "all_sites['slim_start'].to_csv('../slim/admixture_array_coordinates.txt', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save SLiM coordinates of exonic regions only\n",
    "\n",
    "This is required for specification for the `initializeGenomicElement` function in SLiM."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "concat_map.query('type == \"exon\"')[['slim_start', 'slim_end']].to_csv('../slim/exon_coordinates.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br><br><br><br><br>\n",
    "# Testing of the recombination map building functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sample 5 exons/sites for each chromosome to make a testing data set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "test_data = exons_and_sites.groupby('chrom') \\\n",
    "                           .apply(pd.DataFrame.head) \\\n",
    "                           .sort_values(by=['chrom', 'start']) \\\n",
    "                           .reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Test `add_recombination_gaps`\n",
    "\n",
    "The function should add a 1bp gap between each exon or site (with a recombination rate at this gap proportional to the distance between the adjacent features) and it should also insert a 0.5 recombination rate between chromosomes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>recomb_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>70008</td>\n",
       "      <td>70009</td>\n",
       "      <td>gap</td>\n",
       "      <td>6.489200e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>134900</td>\n",
       "      <td>135802</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>135802</td>\n",
       "      <td>135803</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.818000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>137620</td>\n",
       "      <td>139379</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>139379</td>\n",
       "      <td>139380</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.282600e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>367639</td>\n",
       "      <td>368634</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>368634</td>\n",
       "      <td>368635</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.524240e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>621058</td>\n",
       "      <td>622053</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>622053</td>\n",
       "      <td>622054</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end  type   recomb_rate\n",
       "0      1   69090   70008  exon  1.000000e-08\n",
       "1      1   70008   70009   gap  6.489200e-04\n",
       "2      1  134900  135802  exon  1.000000e-08\n",
       "3      1  135802  135803   gap  1.818000e-05\n",
       "4      1  137620  139379  exon  1.000000e-08\n",
       "5      1  139379  139380   gap  2.282600e-03\n",
       "6      1  367639  368634  exon  1.000000e-08\n",
       "7      1  368634  368635   gap  2.524240e-03\n",
       "8      1  621058  622053  exon  1.000000e-08\n",
       "9      1  622053  622054   gap  5.000000e-01"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "add_recombination_gaps(test_data.query('chrom == 1'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Test `concatenate_regions`\n",
    "\n",
    "The function should concatenate all regions into a single continuous segment. This segment should start at 0 and its last position should be equal to the total length of all regions - 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>recomb_rate</th>\n",
       "      <th>width</th>\n",
       "      <th>slim_start</th>\n",
       "      <th>slim_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>918</td>\n",
       "      <td>0</td>\n",
       "      <td>917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>134900</td>\n",
       "      <td>135802</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>902</td>\n",
       "      <td>918</td>\n",
       "      <td>1819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>137620</td>\n",
       "      <td>139379</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>1759</td>\n",
       "      <td>1820</td>\n",
       "      <td>3578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>367639</td>\n",
       "      <td>368634</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>995</td>\n",
       "      <td>3579</td>\n",
       "      <td>4573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>621058</td>\n",
       "      <td>622053</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>995</td>\n",
       "      <td>4574</td>\n",
       "      <td>5568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>32403</td>\n",
       "      <td>32404</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>5569</td>\n",
       "      <td>5569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>34135</td>\n",
       "      <td>34136</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>5570</td>\n",
       "      <td>5570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>35451</td>\n",
       "      <td>35452</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>5571</td>\n",
       "      <td>5571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2</td>\n",
       "      <td>38813</td>\n",
       "      <td>41627</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>2814</td>\n",
       "      <td>5572</td>\n",
       "      <td>8385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2</td>\n",
       "      <td>45439</td>\n",
       "      <td>46385</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>946</td>\n",
       "      <td>8386</td>\n",
       "      <td>9331</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end  type   recomb_rate  width  slim_start  slim_end\n",
       "0      1   69090   70008  exon  1.000000e-08    918           0       917\n",
       "1      1  134900  135802  exon  1.000000e-08    902         918      1819\n",
       "2      1  137620  139379  exon  1.000000e-08   1759        1820      3578\n",
       "3      1  367639  368634  exon  1.000000e-08    995        3579      4573\n",
       "4      1  621058  622053  exon  1.000000e-08    995        4574      5568\n",
       "5      2   32403   32404   snp           NaN      1        5569      5569\n",
       "6      2   34135   34136   snp           NaN      1        5570      5570\n",
       "7      2   35451   35452   snp           NaN      1        5571      5571\n",
       "8      2   38813   41627  exon  1.000000e-08   2814        5572      8385\n",
       "9      2   45439   46385  exon  1.000000e-08    946        8386      9331"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concatenate_regions(test_data.query('chrom == 1 | chrom == 2'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What is the correct end of the single continuous segment?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9331"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(concatenate_regions(test_data.query('chrom == 1 | chrom == 2')).width) - 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Test `create_recomb_map`\n",
    "\n",
    "The function should create a recombination map of all exons and sites in the given DataFrame. A 1 bp gap should be inserted between all regions, and a similar gap with a recombination rate of 0.5 should be inserted between each \"adjacent\" pair of chromosomes.\n",
    "\n",
    "SNPs have no recombination rate and they are not going to be a part of the recombination rate output file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>recomb_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>70008</td>\n",
       "      <td>70009</td>\n",
       "      <td>gap</td>\n",
       "      <td>6.489200e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>134900</td>\n",
       "      <td>135802</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>135802</td>\n",
       "      <td>135803</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.818000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>137620</td>\n",
       "      <td>139379</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>139379</td>\n",
       "      <td>139380</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.282600e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>367639</td>\n",
       "      <td>368634</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>368634</td>\n",
       "      <td>368635</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.524240e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>621058</td>\n",
       "      <td>622053</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>622053</td>\n",
       "      <td>622054</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2</td>\n",
       "      <td>32403</td>\n",
       "      <td>32404</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2</td>\n",
       "      <td>32404</td>\n",
       "      <td>32405</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.731000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2</td>\n",
       "      <td>34135</td>\n",
       "      <td>34136</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2</td>\n",
       "      <td>34136</td>\n",
       "      <td>34137</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.315000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2</td>\n",
       "      <td>35451</td>\n",
       "      <td>35452</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2</td>\n",
       "      <td>35452</td>\n",
       "      <td>35453</td>\n",
       "      <td>gap</td>\n",
       "      <td>3.361000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>2</td>\n",
       "      <td>38813</td>\n",
       "      <td>41627</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>2</td>\n",
       "      <td>41627</td>\n",
       "      <td>41628</td>\n",
       "      <td>gap</td>\n",
       "      <td>3.812000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>2</td>\n",
       "      <td>45439</td>\n",
       "      <td>46385</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>2</td>\n",
       "      <td>46385</td>\n",
       "      <td>46386</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>3</td>\n",
       "      <td>100338</td>\n",
       "      <td>100339</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>3</td>\n",
       "      <td>100339</td>\n",
       "      <td>100340</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.842000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>3</td>\n",
       "      <td>102181</td>\n",
       "      <td>102182</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>3</td>\n",
       "      <td>102182</td>\n",
       "      <td>102183</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.200000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>3</td>\n",
       "      <td>102204</td>\n",
       "      <td>102205</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>3</td>\n",
       "      <td>102205</td>\n",
       "      <td>102206</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.937000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>3</td>\n",
       "      <td>108142</td>\n",
       "      <td>108143</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>3</td>\n",
       "      <td>108143</td>\n",
       "      <td>108144</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.472000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>3</td>\n",
       "      <td>113615</td>\n",
       "      <td>113616</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>3</td>\n",
       "      <td>113616</td>\n",
       "      <td>113617</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>19</td>\n",
       "      <td>282809</td>\n",
       "      <td>282810</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>20</td>\n",
       "      <td>68350</td>\n",
       "      <td>68408</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>20</td>\n",
       "      <td>68408</td>\n",
       "      <td>68409</td>\n",
       "      <td>gap</td>\n",
       "      <td>8.237000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>20</td>\n",
       "      <td>76645</td>\n",
       "      <td>77174</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>20</td>\n",
       "      <td>77174</td>\n",
       "      <td>77175</td>\n",
       "      <td>gap</td>\n",
       "      <td>3.549000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>20</td>\n",
       "      <td>80723</td>\n",
       "      <td>80724</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>20</td>\n",
       "      <td>80724</td>\n",
       "      <td>80725</td>\n",
       "      <td>gap</td>\n",
       "      <td>3.793000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>20</td>\n",
       "      <td>84517</td>\n",
       "      <td>84518</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>20</td>\n",
       "      <td>84518</td>\n",
       "      <td>84519</td>\n",
       "      <td>gap</td>\n",
       "      <td>9.830000e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>20</td>\n",
       "      <td>85501</td>\n",
       "      <td>85502</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>20</td>\n",
       "      <td>85502</td>\n",
       "      <td>85503</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>21</td>\n",
       "      <td>10906200</td>\n",
       "      <td>10907040</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>21</td>\n",
       "      <td>10907040</td>\n",
       "      <td>10907041</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.784000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>21</td>\n",
       "      <td>10908824</td>\n",
       "      <td>10908895</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>21</td>\n",
       "      <td>10908895</td>\n",
       "      <td>10908896</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.411000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>21</td>\n",
       "      <td>10910306</td>\n",
       "      <td>10910399</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>21</td>\n",
       "      <td>10910399</td>\n",
       "      <td>10910400</td>\n",
       "      <td>gap</td>\n",
       "      <td>3.963000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>21</td>\n",
       "      <td>10914362</td>\n",
       "      <td>10914442</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>21</td>\n",
       "      <td>10914442</td>\n",
       "      <td>10914443</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.927000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>21</td>\n",
       "      <td>10916369</td>\n",
       "      <td>10916475</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>21</td>\n",
       "      <td>10916475</td>\n",
       "      <td>10916476</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>22</td>\n",
       "      <td>16256440</td>\n",
       "      <td>16256677</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>22</td>\n",
       "      <td>16256677</td>\n",
       "      <td>16256678</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.507000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>22</td>\n",
       "      <td>16258184</td>\n",
       "      <td>16258303</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>22</td>\n",
       "      <td>16258303</td>\n",
       "      <td>16258304</td>\n",
       "      <td>gap</td>\n",
       "      <td>8.625000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>22</td>\n",
       "      <td>16266928</td>\n",
       "      <td>16267095</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>22</td>\n",
       "      <td>16267095</td>\n",
       "      <td>16267096</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.041000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>22</td>\n",
       "      <td>16268136</td>\n",
       "      <td>16268181</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>22</td>\n",
       "      <td>16268181</td>\n",
       "      <td>16268182</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.691000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>22</td>\n",
       "      <td>16269872</td>\n",
       "      <td>16269943</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>219 rows  5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom     start       end  type   recomb_rate\n",
       "0        1     69090     70008  exon  1.000000e-08\n",
       "1        1     70008     70009   gap  6.489200e-04\n",
       "2        1    134900    135802  exon  1.000000e-08\n",
       "3        1    135802    135803   gap  1.818000e-05\n",
       "4        1    137620    139379  exon  1.000000e-08\n",
       "5        1    139379    139380   gap  2.282600e-03\n",
       "6        1    367639    368634  exon  1.000000e-08\n",
       "7        1    368634    368635   gap  2.524240e-03\n",
       "8        1    621058    622053  exon  1.000000e-08\n",
       "9        1    622053    622054   gap  5.000000e-01\n",
       "10       2     32403     32404   snp           NaN\n",
       "11       2     32404     32405   gap  1.731000e-05\n",
       "12       2     34135     34136   snp           NaN\n",
       "13       2     34136     34137   gap  1.315000e-05\n",
       "14       2     35451     35452   snp           NaN\n",
       "15       2     35452     35453   gap  3.361000e-05\n",
       "16       2     38813     41627  exon  1.000000e-08\n",
       "17       2     41627     41628   gap  3.812000e-05\n",
       "18       2     45439     46385  exon  1.000000e-08\n",
       "19       2     46385     46386   gap  5.000000e-01\n",
       "20       3    100338    100339   snp           NaN\n",
       "21       3    100339    100340   gap  1.842000e-05\n",
       "22       3    102181    102182   snp           NaN\n",
       "23       3    102182    102183   gap  2.200000e-07\n",
       "24       3    102204    102205   snp           NaN\n",
       "25       3    102205    102206   gap  5.937000e-05\n",
       "26       3    108142    108143   snp           NaN\n",
       "27       3    108143    108144   gap  5.472000e-05\n",
       "28       3    113615    113616   snp           NaN\n",
       "29       3    113616    113617   gap  5.000000e-01\n",
       "..     ...       ...       ...   ...           ...\n",
       "189     19    282809    282810   gap  5.000000e-01\n",
       "190     20     68350     68408  exon  1.000000e-08\n",
       "191     20     68408     68409   gap  8.237000e-05\n",
       "192     20     76645     77174  exon  1.000000e-08\n",
       "193     20     77174     77175   gap  3.549000e-05\n",
       "194     20     80723     80724   snp           NaN\n",
       "195     20     80724     80725   gap  3.793000e-05\n",
       "196     20     84517     84518   snp           NaN\n",
       "197     20     84518     84519   gap  9.830000e-06\n",
       "198     20     85501     85502   snp           NaN\n",
       "199     20     85502     85503   gap  5.000000e-01\n",
       "200     21  10906200  10907040  exon  1.000000e-08\n",
       "201     21  10907040  10907041   gap  1.784000e-05\n",
       "202     21  10908824  10908895  exon  1.000000e-08\n",
       "203     21  10908895  10908896   gap  1.411000e-05\n",
       "204     21  10910306  10910399  exon  1.000000e-08\n",
       "205     21  10910399  10910400   gap  3.963000e-05\n",
       "206     21  10914362  10914442  exon  1.000000e-08\n",
       "207     21  10914442  10914443   gap  1.927000e-05\n",
       "208     21  10916369  10916475  exon  1.000000e-08\n",
       "209     21  10916475  10916476   gap  5.000000e-01\n",
       "210     22  16256440  16256677  exon  1.000000e-08\n",
       "211     22  16256677  16256678   gap  1.507000e-05\n",
       "212     22  16258184  16258303  exon  1.000000e-08\n",
       "213     22  16258303  16258304   gap  8.625000e-05\n",
       "214     22  16266928  16267095  exon  1.000000e-08\n",
       "215     22  16267095  16267096   gap  1.041000e-05\n",
       "216     22  16268136  16268181  exon  1.000000e-08\n",
       "217     22  16268181  16268182   gap  1.691000e-05\n",
       "218     22  16269872  16269943  exon  1.000000e-08\n",
       "\n",
       "[219 rows x 5 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "create_recomb_map(test_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Test `create_recomb_map` and `concatenate_regoins` together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>recomb_rate</th>\n",
       "      <th>width</th>\n",
       "      <th>slim_start</th>\n",
       "      <th>slim_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>918</td>\n",
       "      <td>0</td>\n",
       "      <td>917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>70008</td>\n",
       "      <td>70009</td>\n",
       "      <td>gap</td>\n",
       "      <td>6.489200e-04</td>\n",
       "      <td>1</td>\n",
       "      <td>918</td>\n",
       "      <td>918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>134900</td>\n",
       "      <td>135802</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>902</td>\n",
       "      <td>919</td>\n",
       "      <td>1820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>135802</td>\n",
       "      <td>135803</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.818000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>1821</td>\n",
       "      <td>1821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>137620</td>\n",
       "      <td>139379</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>1759</td>\n",
       "      <td>1822</td>\n",
       "      <td>3580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>139379</td>\n",
       "      <td>139380</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.282600e-03</td>\n",
       "      <td>1</td>\n",
       "      <td>3581</td>\n",
       "      <td>3581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>367639</td>\n",
       "      <td>368634</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>995</td>\n",
       "      <td>3582</td>\n",
       "      <td>4576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>368634</td>\n",
       "      <td>368635</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.524240e-03</td>\n",
       "      <td>1</td>\n",
       "      <td>4577</td>\n",
       "      <td>4577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>621058</td>\n",
       "      <td>622053</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>995</td>\n",
       "      <td>4578</td>\n",
       "      <td>5572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>622053</td>\n",
       "      <td>622054</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>5573</td>\n",
       "      <td>5573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2</td>\n",
       "      <td>32403</td>\n",
       "      <td>32404</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>5574</td>\n",
       "      <td>5574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2</td>\n",
       "      <td>32404</td>\n",
       "      <td>32405</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.731000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>5575</td>\n",
       "      <td>5575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2</td>\n",
       "      <td>34135</td>\n",
       "      <td>34136</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>5576</td>\n",
       "      <td>5576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2</td>\n",
       "      <td>34136</td>\n",
       "      <td>34137</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.315000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>5577</td>\n",
       "      <td>5577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2</td>\n",
       "      <td>35451</td>\n",
       "      <td>35452</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>5578</td>\n",
       "      <td>5578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2</td>\n",
       "      <td>35452</td>\n",
       "      <td>35453</td>\n",
       "      <td>gap</td>\n",
       "      <td>3.361000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>5579</td>\n",
       "      <td>5579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>2</td>\n",
       "      <td>38813</td>\n",
       "      <td>41627</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>2814</td>\n",
       "      <td>5580</td>\n",
       "      <td>8393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>2</td>\n",
       "      <td>41627</td>\n",
       "      <td>41628</td>\n",
       "      <td>gap</td>\n",
       "      <td>3.812000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>8394</td>\n",
       "      <td>8394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>2</td>\n",
       "      <td>45439</td>\n",
       "      <td>46385</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>946</td>\n",
       "      <td>8395</td>\n",
       "      <td>9340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>2</td>\n",
       "      <td>46385</td>\n",
       "      <td>46386</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>9341</td>\n",
       "      <td>9341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>3</td>\n",
       "      <td>100338</td>\n",
       "      <td>100339</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>9342</td>\n",
       "      <td>9342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>3</td>\n",
       "      <td>100339</td>\n",
       "      <td>100340</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.842000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>9343</td>\n",
       "      <td>9343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>3</td>\n",
       "      <td>102181</td>\n",
       "      <td>102182</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>9344</td>\n",
       "      <td>9344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>3</td>\n",
       "      <td>102182</td>\n",
       "      <td>102183</td>\n",
       "      <td>gap</td>\n",
       "      <td>2.200000e-07</td>\n",
       "      <td>1</td>\n",
       "      <td>9345</td>\n",
       "      <td>9345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>3</td>\n",
       "      <td>102204</td>\n",
       "      <td>102205</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>9346</td>\n",
       "      <td>9346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>3</td>\n",
       "      <td>102205</td>\n",
       "      <td>102206</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.937000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>9347</td>\n",
       "      <td>9347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>3</td>\n",
       "      <td>108142</td>\n",
       "      <td>108143</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>9348</td>\n",
       "      <td>9348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>3</td>\n",
       "      <td>108143</td>\n",
       "      <td>108144</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.472000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>9349</td>\n",
       "      <td>9349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>3</td>\n",
       "      <td>113615</td>\n",
       "      <td>113616</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>9350</td>\n",
       "      <td>9350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>3</td>\n",
       "      <td>113616</td>\n",
       "      <td>113617</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>9351</td>\n",
       "      <td>9351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>19</td>\n",
       "      <td>282809</td>\n",
       "      <td>282810</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>31909</td>\n",
       "      <td>31909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>20</td>\n",
       "      <td>68350</td>\n",
       "      <td>68408</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>58</td>\n",
       "      <td>31910</td>\n",
       "      <td>31967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>20</td>\n",
       "      <td>68408</td>\n",
       "      <td>68409</td>\n",
       "      <td>gap</td>\n",
       "      <td>8.237000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>31968</td>\n",
       "      <td>31968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>20</td>\n",
       "      <td>76645</td>\n",
       "      <td>77174</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>529</td>\n",
       "      <td>31969</td>\n",
       "      <td>32497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>20</td>\n",
       "      <td>77174</td>\n",
       "      <td>77175</td>\n",
       "      <td>gap</td>\n",
       "      <td>3.549000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>32498</td>\n",
       "      <td>32498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>20</td>\n",
       "      <td>80723</td>\n",
       "      <td>80724</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>32499</td>\n",
       "      <td>32499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>20</td>\n",
       "      <td>80724</td>\n",
       "      <td>80725</td>\n",
       "      <td>gap</td>\n",
       "      <td>3.793000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>32500</td>\n",
       "      <td>32500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>20</td>\n",
       "      <td>84517</td>\n",
       "      <td>84518</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>32501</td>\n",
       "      <td>32501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>20</td>\n",
       "      <td>84518</td>\n",
       "      <td>84519</td>\n",
       "      <td>gap</td>\n",
       "      <td>9.830000e-06</td>\n",
       "      <td>1</td>\n",
       "      <td>32502</td>\n",
       "      <td>32502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>20</td>\n",
       "      <td>85501</td>\n",
       "      <td>85502</td>\n",
       "      <td>snp</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>32503</td>\n",
       "      <td>32503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>20</td>\n",
       "      <td>85502</td>\n",
       "      <td>85503</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>32504</td>\n",
       "      <td>32504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>21</td>\n",
       "      <td>10906200</td>\n",
       "      <td>10907040</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>840</td>\n",
       "      <td>32505</td>\n",
       "      <td>33344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>21</td>\n",
       "      <td>10907040</td>\n",
       "      <td>10907041</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.784000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>33345</td>\n",
       "      <td>33345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>21</td>\n",
       "      <td>10908824</td>\n",
       "      <td>10908895</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>71</td>\n",
       "      <td>33346</td>\n",
       "      <td>33416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>21</td>\n",
       "      <td>10908895</td>\n",
       "      <td>10908896</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.411000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>33417</td>\n",
       "      <td>33417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>21</td>\n",
       "      <td>10910306</td>\n",
       "      <td>10910399</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>93</td>\n",
       "      <td>33418</td>\n",
       "      <td>33510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>21</td>\n",
       "      <td>10910399</td>\n",
       "      <td>10910400</td>\n",
       "      <td>gap</td>\n",
       "      <td>3.963000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>33511</td>\n",
       "      <td>33511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>21</td>\n",
       "      <td>10914362</td>\n",
       "      <td>10914442</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>80</td>\n",
       "      <td>33512</td>\n",
       "      <td>33591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>21</td>\n",
       "      <td>10914442</td>\n",
       "      <td>10914443</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.927000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>33592</td>\n",
       "      <td>33592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>21</td>\n",
       "      <td>10916369</td>\n",
       "      <td>10916475</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>106</td>\n",
       "      <td>33593</td>\n",
       "      <td>33698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>21</td>\n",
       "      <td>10916475</td>\n",
       "      <td>10916476</td>\n",
       "      <td>gap</td>\n",
       "      <td>5.000000e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>33699</td>\n",
       "      <td>33699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>22</td>\n",
       "      <td>16256440</td>\n",
       "      <td>16256677</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>237</td>\n",
       "      <td>33700</td>\n",
       "      <td>33936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>22</td>\n",
       "      <td>16256677</td>\n",
       "      <td>16256678</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.507000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>33937</td>\n",
       "      <td>33937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>22</td>\n",
       "      <td>16258184</td>\n",
       "      <td>16258303</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>119</td>\n",
       "      <td>33938</td>\n",
       "      <td>34056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>22</td>\n",
       "      <td>16258303</td>\n",
       "      <td>16258304</td>\n",
       "      <td>gap</td>\n",
       "      <td>8.625000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>34057</td>\n",
       "      <td>34057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>22</td>\n",
       "      <td>16266928</td>\n",
       "      <td>16267095</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>167</td>\n",
       "      <td>34058</td>\n",
       "      <td>34224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>22</td>\n",
       "      <td>16267095</td>\n",
       "      <td>16267096</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.041000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>34225</td>\n",
       "      <td>34225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>22</td>\n",
       "      <td>16268136</td>\n",
       "      <td>16268181</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>45</td>\n",
       "      <td>34226</td>\n",
       "      <td>34270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>22</td>\n",
       "      <td>16268181</td>\n",
       "      <td>16268182</td>\n",
       "      <td>gap</td>\n",
       "      <td>1.691000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>34271</td>\n",
       "      <td>34271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>22</td>\n",
       "      <td>16269872</td>\n",
       "      <td>16269943</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>71</td>\n",
       "      <td>34272</td>\n",
       "      <td>34342</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>219 rows  8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom     start       end  type   recomb_rate  width  slim_start  \\\n",
       "0        1     69090     70008  exon  1.000000e-08    918           0   \n",
       "1        1     70008     70009   gap  6.489200e-04      1         918   \n",
       "2        1    134900    135802  exon  1.000000e-08    902         919   \n",
       "3        1    135802    135803   gap  1.818000e-05      1        1821   \n",
       "4        1    137620    139379  exon  1.000000e-08   1759        1822   \n",
       "5        1    139379    139380   gap  2.282600e-03      1        3581   \n",
       "6        1    367639    368634  exon  1.000000e-08    995        3582   \n",
       "7        1    368634    368635   gap  2.524240e-03      1        4577   \n",
       "8        1    621058    622053  exon  1.000000e-08    995        4578   \n",
       "9        1    622053    622054   gap  5.000000e-01      1        5573   \n",
       "10       2     32403     32404   snp           NaN      1        5574   \n",
       "11       2     32404     32405   gap  1.731000e-05      1        5575   \n",
       "12       2     34135     34136   snp           NaN      1        5576   \n",
       "13       2     34136     34137   gap  1.315000e-05      1        5577   \n",
       "14       2     35451     35452   snp           NaN      1        5578   \n",
       "15       2     35452     35453   gap  3.361000e-05      1        5579   \n",
       "16       2     38813     41627  exon  1.000000e-08   2814        5580   \n",
       "17       2     41627     41628   gap  3.812000e-05      1        8394   \n",
       "18       2     45439     46385  exon  1.000000e-08    946        8395   \n",
       "19       2     46385     46386   gap  5.000000e-01      1        9341   \n",
       "20       3    100338    100339   snp           NaN      1        9342   \n",
       "21       3    100339    100340   gap  1.842000e-05      1        9343   \n",
       "22       3    102181    102182   snp           NaN      1        9344   \n",
       "23       3    102182    102183   gap  2.200000e-07      1        9345   \n",
       "24       3    102204    102205   snp           NaN      1        9346   \n",
       "25       3    102205    102206   gap  5.937000e-05      1        9347   \n",
       "26       3    108142    108143   snp           NaN      1        9348   \n",
       "27       3    108143    108144   gap  5.472000e-05      1        9349   \n",
       "28       3    113615    113616   snp           NaN      1        9350   \n",
       "29       3    113616    113617   gap  5.000000e-01      1        9351   \n",
       "..     ...       ...       ...   ...           ...    ...         ...   \n",
       "189     19    282809    282810   gap  5.000000e-01      1       31909   \n",
       "190     20     68350     68408  exon  1.000000e-08     58       31910   \n",
       "191     20     68408     68409   gap  8.237000e-05      1       31968   \n",
       "192     20     76645     77174  exon  1.000000e-08    529       31969   \n",
       "193     20     77174     77175   gap  3.549000e-05      1       32498   \n",
       "194     20     80723     80724   snp           NaN      1       32499   \n",
       "195     20     80724     80725   gap  3.793000e-05      1       32500   \n",
       "196     20     84517     84518   snp           NaN      1       32501   \n",
       "197     20     84518     84519   gap  9.830000e-06      1       32502   \n",
       "198     20     85501     85502   snp           NaN      1       32503   \n",
       "199     20     85502     85503   gap  5.000000e-01      1       32504   \n",
       "200     21  10906200  10907040  exon  1.000000e-08    840       32505   \n",
       "201     21  10907040  10907041   gap  1.784000e-05      1       33345   \n",
       "202     21  10908824  10908895  exon  1.000000e-08     71       33346   \n",
       "203     21  10908895  10908896   gap  1.411000e-05      1       33417   \n",
       "204     21  10910306  10910399  exon  1.000000e-08     93       33418   \n",
       "205     21  10910399  10910400   gap  3.963000e-05      1       33511   \n",
       "206     21  10914362  10914442  exon  1.000000e-08     80       33512   \n",
       "207     21  10914442  10914443   gap  1.927000e-05      1       33592   \n",
       "208     21  10916369  10916475  exon  1.000000e-08    106       33593   \n",
       "209     21  10916475  10916476   gap  5.000000e-01      1       33699   \n",
       "210     22  16256440  16256677  exon  1.000000e-08    237       33700   \n",
       "211     22  16256677  16256678   gap  1.507000e-05      1       33937   \n",
       "212     22  16258184  16258303  exon  1.000000e-08    119       33938   \n",
       "213     22  16258303  16258304   gap  8.625000e-05      1       34057   \n",
       "214     22  16266928  16267095  exon  1.000000e-08    167       34058   \n",
       "215     22  16267095  16267096   gap  1.041000e-05      1       34225   \n",
       "216     22  16268136  16268181  exon  1.000000e-08     45       34226   \n",
       "217     22  16268181  16268182   gap  1.691000e-05      1       34271   \n",
       "218     22  16269872  16269943  exon  1.000000e-08     71       34272   \n",
       "\n",
       "     slim_end  \n",
       "0         917  \n",
       "1         918  \n",
       "2        1820  \n",
       "3        1821  \n",
       "4        3580  \n",
       "5        3581  \n",
       "6        4576  \n",
       "7        4577  \n",
       "8        5572  \n",
       "9        5573  \n",
       "10       5574  \n",
       "11       5575  \n",
       "12       5576  \n",
       "13       5577  \n",
       "14       5578  \n",
       "15       5579  \n",
       "16       8393  \n",
       "17       8394  \n",
       "18       9340  \n",
       "19       9341  \n",
       "20       9342  \n",
       "21       9343  \n",
       "22       9344  \n",
       "23       9345  \n",
       "24       9346  \n",
       "25       9347  \n",
       "26       9348  \n",
       "27       9349  \n",
       "28       9350  \n",
       "29       9351  \n",
       "..        ...  \n",
       "189     31909  \n",
       "190     31967  \n",
       "191     31968  \n",
       "192     32497  \n",
       "193     32498  \n",
       "194     32499  \n",
       "195     32500  \n",
       "196     32501  \n",
       "197     32502  \n",
       "198     32503  \n",
       "199     32504  \n",
       "200     33344  \n",
       "201     33345  \n",
       "202     33416  \n",
       "203     33417  \n",
       "204     33510  \n",
       "205     33511  \n",
       "206     33591  \n",
       "207     33592  \n",
       "208     33698  \n",
       "209     33699  \n",
       "210     33936  \n",
       "211     33937  \n",
       "212     34056  \n",
       "213     34057  \n",
       "214     34224  \n",
       "215     34225  \n",
       "216     34270  \n",
       "217     34271  \n",
       "218     34342  \n",
       "\n",
       "[219 rows x 8 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concatenate_regions(create_recomb_map(test_data))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
