{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The basic idea is to find out which categories of genomic regions (exons, promoters, enhancers, etc) contribute the most to the Nea. ancestry observed in present day Europeans. For example: have coding regions or enhancers been the main drivers of selection against Nea. introgression?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Links:\n",
    "\n",
    "* info about different coordinate types: https://groups.google.com/forum/#!msg/biomart-users/OtQbAx3y9CA/wrF19ID1AgAJ\n",
    "* https://www.biostars.org/p/2005/\n",
    "* http://www.bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/biomaRt.html#attribute-pages\n",
    "* http://www.ensembl.org/info/data/biomart/biomart_r_package.html\n",
    "* http://www.ensembl.info/blog/2015/06/01/biomart-or-how-to-access-the-ensembl-data-from-r/\n",
    "* for checking with manually downloaded data: http://www.ensembl.org/info/data/biomart/how_to_use_biomart.html"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "source(\"http://bioconductor.org/biocLite.R\")\n",
    "biocLite(\"biomaRt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Following [this](http://www.ensembl.org/info/data/biomart/biomart_r_package.html) tutorial, I want to extract coordinates of exonic and regulatory regions from the Ensembl database and then calculate the density of such regions in a defined window around each SNP.\n",
    "\n",
    "These densities will they be used as predictors in a linear model, predicting the Nea. ancestry at each site.\n",
    "\n",
    "Alternatively, I could just test if the distribution of densities for different regions differ based on frequency of Nea. alleles at each site."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(tidyverse)\n",
    "library(stringr)\n",
    "library(magrittr)\n",
    "library(parallel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "library(biomaRt)\n",
    "library(rtracklayer)\n",
    "library(BSgenome.Hsapiens.UCSC.hg19)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get coordinates of admixture array sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "source(\"../R/utils.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Missing column names filled in: 'X12' [12], 'X13' [13], 'X14' [14], 'X15' [15], 'X16' [16], 'X17' [17], 'X18' [18], 'X19' [19], 'X20' [20], 'X21' [21], 'X22' [22], 'X23' [23], 'X24' [24]”Parsed with column specification:\n",
      "cols(\n",
      "  .default = col_character(),\n",
      "  Latitude = col_double(),\n",
      "  Longitude = col_double(),\n",
      "  Coverage = col_integer(),\n",
      "  HetRateAuto = col_double()\n",
      ")\n",
      "See spec(...) for full column specifications.\n",
      "Parsed with column specification:\n",
      "cols(\n",
      "  .default = col_integer(),\n",
      "  ref = col_character(),\n",
      "  alt = col_character()\n",
      ")\n",
      "See spec(...) for full column specifications.\n",
      "Parsed with column specification:\n",
      "cols(\n",
      "  .default = col_integer(),\n",
      "  ref = col_character(),\n",
      "  alt = col_character()\n",
      ")\n",
      "See spec(...) for full column specifications.\n",
      "Parsed with column specification:\n",
      "cols(\n",
      "  chrom = col_integer(),\n",
      "  pos = col_integer(),\n",
      "  ref = col_character(),\n",
      "  alt = col_character(),\n",
      "  Altai = col_integer(),\n",
      "  Vindija = col_integer(),\n",
      "  Denisovan = col_integer()\n",
      ")\n",
      "Joining, by = c(\"chrom\", \"pos\", \"ref\", \"alt\")\n",
      "Joining, by = c(\"chrom\", \"pos\", \"ref\", \"alt\")\n"
     ]
    }
   ],
   "source": [
    "european_ids <- get_european_ids(\"../raw_data/10_24_2014_SGDP_metainformation_update.txt\")\n",
    "\n",
    "europeans <-\n",
    "    load_dataset(\"../clean_data/ice_age.tsv\", \"../clean_data/sgdp.tsv\", \"../clean_data/archaics.tsv\") %>%\n",
    "    dplyr::select(c(\"chrom\", \"pos\", european_ids)) %>% \n",
    "    dplyr::mutate(chrom=paste0(\"chr\", chrom), start=pos, end=pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "sites <- makeGRangesFromDataFrame(europeans)\n",
    "seqinfo(sites) <- seqinfo(BSgenome.Hsapiens.UCSC.hg19)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "484016"
      ],
      "text/latex": [
       "484016"
      ],
      "text/markdown": [
       "484016"
      ],
      "text/plain": [
       "[1] 484016"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(sites)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 6 ranges and 0 metadata columns:\n",
       "      seqnames           ranges strand\n",
       "         <Rle>        <IRanges>  <Rle>\n",
       "  [1]     chr1 [847983, 847983]      *\n",
       "  [2]     chr1 [853089, 853089]      *\n",
       "  [3]     chr1 [853596, 853596]      *\n",
       "  [4]     chr1 [854793, 854793]      *\n",
       "  [5]     chr1 [867552, 867552]      *\n",
       "  [6]     chr1 [871401, 871401]      *\n",
       "  -------\n",
       "  seqinfo: 93 sequences (1 circular) from hg19 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(sites)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extracting data from biomart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>biomart</th><th scope=col>version</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>ensembl              </td><td>Ensembl Genes 75     </td></tr>\n",
       "\t<tr><td>snp                  </td><td>Ensembl Variation 75 </td></tr>\n",
       "\t<tr><td>regulation           </td><td>Ensembl Regulation 75</td></tr>\n",
       "\t<tr><td>vega                 </td><td>Vega 55              </td></tr>\n",
       "\t<tr><td>pride                </td><td>PRIDE (EBI UK)       </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ll}\n",
       " biomart & version\\\\\n",
       "\\hline\n",
       "\t ensembl               & Ensembl Genes 75     \\\\\n",
       "\t snp                   & Ensembl Variation 75 \\\\\n",
       "\t regulation            & Ensembl Regulation 75\\\\\n",
       "\t vega                  & Vega 55              \\\\\n",
       "\t pride                 & PRIDE (EBI UK)       \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "biomart | version | \n",
       "|---|---|---|---|---|\n",
       "| ensembl               | Ensembl Genes 75      | \n",
       "| snp                   | Ensembl Variation 75  | \n",
       "| regulation            | Ensembl Regulation 75 | \n",
       "| vega                  | Vega 55               | \n",
       "| pride                 | PRIDE (EBI UK)        | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  biomart    version              \n",
       "1 ensembl    Ensembl Genes 75     \n",
       "2 snp        Ensembl Variation 75 \n",
       "3 regulation Ensembl Regulation 75\n",
       "4 vega       Vega 55              \n",
       "5 pride      PRIDE (EBI UK)       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "listEnsembl(version=\"75\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ensembl <- useEnsembl(biomart=\"ensembl\", version=75)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find the name of the _H. sapiens_ dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>dataset</th><th scope=col>description</th><th scope=col>version</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>hsapiens_gene_ensembl          </td><td>Homo sapiens genes (GRCh37.p13)</td><td>GRCh37.p13                     </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lll}\n",
       " dataset & description & version\\\\\n",
       "\\hline\n",
       "\t hsapiens\\_gene\\_ensembl       & Homo sapiens genes (GRCh37.p13) & GRCh37.p13                     \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "dataset | description | version | \n",
       "|---|\n",
       "| hsapiens_gene_ensembl           | Homo sapiens genes (GRCh37.p13) | GRCh37.p13                      | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  dataset               description                     version   \n",
       "1 hsapiens_gene_ensembl Homo sapiens genes (GRCh37.p13) GRCh37.p13"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "listDatasets(ensembl) %>% filter(str_detect(dataset, \"sapiens\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting exon coordinates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Connect to the human gene Ensembl dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genes <- useDataset(dataset=\"hsapiens_gene_ensembl\", mart=ensembl)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Show all available attributes that can be pulled out of biomart:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>name</th><th scope=col>description</th><th scope=col>page</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>ensembl_gene_id        </td><td>Ensembl Gene ID        </td><td>structure              </td></tr>\n",
       "\t<tr><td>ensembl_transcript_id  </td><td>Ensembl Transcript ID  </td><td>structure              </td></tr>\n",
       "\t<tr><td>ensembl_peptide_id     </td><td>Ensembl Protein ID     </td><td>structure              </td></tr>\n",
       "\t<tr><td>chromosome_name        </td><td>Chromosome Name        </td><td>structure              </td></tr>\n",
       "\t<tr><td>start_position         </td><td>Gene Start (bp)        </td><td>structure              </td></tr>\n",
       "\t<tr><td>end_position           </td><td>Gene End (bp)          </td><td>structure              </td></tr>\n",
       "\t<tr><td>transcript_start       </td><td>Transcript Start (bp)  </td><td>structure              </td></tr>\n",
       "\t<tr><td>transcript_end         </td><td>Transcript End (bp)    </td><td>structure              </td></tr>\n",
       "\t<tr><td>strand                 </td><td>Strand                 </td><td>structure              </td></tr>\n",
       "\t<tr><td>external_gene_id       </td><td>Associated Gene Name   </td><td>structure              </td></tr>\n",
       "\t<tr><td>external_gene_db       </td><td>Associated Gene DB     </td><td>structure              </td></tr>\n",
       "\t<tr><td>5_utr_start            </td><td>5' UTR Start           </td><td>structure              </td></tr>\n",
       "\t<tr><td>5_utr_end              </td><td>5' UTR End             </td><td>structure              </td></tr>\n",
       "\t<tr><td>3_utr_start            </td><td>3' UTR Start           </td><td>structure              </td></tr>\n",
       "\t<tr><td>3_utr_end              </td><td>3' UTR End             </td><td>structure              </td></tr>\n",
       "\t<tr><td>cds_length             </td><td>CDS Length             </td><td>structure              </td></tr>\n",
       "\t<tr><td>transcript_count       </td><td>Transcript count       </td><td>structure              </td></tr>\n",
       "\t<tr><td>description            </td><td>Description            </td><td>structure              </td></tr>\n",
       "\t<tr><td>gene_biotype           </td><td>Gene Biotype           </td><td>structure              </td></tr>\n",
       "\t<tr><td>exon_chrom_start       </td><td>Exon Chr Start (bp)    </td><td>structure              </td></tr>\n",
       "\t<tr><td>exon_chrom_end         </td><td>Exon Chr End (bp)      </td><td>structure              </td></tr>\n",
       "\t<tr><td>is_constitutive        </td><td>Constitutive Exon      </td><td>structure              </td></tr>\n",
       "\t<tr><td>rank                   </td><td>Exon Rank in Transcript</td><td>structure              </td></tr>\n",
       "\t<tr><td>phase                  </td><td>phase                  </td><td>structure              </td></tr>\n",
       "\t<tr><td>cdna_coding_start      </td><td>cDNA coding start      </td><td>structure              </td></tr>\n",
       "\t<tr><td>cdna_coding_end        </td><td>cDNA coding end        </td><td>structure              </td></tr>\n",
       "\t<tr><td>genomic_coding_start   </td><td>Genomic coding start   </td><td>structure              </td></tr>\n",
       "\t<tr><td>genomic_coding_end     </td><td>Genomic coding end     </td><td>structure              </td></tr>\n",
       "\t<tr><td>ensembl_exon_id        </td><td>Ensembl Exon ID        </td><td>structure              </td></tr>\n",
       "\t<tr><td>cds_start              </td><td>CDS Start              </td><td>structure              </td></tr>\n",
       "\t<tr><td>cds_end                </td><td>CDS End                </td><td>structure              </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lll}\n",
       " name & description & page\\\\\n",
       "\\hline\n",
       "\t ensembl\\_gene\\_id       & Ensembl Gene ID           & structure                \\\\\n",
       "\t ensembl\\_transcript\\_id & Ensembl Transcript ID     & structure                \\\\\n",
       "\t ensembl\\_peptide\\_id    & Ensembl Protein ID        & structure                \\\\\n",
       "\t chromosome\\_name       & Chromosome Name         & structure              \\\\\n",
       "\t start\\_position        & Gene Start (bp)         & structure              \\\\\n",
       "\t end\\_position          & Gene End (bp)           & structure              \\\\\n",
       "\t transcript\\_start      & Transcript Start (bp)   & structure              \\\\\n",
       "\t transcript\\_end        & Transcript End (bp)     & structure              \\\\\n",
       "\t strand                  & Strand                  & structure              \\\\\n",
       "\t external\\_gene\\_id      & Associated Gene Name      & structure                \\\\\n",
       "\t external\\_gene\\_db      & Associated Gene DB        & structure                \\\\\n",
       "\t 5\\_utr\\_start           & 5' UTR Start              & structure                \\\\\n",
       "\t 5\\_utr\\_end             & 5' UTR End                & structure                \\\\\n",
       "\t 3\\_utr\\_start           & 3' UTR Start              & structure                \\\\\n",
       "\t 3\\_utr\\_end             & 3' UTR End                & structure                \\\\\n",
       "\t cds\\_length            & CDS Length              & structure              \\\\\n",
       "\t transcript\\_count      & Transcript count        & structure              \\\\\n",
       "\t description             & Description             & structure              \\\\\n",
       "\t gene\\_biotype          & Gene Biotype            & structure              \\\\\n",
       "\t exon\\_chrom\\_start      & Exon Chr Start (bp)       & structure                \\\\\n",
       "\t exon\\_chrom\\_end        & Exon Chr End (bp)         & structure                \\\\\n",
       "\t is\\_constitutive       & Constitutive Exon       & structure              \\\\\n",
       "\t rank                    & Exon Rank in Transcript & structure              \\\\\n",
       "\t phase                   & phase                   & structure              \\\\\n",
       "\t cdna\\_coding\\_start     & cDNA coding start         & structure                \\\\\n",
       "\t cdna\\_coding\\_end       & cDNA coding end           & structure                \\\\\n",
       "\t genomic\\_coding\\_start  & Genomic coding start      & structure                \\\\\n",
       "\t genomic\\_coding\\_end    & Genomic coding end        & structure                \\\\\n",
       "\t ensembl\\_exon\\_id       & Ensembl Exon ID           & structure                \\\\\n",
       "\t cds\\_start             & CDS Start               & structure              \\\\\n",
       "\t cds\\_end               & CDS End                 & structure              \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "name | description | page | \n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| ensembl_gene_id         | Ensembl Gene ID         | structure               | \n",
       "| ensembl_transcript_id   | Ensembl Transcript ID   | structure               | \n",
       "| ensembl_peptide_id      | Ensembl Protein ID      | structure               | \n",
       "| chromosome_name         | Chromosome Name         | structure               | \n",
       "| start_position          | Gene Start (bp)         | structure               | \n",
       "| end_position            | Gene End (bp)           | structure               | \n",
       "| transcript_start        | Transcript Start (bp)   | structure               | \n",
       "| transcript_end          | Transcript End (bp)     | structure               | \n",
       "| strand                  | Strand                  | structure               | \n",
       "| external_gene_id        | Associated Gene Name    | structure               | \n",
       "| external_gene_db        | Associated Gene DB      | structure               | \n",
       "| 5_utr_start             | 5' UTR Start            | structure               | \n",
       "| 5_utr_end               | 5' UTR End              | structure               | \n",
       "| 3_utr_start             | 3' UTR Start            | structure               | \n",
       "| 3_utr_end               | 3' UTR End              | structure               | \n",
       "| cds_length              | CDS Length              | structure               | \n",
       "| transcript_count        | Transcript count        | structure               | \n",
       "| description             | Description             | structure               | \n",
       "| gene_biotype            | Gene Biotype            | structure               | \n",
       "| exon_chrom_start        | Exon Chr Start (bp)     | structure               | \n",
       "| exon_chrom_end          | Exon Chr End (bp)       | structure               | \n",
       "| is_constitutive         | Constitutive Exon       | structure               | \n",
       "| rank                    | Exon Rank in Transcript | structure               | \n",
       "| phase                   | phase                   | structure               | \n",
       "| cdna_coding_start       | cDNA coding start       | structure               | \n",
       "| cdna_coding_end         | cDNA coding end         | structure               | \n",
       "| genomic_coding_start    | Genomic coding start    | structure               | \n",
       "| genomic_coding_end      | Genomic coding end      | structure               | \n",
       "| ensembl_exon_id         | Ensembl Exon ID         | structure               | \n",
       "| cds_start               | CDS Start               | structure               | \n",
       "| cds_end                 | CDS End                 | structure               | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "   name                  description             page     \n",
       "1  ensembl_gene_id       Ensembl Gene ID         structure\n",
       "2  ensembl_transcript_id Ensembl Transcript ID   structure\n",
       "3  ensembl_peptide_id    Ensembl Protein ID      structure\n",
       "4  chromosome_name       Chromosome Name         structure\n",
       "5  start_position        Gene Start (bp)         structure\n",
       "6  end_position          Gene End (bp)           structure\n",
       "7  transcript_start      Transcript Start (bp)   structure\n",
       "8  transcript_end        Transcript End (bp)     structure\n",
       "9  strand                Strand                  structure\n",
       "10 external_gene_id      Associated Gene Name    structure\n",
       "11 external_gene_db      Associated Gene DB      structure\n",
       "12 5_utr_start           5' UTR Start            structure\n",
       "13 5_utr_end             5' UTR End              structure\n",
       "14 3_utr_start           3' UTR Start            structure\n",
       "15 3_utr_end             3' UTR End              structure\n",
       "16 cds_length            CDS Length              structure\n",
       "17 transcript_count      Transcript count        structure\n",
       "18 description           Description             structure\n",
       "19 gene_biotype          Gene Biotype            structure\n",
       "20 exon_chrom_start      Exon Chr Start (bp)     structure\n",
       "21 exon_chrom_end        Exon Chr End (bp)       structure\n",
       "22 is_constitutive       Constitutive Exon       structure\n",
       "23 rank                  Exon Rank in Transcript structure\n",
       "24 phase                 phase                   structure\n",
       "25 cdna_coding_start     cDNA coding start       structure\n",
       "26 cdna_coding_end       cDNA coding end         structure\n",
       "27 genomic_coding_start  Genomic coding start    structure\n",
       "28 genomic_coding_end    Genomic coding end      structure\n",
       "29 ensembl_exon_id       Ensembl Exon ID         structure\n",
       "30 cds_start             CDS Start               structure\n",
       "31 cds_end               CDS End                 structure"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "listAttributes(genes) %>% filter(page==\"structure\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Show the available filters for extracting protein coding genes from the given mart:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>name</th><th scope=col>description</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>type                          </td><td>Type                          </td></tr>\n",
       "\t<tr><td>with_exp_est_celltype         </td><td>with Expression est cell type </td></tr>\n",
       "\t<tr><td>with_exp_atlas_celltype       </td><td>with GNF/Atlas cell type ID(s)</td></tr>\n",
       "\t<tr><td>biotype                       </td><td>Type                          </td></tr>\n",
       "\t<tr><td>phenotype_description         </td><td>Phenotype description         </td></tr>\n",
       "\t<tr><td>event_type                    </td><td>Event Type                    </td></tr>\n",
       "\t<tr><td>cell_type_term                </td><td>Cell type                     </td></tr>\n",
       "\t<tr><td>atlas_celltype                </td><td>                              </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ll}\n",
       " name & description\\\\\n",
       "\\hline\n",
       "\t type                           & Type                          \\\\\n",
       "\t with\\_exp\\_est\\_celltype    & with Expression est cell type \\\\\n",
       "\t with\\_exp\\_atlas\\_celltype  & with GNF/Atlas cell type ID(s)\\\\\n",
       "\t biotype                        & Type                          \\\\\n",
       "\t phenotype\\_description        & Phenotype description         \\\\\n",
       "\t event\\_type                   & Event Type                    \\\\\n",
       "\t cell\\_type\\_term             & Cell type                     \\\\\n",
       "\t atlas\\_celltype               &                               \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "name | description | \n",
       "|---|---|---|---|---|---|---|---|\n",
       "| type                           | Type                           | \n",
       "| with_exp_est_celltype          | with Expression est cell type  | \n",
       "| with_exp_atlas_celltype        | with GNF/Atlas cell type ID(s) | \n",
       "| biotype                        | Type                           | \n",
       "| phenotype_description          | Phenotype description          | \n",
       "| event_type                     | Event Type                     | \n",
       "| cell_type_term                 | Cell type                      | \n",
       "| atlas_celltype                 |                                | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  name                    description                   \n",
       "1 type                    Type                          \n",
       "2 with_exp_est_celltype   with Expression est cell type \n",
       "3 with_exp_atlas_celltype with GNF/Atlas cell type ID(s)\n",
       "4 biotype                 Type                          \n",
       "5 phenotype_description   Phenotype description         \n",
       "6 event_type              Event Type                    \n",
       "7 cell_type_term          Cell type                     \n",
       "8 atlas_celltype                                        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "listFilters(genes) %>% filter(str_detect(name, \"type\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The \"biotype\" filter is what we are interested in, specifically its \"protein_coding\" subset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "exons <-\n",
    "    getBM((listAttributes(genes) %>% filter(page==\"structure\"))$name,    # get all attributes\n",
    "          filters=c(\"chromosome_name\", \"biotype\"),                       # filter for chromosome and biotype\n",
    "          values=list(1:22, \"protein_coding\"),                           # chromosomes 1:22 and \"protein_coding\"\n",
    "          mart=genes) %>%\n",
    "    dplyr::select(chromosome_name, exon_chrom_start, exon_chrom_end) %>% # take only exon coordinates\n",
    "    setNames(c(\"chrom\", \"start\", \"end\")) %>%                             # rename the columns to BED names\n",
    "    mutate(chrom=paste0(\"chr\", chrom)) %>%                               # convert chromosome IDs to USCS format\n",
    "    makeGRangesFromDataFrame %>%                                         # convert to GRanges object\n",
    "    reduce                                                               # merge overlapping exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "seqinfo(exons) <- seqinfo(BSgenome.Hsapiens.UCSC.hg19)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many exons did I get? What is the total length of the exonic sequence (in Mb) that we extracted?\n",
    "\n",
    "Bioinformatic databases are a mess - are the numbers I got reasonable?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "186343"
      ],
      "text/latex": [
       "186343"
      ],
      "text/markdown": [
       "186343"
      ],
      "text/plain": [
       "[1] 186343"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "75.235862"
      ],
      "text/latex": [
       "75.235862"
      ],
      "text/markdown": [
       "75.235862"
      ],
      "text/plain": [
       "[1] 75.23586"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(exons)\n",
    "sum(width(exons)) / 1e6"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculate the overlap of each window with each annotation feature"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Those black magic GRanges overlap tricks are from [here](https://support.bioconductor.org/p/56880/)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Developing the code on a simple test example"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Testing sites, testing windows and testing exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 4 ranges and 0 metadata columns:\n",
       "      seqnames     ranges strand\n",
       "         <Rle>  <IRanges>  <Rle>\n",
       "  [1]     chr1 [ 10,  10]      *\n",
       "  [2]     chr1 [ 16,  16]      *\n",
       "  [3]     chr1 [ 22,  22]      *\n",
       "  [4]     chr1 [100, 100]      *\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "GRanges object with 4 ranges and 0 metadata columns:\n",
       "      seqnames    ranges strand\n",
       "         <Rle> <IRanges>  <Rle>\n",
       "  [1]     chr1 [ 5,  14]      *\n",
       "  [2]     chr1 [11,  20]      *\n",
       "  [3]     chr1 [17,  26]      *\n",
       "  [4]     chr1 [95, 104]      *\n",
       "  -------\n",
       "  seqinfo: 93 sequences (1 circular) from hg19 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "GRanges object with 4 ranges and 0 metadata columns:\n",
       "      seqnames     ranges strand\n",
       "         <Rle>  <IRanges>  <Rle>\n",
       "  [1]     chr1 [  7,   9]      *\n",
       "  [2]     chr1 [ 12,  13]      *\n",
       "  [3]     chr1 [ 16,  21]      *\n",
       "  [4]     chr1 [200, 220]      *\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# sites\n",
    "(s <- GRanges(seqnames=\"chr1\", ranges=IRanges(c(10, 16, 22, 100), width=1)))\n",
    "seqinfo(s) <- seqinfo(BSgenome.Hsapiens.UCSC.hg19)\n",
    "\n",
    "# windows around sites\n",
    "(w <- flank(s, 5, both=TRUE))\n",
    "seqinfo(w) <- seqinfo(BSgenome.Hsapiens.UCSC.hg19)\n",
    "\n",
    "# regions of interest (like exons)\n",
    "(r <- GRanges(seqnames=\"chr1\", ranges=IRanges(start=c(7, 12, 16, 200), end=c(9, 13, 21, 220))))\n",
    "seqinfo(r) <- seqinfo(BSgenome.Hsapiens.UCSC.hg19)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate the # of overlapping bases for each window:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Hits object with 5 hits and 0 metadata columns:\n",
       "      queryHits subjectHits\n",
       "      <integer>   <integer>\n",
       "  [1]         1           1\n",
       "  [2]         1           2\n",
       "  [3]         2           2\n",
       "  [4]         2           3\n",
       "  [5]         3           3\n",
       "  -------\n",
       "  queryLength: 4 / subjectLength: 4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# get the hit table between windows and regions they overlap with\n",
    "(hits <- findOverlaps(w, r))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 5 ranges and 1 metadata column:\n",
       "      seqnames    ranges strand |       hit\n",
       "         <Rle> <IRanges>  <Rle> | <logical>\n",
       "  [1]     chr1  [ 7,  9]      * |      TRUE\n",
       "  [2]     chr1  [12, 13]      * |      TRUE\n",
       "  [3]     chr1  [12, 13]      * |      TRUE\n",
       "  [4]     chr1  [16, 20]      * |      TRUE\n",
       "  [5]     chr1  [17, 21]      * |      TRUE\n",
       "  -------\n",
       "  seqinfo: 93 sequences (1 circular) from hg19 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# get all overlapping regions (duplicates for regions\n",
    "# overlapping several windows/sites)\n",
    "(w_over <- pintersect(w[queryHits(hits)], r[subjectHits(hits)]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl class=dl-horizontal>\n",
       "\t<dt>1</dt>\n",
       "\t\t<dd>5</dd>\n",
       "\t<dt>2</dt>\n",
       "\t\t<dd>7</dd>\n",
       "\t<dt>3</dt>\n",
       "\t\t<dd>5</dd>\n",
       "</dl>\n"
      ],
      "text/plain": [
       "1 2 3 \n",
       "5 7 5 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# for each window count the number of overlapping bases in regions\n",
    "# as an integer vector\n",
    "(w_counts <- tapply(w_over, queryHits(hits), FUN=function(x) sum(width(x))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 4 ranges and 1 metadata column:\n",
       "      seqnames     ranges strand |   overlap\n",
       "         <Rle>  <IRanges>  <Rle> | <numeric>\n",
       "  [1]     chr1 [ 10,  10]      * |       0.5\n",
       "  [2]     chr1 [ 16,  16]      * |       0.7\n",
       "  [3]     chr1 [ 22,  22]      * |       0.5\n",
       "  [4]     chr1 [100, 100]      * |         0\n",
       "  -------\n",
       "  seqinfo: 93 sequences (1 circular) from hg19 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# add the column with % overlap to the original sites object\n",
    "s$overlap <- 0\n",
    "s$overlap[as.numeric(names(w_counts))] <- unname(w_counts) / WINDOW_SIZE\n",
    "s"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The code seems to work, let's construct a function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "calc_window_overlaps <- function(sites, regions, window_size) {\n",
    "    # construct windows around each site\n",
    "    windows <- flank(sites, window_size, both=TRUE)\n",
    "    seqinfo(windows) <- seqinfo(sites)\n",
    "\n",
    "    # get the hit table between windows and regions they overlap with\n",
    "    hits <- findOverlaps(windows, regions)\n",
    "\n",
    "    # get the all overlapping regions (duplicates for regions\n",
    "    # overlapping several windows/sites)\n",
    "    w_over <- pintersect(windows[queryHits(hits)], regions[subjectHits(hits)])\n",
    "\n",
    "    # for each window count the number of overlapping bases in regions\n",
    "    # as an integer vector\n",
    "    w_counts <- tapply(w_over, queryHits(hits), FUN=function(x) { sum(width(x)) })\n",
    "\n",
    "    # add the column with % overlap to the original sites object\n",
    "    sites$overlap <- 0 # by default zero overlap for each site\n",
    "    sites$overlap[as.numeric(names(w_counts))] <- unname(w_counts) / window_size\n",
    "\n",
    "    mcols\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Test of this function on the toy dataset from above:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 4 ranges and 1 metadata column:\n",
       "      seqnames     ranges strand |   overlap\n",
       "         <Rle>  <IRanges>  <Rle> | <numeric>\n",
       "  [1]     chr1 [ 10,  10]      * |         1\n",
       "  [2]     chr1 [ 16,  16]      * |       1.4\n",
       "  [3]     chr1 [ 22,  22]      * |         1\n",
       "  [4]     chr1 [100, 100]      * |         0\n",
       "  -------\n",
       "  seqinfo: 93 sequences (1 circular) from hg19 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "calc_window_overlaps(s, r, 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "WINDOW_SIZE <- 100000\n",
    "\n",
    "exon_densities <-\n",
    "    mclapply(paste0(\"chr\", 1:22),\n",
    "             function(chr) {\n",
    "                 calc_window_overlaps(sites[seqnames(sites) == chr], exons, WINDOW_SIZE)\n",
    "             }, mc.cores=22) %>%\n",
    "    do.call(c, .)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
       "0.00000 0.00709 0.07943 0.09700 0.15620 0.64220 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "summary(mcols(x)$overlap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.\n"
     ]
    },
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAIAAAByhViMAAAABmJLR0QA/wD/AP+gvaeTAAAg\nAElEQVR4nOzdd4BcZaHw4XfO1G1JNoQSSMQAQgJ4aZEiIF2aYAGpSgcpEhEpGtTL9cIVBRWi\noF7kA9QLGEGKIoigiEr1IohAKNeAQIgIaZvM1pn5/hhYVk3ZhczO5p3n+WvnTDnvzJzy27Nn\nZlOVSiUAALDqS+o9AAAAVg5hBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQiUy9\nB1ArnZ2dXV1dNZ1FOp1ubW0NIXR0dJTL5ZrOa8TK5/OZTGbJkiX1HkjdtLW1JUnS1dXV3d1d\n77HURyqVGjVq1KJFixr2287z+XyhUCiXyx0dHfUeS920tLT09fU17FqQJElbW1sIYfHixaVS\nqd7DqY9cLpfL5RYvXlzvgdRNa2trOp3u6enp7Oys9bza29uXdVW0YVepVGq9diVJkiRJCKFc\nLjfsmlypVFKpVMM+/RBCKpWqLgYN+yJUV4Ryudywv96EEJIkGYZtzkiWSqUa/BWwO7A7GCG7\nA3+KBQCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLAD\nAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiERm\nGObRt+Sv11x2+e/+/Oy83vw7N9jikOOPmzqxJYQQQuW+ay+9/rcPv9CR2XjTrU447bi18+m3\nNB0AgGE4YlcpXX76WTc+8PIWu3zgo/u+r+//fvtfn/nc892lEMJjV07/8rV3FCZN3X+nd7/4\n0O2nT7uws1x5C9MBAAjDcMRuydwf3vZy8T3nzDhpmzVCCHu/t/SxT/308odf/dLU1Pm3PDFu\nq2nnn7l7CGGfrTNHff62GY/NO3Pj0pCmn73ZarV+CgAAq4SaH7HrXVjebLPNdpsypnox17ZG\nCKFSriyZc22xVNnksPdUp7dvckRzOjVr5uyhTq/1+AEAVhU1P2I3ZvLR//mfIYTw5K/veGre\nq3/61c+a1tz8xKmrL/rNnBDC+mPz1Zul0i3r5jPPz31y0VNDmx7C1OqUrq6uH/3oR/3z3WST\nTaZMmVLTp5ZOv36GX6FQKJfLNZ3XiJXNZpMkaWpqqvdA6iZJkhBCNptt2BchlUqFEAqFQqXS\noKdGZLPZEEIqlWrYZSCEkCRJ9XVoTNXtQAghn8837Otgd1BdDNLpdK1fhOVvbIfjwxNV91z9\nvZ/P765UKlP2mbpGLpmzqDeEMCaT6r9BeyaZ3Te/Z4jT+y92dnZ+85vf7L94wgknTJ06tZZP\n6E2NvBxXtbS01HsIdZbNZht2a17V3Nxc7yHUWZIkDb4ipNPpXC5X71HUmd1Bg68FYVh2B6VS\naTnXDl/YfeKqmZ8I4bVnf3XCZy75XNvGp4/LhBAW9r1ZnfP7ykluVLZtaNP7L6bT6XXWWaf/\nYltb2/Kf+duXSqWqeV7rGY1k1VegYQ9YhjcO3FYqlYZ9EaorQrlcbtgjdkmSVA9b2hQ07FoQ\n3tgUNPKKkEqlUqlUIy8D1U3BMOwOyuVy/98M/1XNw+7p733l6tnFL533H+lUCCGstsGuGzZd\nOvvXD47+wtohzJq9qCesVgghVCpdL3b3FdbdePTkBUOa3j+jUaNG3Xzzzf0Xi8Xi/PnzQy1l\ns9nRo0eHEBYtWtSwG/SmpqZcLrdw4cJ6D6Ru2tvb0+l0Z2dnsVis91jqI0mSsWPHLliwoGE3\n6M3Nzc3NzaVSqdbbnJFs9OjRPT09nZ2d9R5IfaTT6fb29hDCokWL+vr66j2c+igUCoVCYcGC\nBfUeSN2MGTMmk8l0d3cvXry41vMaN27csq6q+Ycnmie89thjf7x17uv7vFLXX/6vq68wbmLL\n+EOaktQTMx+vTu+YPbOjVNno4PWGOr3W4wcAWFWkzz333JrOoHXdzR7+6c9/c+eDi5YseObP\n/3vtd747t9h8+oUnTWgds2HxkZ/dedOcYmr+c49eOeOmrtW2ueC49+cybUOank2lljrf3t7e\n3t7emj61dDpdKBRCCF1dXQ177D2bzabT6e7u7noPpG6ampqSJBmG5W3Eqn5ooLOzs5HXgmw2\nW6lUurq66j2WuikUCqVSqWEPVvV/aKCrq6thD11nMplMJtPga0GSJH19fT09PbWe13JOa655\n2KXSLTvsOHne7CcefPD+R555efQ7tjpx+he2Xqs5hLDWFru/o/zKA/f/7p5H/7rOZjud86VT\nx2aTtzB9qYTd8BB2wk7YCbsg7ISdsBsxYZeKdVtcLBZrfc5T/zl28+fPd45dvQdSN9Vz7IZh\neRuxqufYzZs3r2H3Z86xC86xe+McuwULFjRs3TrHrnqOXVdXV+Tn2AEAMDyEHQBAJIQdAEAk\nhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBA\nJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0A\nQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQd\nAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSE\nHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAk\nhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBA\nJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0A\nQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQd\nAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSE\nHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAk\nMvUeQK0kSdLU1FTTWaTT6eoPhUKhXC7XdF4jVjabHYaXeiRLkiSEkM1mG/ZFSKVSIYRCoVCp\nVOo9lvrIZrMhhFQq1bDLQAghSZLq69CYqtuBEEI+n2/Y1yGTydgdhBDS6XStX4Tlb2xjDrv+\n8KqR6v4shJDNZht2l5YkSZIkuVyu3gOps3Q63eAvQsPuzMIbv+OlUqlGXgaqu7T+rWKjsTsI\nISRJ0uBrQXUxGIbdwfKPJUUbdn19fcVisaazyGazo0ePDiF0dHSUSqWazmvEampqyuVyCxcu\nrPdA6qa9vT2dTnd1ddV6eRuxkiQZO3ZsR0dHwx63bm5ubm5uLpfLjbwijB49uqenp7Ozs94D\nqY90Ot3e3h5CWLx4cV9fX72HUx+FQqFQKDTyWjBmzJhMJtPT07N48eJazyufzy/rKufYAQBE\nQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEA\nRELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgB\nAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELY\nAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC\n2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBE\nQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEA\nRELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgB\nAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELY\nAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC\n2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBE\nQtgBAERC2AEARCIzDPOo9M2/9erv3nb/k3MXliZO2nD/j5+466ZrVK+579pLr//twy90ZDbe\ndKsTTjtu7Xz6LU0HAGBYjtjNPOe0y295aJ0tdzr4w3uMWfzkJeec+NO/dIQQHrty+pevvaMw\naer+O737xYduP33ahZ3lyluYDgBAGIYjdj2Lfv8/T86fuO9/Tf/EpiGE8oH7nnjo8ddccNu+\nl+58/i1PjNtq2vln7h5C2GfrzFGfv23GY/PO3Lg0pOlnb7ZarZ8CAMAqoeZH7Dpf+XUIYeMP\nvPP1+WXHbdaS7ev865I51xZLlU0Oe091evsmRzSnU7Nmzh7q9FqPHwBgVVHzI3ajN/j8Lbe8\neXHe4z+/a0H3uB12XvTUj0MI64/NV6en0i3r5jPPz31y0VNzhjQ9hKnVKcVi8eKLL+6f0Xvf\n+95tt922pk8tSV7P4ubm5kqlQf8onMlkkiRpbW2t90DqproY5HK5/uWh0aRSqRBCS0tLI68F\nIYQGXxHS6XQ+n0+nG/S85+paEEJobm4ul8v1HUy9pNPpBl8LqnuBbDZb6xdh+cvYcHx4oqpS\n6vjNjT+47Ie/SNo2OfvULXp+dk0IYUwm1X+D9kwyu29+z6LeIU3vv9jd3f2Tn/yk/+K4ceN2\n3nnnGj6fAfL5/PDMaMQqFAr1HkKdZTKZ6t69YVkLUqlUg68ISZI0+FoQQsjlcvUeQp01+FoQ\nQkin07X+DadUKi3n2mFaCV957I6Lvva9p+Z3T97hgE+ecvjEQvrFtkwIYWHfm7/iz+8rJ7lR\n2SFO77+YzWa33nrr/ovjx4/v7e2t5XMKqVSquhXr6+tr2GMVSZIkSdLX11fvgdRNJpNJpVLl\ncnn5a1rcstlsrVe3kax6oKJSqTT4ilAulxv2YJXdQbA7GMbdQblcXk47DkfYvXj3d075xm2t\n6257zhdO3Hr99urE0ZPXDmHW7EU9YbVCCKFS6Xqxu6+w7sajJy8Y0vT+ubS2tl522WX9F4vF\n4sKFC2v6vLLZ7OjRo0MIHR0dDbtTb2pqyuVytX6pR7L29vZ0Ot3V1VUsFus9lvpIkmTs2LEd\nHR0Nu1Nvbm6u/gGukVeE0aNH9/T0dHZ21nsg9ZFOp9vb20MIixcvbtiyKRQKhUKhkdeCMWPG\nZDKZnp6exYsX13pey/kjSc3PCip1zT7jktvW3OGY71382f6qCyG0jD+kKUk9MfPx6sWO2TM7\nSpWNDl5vqNNrPX4AgFVFzY/YvfLA5cVSZZ3081d+59v9E7PNU44/apfp+0/+4s0XfO2KQye3\nl+6aeXPzWtudtvlqSZIa0vRajx8AYFVR87Cb99BrIYRn7r7zmQETC+2l44/aZbNjLjgrP+OG\n391+f0dmyhZ7fv20Y5uSVAhhqNMBAAghpGI9zbNYLNb6nKf+c+zmz5/vHLt6D6RuqufYDcPy\nNmJVz7GbN29eg59jVyqV5s+fv+JbR8o5dtVz7BYsWNDg59gtWLCg3gOpm+o5dl1dXcNwjt24\nceOWdVWDfvMWAEB8hB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQd\nAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSE\nHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAk\nhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBA\nJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0A\nQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQd\nAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSE\nHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAk\nhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJIQdAEAkhB0AQCSEHQBAJAYbdhMmTDj07jn/Ov3R\n8/aYtNHeK3VIAAC8FZnlX33nnXdWf3jppZeevu/uO/vW+IerK30P3P/UX/+yqEaDAwBg8FYQ\ndnvssUf/zw9PP3yPpd1m1LqfXKlDAgDgrVhB2J1zzjnVH84///xJh5x62Pqj/ukG6Xz7Piee\nUpOhAQAwFCsIu/POO6/6w3XXXbf9SZ87733jaz8kAADeihWEXb9nn3223Lvwsft+9dLCnn+9\ndsOd91ivkF6pAwMAYGgGG3bFv9265+YH/W5ucanX3vpap7ADAKivwYbd5R869t6/Jx886uRN\n12n/12t3HpNfqaMCAGDIBht25//vq+/77p9vOnZyTUcDAMBbNtgvKG5Kp3bbY52aDgUAgLdj\nsGF3zpbjbp35VE2HAgDA2zHYsDvyZz+uXLjnJy/4/jMvvNJbqemQAAB4KwYbdhtt8bGXUp2X\nfu7IDd+xZj7zz2o6RAAABmOwTTZmzJgQxqy2Vk0HAwDAWzfYsHvkkUdqOg4AAN6mlRB2mfyE\nTaeMW0njAQDgLRps2G2xxRbLuqp9g0vnPXPyShoPAABv0WDD7thjjx14sdzT8dyzf/71fU+0\nTnz/f5z93hoMDACAoUlVKm/9y0v+eu/lm+904j5XPvXDj22wEse0UnR3d7+dpzYYSZLkcrnh\nmdeIlclkkiTp6emp90DqJp/Pp1Kpvr6+vr6+eo+lPlKpVD6fb/C1IJPJVCqV7u7ueo+lbnK5\nXLlcbvC1IITQ09NTLpfrPZz6SKfT6XS6kXcHuVwuSZJSqdTb21vTGZXL5ebm5mVdO9ivO1mq\nd7z3+OsPXO/Gz3zh7TwIAAArxdv9CrpRE5u7Xrt1pQxl5SqVSsVisaazyGaz1SN2xWKxVCrV\ndF4jVlNTUy6XW7x4cb0HUjfZbLb6S2qtl7cRK0mSfD6/ZMmShj1Q0dzcnMlkyuVyI68Io0eP\n7unp6ezsrPdA6iOdTleP2BWLxYY9bFkoFAqFQiOvBWPGjEmSpLe3dxhehFodseue96cvXvlM\npnmjt/MgAACsFIM9YjdhwoR/nlTp+/vcV3rKlS0+e8lKHhQAAEM32LBb6rHlNdZ793v3PeF7\n5223UocEAMBbMdiwmzt3bk3HAQDA2zS0D08sePquK35899NPP/333tYpU6bseMAxe208tkYj\nAwBgSAYfduWrzvjoyRff2Fl6/auqbgwhde4XPnz6d66/8MhUjUYHAMCgDfZTsf93zUFHf+0n\n+XftesH/m3n3vf/7x/vv+fGVX91lg/xPLjrqkB/9paZDBABgMAZ7xO6iM36RH7XN/f97+0bN\nr99l8212/NBB+79v/GY/P/3r4eBv1WyEAAAMymCP2M18tbjWDv/ZX3VVmeaNztt17eLfr63B\nwAAAGJrBhl1LkvQs7PjX6YsX9CTp1pU6JAAA3orBht209Ua98uAJ/zNrwcCJC5/50fH3zh01\naVoNBgYAwNAM9hy7Y2ae8/nNzjhi04k/OOzj205erzUsnv30gz/44S8WV7IX/ujomg4RAIDB\nGGzYjd309D//tHDMKZ/9xQ++/Ys3Jratu8M3v/X/Tnm3r7IDAKi/IXxB8Qb7nPybZz/+yIN/\nmDVr1oIw5l3rbrD1TlNHpX2HHQDAiDDYc+xCCK/+8UeHf3DfT/+49dCjTzrp6EPDRYdsvd2e\nV9/vX40BAIwIgz1it/DZb6+/9ScX9ZW3/bfXW3DUhpOe/8WdR++wXu8zc4+bNKpmIwQAYFAG\ne8Tuso98YUlq7NX3vnDff21VnbL1xXe+8szt6+d6P/fh79ZseAAADNZgw+7ipxasvtV/H7Hd\nhIET29Z7/7d2GD//yW/UYGAAAAzNYMMuFUK2rWkp07NJCKWVOiQAAN6KwYbdKeuOmvu7T939\natfAid0L7jvt7jmtax9Xg4EBADA0g/3wxEk/OvM/p57z/kkbH3bikVtvNGlMpuu5Zx++6ltX\nPdvZd+b3T63pEAEAGIzBht24LT736MzsEZ/896svOvfqNybm2t41/fLvn7fjWjUaHAAAgzeE\nLyiecsAZD+x/5K9+8ZtZs2b9rbtpypQp2+35/kktQ3gEAABqZ2hZlmRX3/0DB+7+gRoNBgCA\nt24I/3kCAICRTNgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBE\nQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEA\nRELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgB\nAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELYAQBEQtgBAERC2AEARELY\nAQBEQtgBAERC2AEARCJT7wGwqpo2bdpKeZwZM2aslMcBAByxAwCIhLADAIiEsAMAiISwAwCI\nhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIRKbeA2C4TZs2\nrd5DAABqwhE7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7\nAIBICDsAgEj4X7GrDP/jFQBYPkfsAAAi4YhdzTnSBgAMD0fsAAAiIewAACIh7AAAIiHsAAAi\nIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAikRnGeVU+d8iB\n2bO++6Utx/VPue/aS6//7cMvdGQ23nSrE047bu18+i1NBwBguI7YVUpLHrjpgseLvQMnPnbl\n9C9fe0dh0tT9d3r3iw/dfvq0CzvLlbcwHQCAMDxH7GZdesY5dzzTW/mHCCv3vnL+LU+M22ra\n+WfuHkLYZ+vMUZ+/bcZj887cuDSk6WdvttowPAUAgJFvOI7YrbnjR06eNu2kY3ccOHHJnGuL\npcomh72nerF9kyOa06lZM2cPdfowjB8AYJUwHEfs2v/tvbuF0PXqom9f8dv+iYuemhNCWH9s\nvnoxlW5ZN595fu6TQ50ewtTqlMWLF5911ln9j7/33nvvtddeNX1eqVSq+kNbW1ul4o/Cb9Ho\n0aPrPYS3JUmSEEKhUMhms/UeSz21tbXVewh1k06nQwhJkqzqC/PbkclkkiTJ5XL1Hkh99O8O\nWltbG3Z3kCRJg68F1U1BLper9YtQLpeXc+1wfnjiH/Qs6g0hjMmk+qe0Z5LZffOHOr3/Ym9v\n74MPPth/cfPNNx+2HW0mU7eXMQJx9FB1i1bvUdRTHO/j25FKpRr8RUin09UdWyOzO2jwtSAM\ny+6gVCot59q6LYLZtkwIYWHfm7/ZzO8rJ7lRQ53efzGfz3/kIx/pv7jhhht2dXXV8hmE/l9P\nu7u7G/ZXtLev1m9TreXz+VQq1dfX19fXV++x1Ecqlcrn8428FmQymUwmU6lUuru76z2Wusnl\ncuVyucHXghBCT0/P8o+mRKxa9j09PfUeSN3kcrkkSUqlUm9v74pv/TaUy+Xm5uZlXVu3sBs9\nee0QZs1e1BNWK4QQKpWuF7v7CutuPHrygiFN73/A5ubm6dOn918sFouLFy+u6VPIZrPVsCsW\ni8vPZ5aj1m9TrWWz2eq2rFgs1nss9ZEkST6fX7JkScPuz5qbmzOZTLlcXtUX5rdj9OjRPT09\nnZ2d9R5IfaTT6WrYFYvFhq3bQqFQKBQaeS0YM2ZMkiS9vb3D8CIsJ+zq9sejlvGHNCWpJ2Y+\nXr3YMXtmR6my0cHrDXV6fUYPADDy1O2IXZJba/r+k7948wVfu+LQye2lu2be3LzWdqdtvlqS\npIY0vV7jBwAYaep5mudmx1xwVn7GDb+7/f6OzJQt9vz6acc2Jam3MB0AgDCcYVcY9+Fbbvnw\nP05LbX/4p7Y//F9vO9TpAADU7xw7AABWLmEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQ\nCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcAEAlhBwAQCWEHABAJYQcA\nEAlhBwAQCWEHABAJYQcAEAlhBwAQiUy9B0CjmzZt2kp5nBkzZqyUxwGAVZcjdgAAkRB2AACR\nEHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAA\nkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYA\nAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2\nAACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQ\ndgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACR\nEHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AACREHYAAJEQdgAAkRB2AJR12RoAAB0qSURB\nVACREHYAAJEQdgAAkRB2AACREHYAAJHI1HsAsHJMmzZtpTzOjBkzVsrjAMDwc8QOACASwg4A\nIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIO\nACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEpl6D6BWMpnM6NGjazqLVCpV/aGtra1SqdR0\nXgyboS42SZKEEAqFQjabrc2IVg1tbW31HkLdpNPpEEKSJLXe5oxkmUwmSZJcLlfvgdRH/+6g\ntbW1YXcHSZI0+FpQ3RTkcrlavwjlcnk510YbduVyuVQq1XQW6XQ6k8mEEHp7e5f/KrMK6enp\nGdLtq8tAqVQa6h2jkUqlstlsb29vw+7PcrlckiSVSqVhl4EQQpIkpVKpt7e33gOpjyRJ7A4y\nmUw2m23ktSCdTqdSqWHYHVQqlXw+v6xrYw67zs7Oms4im80WCoUQQldXV60jkmEz1MWmUCik\n0+ne3t5aL28jVpIkzc3NXV1dDbs/q6ZtpVJp2GUghJDL5Rp5LUin001NTSGE7u7uvr6+eg+n\nPqobw4ZdBkII+Xy++hvOMLwIra2ty7oq2rCDt2batGkr5XFmzJixUh4HAAbPhycAACIh7AAA\nIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewA\nACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHs\nAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh\n7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAi\nIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIpGp9wAgTtOmTVspjzNj\nxoyV8jgANAJH7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh\n7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAi\nIewAACIh7AAAIiHsAAAiIewAACKRqfcAgOWZNm3a23+QGTNmvP0HAWDkc8QOACASwg4AIBLC\nDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACASwg4AIBLCDgAgEsIOACAS\nwg4AIBKZeg8AqLlp06atlMeZMWPGSnkcAGrEETsAgEgIOwCASAg7AIBICDsAgEgIOwCASAg7\nAIBICDsAgEgIOwCASAg7AIBICDsAgEgIOwCASPhfscBg+Z+zACOcI3YAAJEQdgAAkRB2AACR\nEHYAAJFYtT48Ubnv2kuv/+3DL3RkNt50qxNOO27tfLreQwIAGClWpSN2j105/cvX3lGYNHX/\nnd794kO3nz7tws5ypd6DAgAYKVaZI3bl3lfOv+WJcVtNO//M3UMI+2ydOerzt814bN7Zm61W\n76EBQ+NrUwBqZJU5YrdkzrXFUmWTw95Tvdi+yRHN6dSsmbPrOyoAgJFjlTlit+ipOSGE9cfm\nqxdT6ZZ185nn5z4ZwtTXb7Bo0cc//vH+2x9yyCEHHXRQTYeUSqWqP4waNaqmMwKWqr29vd5D\nCCGEJElCCOl0+u2PZ+BGLCY/+MEP6j2EYTJq1KhKpUHPEUqlUqlUaoSslXVR3RTk8/lsNlvT\nGZXL5eVcu8qEXc+i3hDCmEyqf0p7JpndN7//YqlUeumll/ovdnR0pNPD9NGKYZsRMFCsGRQf\n7xRxuOaaa1Z4m1QqVd8qWGXCLtuWCSEs7HvzN6H5feUk9+ahskKhcOSRR/Zf3HTTTTs7O2s6\npCRJ8vl8CKGrq2s5v6JdccUVNR1GfWUymXQ63d3dXe+B1E2hUEilUn19fb29vfUeS32kUqlC\nobD8tSBu2Ww2k8lUKpWurq56j6Vu8vl8qVTq6+ur90Dqo7oWhBC6u7uXfzQlYul0OpPJxL07\nWH5X5PP5JEmGYXdQLpdbWlqWde0qE3ajJ68dwqzZi3rCaoUQQqXS9WJ3X2Hdjftv0NTUdOqp\np/ZfLBaLS5YsqemQstlsNew6OztLpVJN5zViNTU15XK5Wr/UI1kul0un0z09PcVisd5jqY8k\nSQqFQrFYbNj9WXNzcyaTKZfLjbwiZDKZnp6eWv86PWKl0+lq2HV2djZs3RYKhSRJGnktyGaz\n1bAbhhdhOWG3ynx4omX8IU1J6omZj1cvdsye2VGqbHTwevUdFQDAyLHKHLFLcmtN33/yF2++\n4GtXHDq5vXTXzJub19rutM191wkAwOtWmbALIWx2zAVn5Wfc8Lvb7+/ITNliz6+fdmxTklrx\n3QAAGsOqFHYhpLY//FPbH17vUQAAjEirzDl2AAAsn7ADAIiEsAMAiISwAwCIhLADAIiEsAMA\niISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLAD\nAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISwAwCIhLADAIiEsAMAiISw\nAwCIRKpSqdR7DDVRLBaLxWJNZzFv3rx77rknhLD77ru3trbWdF4jVjabTafTXV1d9R5I3fzy\nl79csmTJlClTNtpoo3qPpT5SqVRzc3OxWIx1Y7JCTz311JNPPtnS0rLHHnvUeyx1UygUSqVS\nb29vvQdSH0uWLPnlL38ZQthpp53a29vrPZz6yGQy2Wy2s7Oz3gOpm9/85jfz589ff/313/3u\nd9d6XuPGjVvWVZlaz7tempubm5ubazqLOXPmfOc73wkh7Lnnnst5iRtBw3ZtCOG666576aWX\nTjjhhO23377eY6mnpqameg+hbm644YbLL7984sSJhx56aL3HQn0Ui8Xq7mCbbbZp8N1BS0tL\nvYdQN7fccsvjjz9+0EEH7bLLLnUchj/FAgBEQtgBAERC2AEARCLaD08Mg1KptGTJkhBCa2tr\nkkjkBrV48eJyuZzP5/P5fL3HQn10d3d3d3cnSdLIJ5s2uHK5vHjx4hBCS0tLOp2u93Coj2Kx\n2NfXl8vlCoVCHYch7AAAIuE4EwBAJIQdAEAkov0eu5Whct+1l17/24df6MhsvOlWJ5x23Nr5\nfz1zYlm3Gcx9GflW/D5W+ubfevV3b7v/ybkLSxMnbbj/x0/cddM1QgjP/L9TP3PT8wNvufUl\nP/j8pNHDN3ZWmhUvBst+u20K4rCC9/HF2886+bJZ/3q362++5fkrbQpiUvncIQdmz/rul7Zc\n6rcVjogkEHbL9NiV07984+Pv3nGv/dtLd99+++nT5l/57bObktRgbjOY+zLyDeZ9nHnOadfM\nWrzNXvvuNCb9xO9uv+ScE5d84+r91mt75U8L0rm1dt3pze8fn9SWG/ZnwEowmMVgWW+3TUEc\nVvg+tqyz9R57TBxwj8r9v7qrd9TW2dQylw1WOZXSkgd/OuPxYu/my7jBSEmCCktT6vnbwR/a\n/+hzf1m9+Nqjl+23334XPPLqYG4zmPsy8g3mfexe+Lv99tvv5O889sZd/n78AR865PgfVSqV\nS4/46CHH3zLMY2alG+TqvNS326YgDm/hfXzxVxfu/6GDfz13ScWmIBZPfuszH9l///3222+/\n/fb7wv/+/V9vMHKSwDl2S7dkzrXFUmWTw95Tvdi+yRHN6dSsmbMHc5vB3JeRbzDvY+crvw4h\nbPyBd1YvJtlxm7Vk+zr/GkL4w+Le1klj//D7X996268ee/qlYR06K88gV+elvt02BXEY6vvY\nW3z889/63aZHnr/zms3BpiAWa+74kZOnTTvp2B2XdYORkwT+FLt0i56aE0JYf+zr30yWSres\nm888P/fJEKau8DaDuS8j32Dex9EbfP6WW968y7zHf37Xgu5xO+xc6pr9am8p3PeVL933+lXr\nbX/gV876eD7lb3CrmMEsBst6u20K4jDU9/H+S77Rkdv0nP3XC8teNmwKVjnt//be3ULoenXR\nt6/47VJvMHKSwBG7petZ1BtCGJN5c91rzyTlvvmDuc1g7svIN6T3sVLquPv6y04857tJ2yZn\nn7pFz+KHkySZuP1hl13xPz/+n6vO/Ph2f/n99ef+2KGaVc+gNgXLeLttCuIwpPexe8Hvv37/\n37f79LTmJBWWvWwMz8gZTiMnCRyxW7psWyaEsLDvzW9vnt9XTnKjBnObwdyXkW/w7+Mrj91x\n0de+99T87sk7HPDJUw6fWEiHwgE33XRA/w12/Ohnf3njR5/86Y3hoM8Mw8hZiQazGDSNW/rb\nnf2YTUEMhrRJ/9O3rwr5d5zyntWrF5e1bNgUxGfkJIEjdks3evLaIYTZi3qqFyuVrhe7+wqr\nbzyY2wzmvox8g3wfX7z7O8d//tKXRm1xztev+sqZR0xsTocQii899+yz/zfgVqn1C5lKaeHw\njJyVaDCLwbLebpuCOAz+fSz3vfath/4+fqcTC2/8pdWmoHGMnCQQdkvXMv6QpiT1xMzHqxc7\nZs/sKFU2Oni9wdxmMPdl5BvM+1jqmn3GJbetucMx37v4s1uv394/ff6Tl59++qevf77j9Zv1\nzLlnUXfbpN2HbfCsLINZDJb1dtsUxGHw7+Nrj3x7fl95z4PX759iU9A4Rk4SpM8999zaPfqq\nK5Vu3bD4yM/uvGlOMTX/uUevnHFT12rbXHDc+7Op1JVnnfbfP7l51332zWXalnqbZU3POlt2\nlTKYZWDevRfe+PtXxq+32vOPPPzQGx55fPH79vvg07+8/c6f37ews+PZP//hum9f+nyx7Zyv\nnbxmzpfTrmIGsxi0v2Ozpb7dazWNtimIwGCWgep7+oeLr3hgQWH6UQem33iLWycufdmwKVhF\n9RVn/fjmR9ba+YO7jG+uThmBSeAcu2Xa7JgLzsrPuOF3t9/fkZmyxZ5fP+3Y6tcJzn95zksL\nuyqV5d1mWdNZtaxwGZj30GshhGfuvvOZAfcqtJeOP2qX6Zd86ervXn3vHTd2VFo22PC9X/rP\n49/dmq3T8+BtWeFikGTHLevttimIw2B2B6HSd+1fFjWN+1huwA57OcsGcRiBSZCqVCorvhUA\nACOec+wAACIh7AAAIiHsAAAiIewAACIh7AAAIiHsAAAiIewAACIh7ABWsvtO2rhtnVPrPQqg\nEQk7IH4Hrt6Sya89yBuXuv+6zaj80be/sJzbLHn52lxuzC1/K/7T9L/c+d8f22+nD1/1zJK5\n3508dZdPf/3m3mH/DvghPVkgMsIO4B/ceuqeT4w6+PI9Jy7nNi3jD/3eHqOP3etL5QETn7zi\nsHe9/8TbXyjsNHl0rnWrndcvXvyZD21y0JW1HjBAP2EH8KbuBXceeMVT+1715cyK/pfjAd/7\nj1cf+cqZf3y1erHcM2ePU2a2rHPU8w/fftq2a2Rbp37nRw/88IPvfOb6Y7750uKajxsghCDs\nAAZ64JxPlbPjZ+w0foW3bBl/1JFrtnz/mNcPyBVfueal7tLED57QMuDfe+/9X/tvtNFGDz02\nv1bDHaB7/rP3/OqeeX3+/Tc0NGEH1M2ZE0c1te/26h9nHr7/ruuOa5kweavDP31pZ7ly1zc/\nteOWk0cVWidtst2F188aeJd5f77p8P1322D8qLbVJ267xxE/efDl/qsWzPr5kR/aY/I7Vm9e\nbeJ2u37gKzc8utSZlntf+X/nnbzlxuuPLhTWfMeGex11zkN/73r9ukr3cVc+s8bUb6yRTUII\nc351diqVmrD71/rvO/v6Y1Op1JRjZ1Yvnv7pKa89+tl7FvaEEJLc+BDCa394ZOC8xm58yaxZ\ns76/18Tlj/CMdUenUqlLXhx4YK+045hCki480NETQuh+7Y9nfuxDm284sdDSvtHmu/z7t2/p\n77fjxrc2r/aB2bf8xzvWmrzTbjv9ekFX+EcP3/CtD22/xYTVR+da2tebvPmJ/37ZvL5yCKF3\n8R9TqdS7Dr3njzO/cuCe26/R2rbhFjse/4UfdCtDWKVVAOrkjAltSbp1QvsmF1x5w2/vuvXU\nXdcOIay740bjtjj4mpvv+sVP/ntqWy7JjLp1Xmf19s/ddNaoTJJtnrTfwcccfej+q2XTSWbU\nlbPmVyqVF24/Z0wmSbJjd9//0OMP/8imqxVCCDtP+0n1jgeMa07nxlcqlUql/B/vnxBCmLDJ\ntocde/SuUzcKIbSu86FXe0uVSmXRCxeFEN73P8/0j/Cqg9cLIZxw03OVSqVn0R82aMo0rb7X\n3J5S9doF/3d2CGHvnz9fqVQqldJH125JpVIf+PTXLv3IO1vX/uQ/PdnljPDpq3YOIWzyyXv7\nb7zg2S+FEFbf/OJKpTJ/1pUbtGSTdMtuHzzkE8ccts2kthDC5AO+Uq5UKpXKsWu1JJlR47Lp\n8e95/zEnn/7Ekt4BT7by3I2npVKp/JgNDznyuOMO++Bm640JIUz68H9XKpWejodDCPn2TZJM\n+ylfvOi6a6/43Am7hRAm7np6d3llvcPAcBN2QN2cMaEthHDqPS9XLy6Ze3UIIZ0b/+clvdUp\nj120dQhhj9v+WqlUSt1zNmjKZJrW/82cJdVrX3v0khDCmA3OLfct2LI1l6Rbrn30tepVPYuf\n+sgazSGE777QURkQdsVXrgshjF7v3P50uXybtUIIn3h6Xv/sLnqxo3+EfV2ztx+VzzZv/Oji\nngt3XSeV5C59fN6b13a/kKRSa+/4ej52PH/r3lu+fnAuSbfuetBJV//svmKpUqlUlj/CnsUP\n55NUoX3P/lHdcfD6IYRjfv9ypVL62PjWJDPqqj+9Wr2q3LfoP3dZO4Rw/N1zKpXKsWu1hBC2\nOHNm3xv3HRh2p67TlkolP/9b8Y37dmw3Kp+k2ypvhF0I4aQ7Xuh/Rr88899CCAfc/NwQ30lg\npBB2QN1Uw67/AFjvksdDCKPe8YX+G7z46z1DCDtd92ylUpl738EhhCkn3jPwEb4+/awzP3v+\n/KdPDyFM2O3agVc9d8u+IYRNP/1AZWDY/f3HIYSm1T74uydfqd6s3Nvd2dlZHcIt240PIdy7\nqHvg4/ztvi+lUqnVtpoaQtj6s3f901N4ZyHTstaxA6fMfvDW4zdfLUm3ZlKpEMJqWxz0bGff\nCkf4lSljQwjfnrO4UqmUS8VNWrLZlk06+spL5n4vhLDm1B8MvGPHi5eEECbscmvljbC7c35X\n/7UDw+7pPz36yKNP9F9V7ut4f3uh+rea14/Yjd6hb8Aj93T8IYQwdsMZFWDV5Bw7oM7WzPZv\niFIhhHRurf6rUqk3P4gw57anQwiTjnjXwPt++vyvfPXL0+c98nAI4R0f3XzgVePes18IYe4d\nzw+c2DTuwGtO2aNn3i07TFljvc13POKkM6/+2b2dmVx1CLNf6QohbNiUGXiXNbb9wsyj3vXa\n//6hZfyhd523yz8NfqOmTF/xzwOnvPM9+xy97RrNax41/69/POejG772x5n7nPybFY7wkAt3\nCCFc9rUnQgjznpj++JLe9Q+9rDWdWjLnpyGEv/3h46kB2iZ8KoSw5IXZb7xKybajckt5ZUN4\n17v/bc3Kq9dc8a3pp31iv923XbOt/Y75/3ASXtO4A9MDLmZbt1q/KbNk7nVLfTRg5Mus+CYA\nI0C5pxxCSGVX9DUkb0ilMiGESrn3n6Yf+q07dj3+V9fe9LNf3nX3LVd84wffuahlnW2vffDO\n/dZuef2O/zLnPz+1KITQveChWcXeqW3/kFCpVKhUKiGE1/5w442PvLblIUdu2ZqtXtU6YbMv\nXfP7i3+y5ku3XRP2XsEI19l9xpjMz5696vxw0U2/OeOGEML086aGEEKShBDWef855370nf90\n91zrlm88UG7gR3EH+uGndj3im3cnubHv3WPP9+1y8FGnX3D90fte98qA71Wu9PzTXbrKlUr5\nnz+BAawqHLEDVg1r7johhPDCzH84Avf9z5xy3HHHZTbdLITwwg2PDbzqtYdvCSGssdOkgROX\nvPjw3Xff/fL47U7796/fes/Dr3XMueGSA5a8dP/xH7olhDBp9XwI4enOvoF3eeJ7B/3HvXM3\n/8ThfZ3PfnD/b/zTqJ4u9mVbNgkhvPzbC44//vivzl448NokM7Y5SaVS2fbNNl/+CNP5dS/a\ncvXO126+8sUXT7v75bYJp3x8zeYQQsta+4UQyh0bHzfAkYdsM2HChPW32GD5r1jxb1d9fMav\nx2x0xqsdr9zz0/8575xPH7DPzvnUPyTgkr99v6P05udgi3Oveam7VFjtA8t/ZGDEEnbAqmGt\n7b+6ejb95GXHPPTq68eTii//9MSLv33dz7omvuvczVpzL/3q2J/MWlC9qq/47GlH3xlCOPns\nTQY+yPwnv7jLLrvsf8SPqxfT+TV23u3fQgiVvkoI4Z0HTAwh3LfozYNYna/ctsspN42adMS9\nl/3g+wevN+fuzx5z/ez+a8s9Lz/XXRq1/r4hhAn77RFCuO9z1w/8XxQv33P233tLE/c/fMx6\n/77CEe77jb1CCNOP3OeF7r7tvvrp6sTmNY/86BrNcx849uo/vfbGo5YuO+QDe++99486upf/\ninUvujeEMHrD3ca88cfuVx649Nq/F0MI/YPsXfLnA7/52+rPldLCrx56Zghhmy8evvxHBkau\nep/kBzSu6ocn+i/2LnkihNC+waX9U166e6/wxocnKpXKY989Opuksq3v+sjHTjj5uI+t15pN\npZu+9cirlUrl+Z+dPSqTpHNr7PvRI0866qDNVm8KIbzvlJnVO/Z/nqCva/YuqzWFEKbufcgn\nP3XqR/d+3zqFTJJpv+zR1yqVyqIXvjpwdpVy10lT2pN0yw+fW1SpVHo7n9myNZdpWv/BNz5d\nseAv00MIe/20+hnS0vQd1wohTPnAsae9b3yhfffzzjhmbDYpjN36/zr7VjjCSqVS6n1tYj4T\nQkiyY//S+eZHGl595FsT8pkk3brXAR879ZOf2HvqhBDC5MO/Wf3IybFrtaSSwsBXdcCTfW7r\nUfkk3faBg4/9wrnTjztkt5ZMdsNRuRDCXkec+OuX7g8hpPPrpFKpzXbZ7/jjDt92o9EhhDW2\nOaHj/7dzdyFNhXEcx59Tc8w1gy21NWRlvmRGJpHVMFyFZsOKCLxqZbK8sIhuAgPFKOgFgroI\nsou0N7qoZL1R6Y0QDiUICpGigoJma4MyxzYFX3a62FgvgitwOg/fz91z/g/n/K8OvwPP/4zz\nvxNgriLYAZg1/xvsZFkecF3fXVFqTl+wMNNsqbA7X36NlQb7H+3dsSUvy6DRm0qs28+1v46V\nfh8UDbq7G/ZvyzMbNSp1ZlaOdU/9/Rfe6L6JkZxUlamsPbLqPb1ZCGFp7o7dx91xRAhhsp6N\nLPvOl0iS1DUUnUgNjwdunDiwJtcYmfnQGXOr7I1v/aP/0mGEs9IsJg3PyrIc8vQcrratyjaq\ntfqC4tKmlscj0UniqYKdLMv+908dtvUmgzZtcXaZrfqWa8D/wVmza1OxZeszd68QwpDf1n/v\nVLmlyJCqXb56Y03D1eAEqQ6YwyRZ5i/jAJJVeMz7+eOgbmlhumZmHvi8vrD8WsgX/GRQxT+p\ncnCJ7kFG07e+439dd9WttD2xBjxXEtPj9BgLvlKnrTXkt31/VzvbvQCYNpyxA5DE5qUYl62Y\nsVQnhNhw5qI06j7a44u7c9h3u9Ubsrc5JpfmqyQhpSSgOwCIg2AHAL9o9JV3a/Mf7msOx9vp\nrGtcVHTswrqMySVLy5vAl0uJaA8ApkawA4A/7LzcWfDj5qEuzxR7hr13HB1DrZ0neYcCSCqc\nsQMAAFAIvjYBAAAUgmAHAACgEAQ7AAAAhSDYAQAAKATBDgAAQCEIdgAAAApBsAMAAFAIgh0A\nAIBCEOwAAAAU4idrLtLsfi+6tAAAAABJRU5ErkJggg==",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "qplot(mcols(x)$overlap, xlim=c(0, 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
