{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The basic idea is to find out which categories of genomic regions (exons, promoters, enhancers, etc) contribute the most to the Nea. ancestry observed in present day Europeans. For example: have coding regions or enhancers been the main drivers of selection against Nea. introgression?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Links:\n",
    "\n",
    "* info about different coordinate types: https://groups.google.com/forum/#!msg/biomart-users/OtQbAx3y9CA/wrF19ID1AgAJ\n",
    "* https://www.biostars.org/p/2005/\n",
    "* http://www.bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/biomaRt.html#attribute-pages\n",
    "* http://www.ensembl.org/info/data/biomart/biomart_r_package.html\n",
    "* http://www.ensembl.info/blog/2015/06/01/biomart-or-how-to-access-the-ensembl-data-from-r/\n",
    "* for checking with manually downloaded data: http://www.ensembl.org/info/data/biomart/how_to_use_biomart.html"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "source(\"http://bioconductor.org/biocLite.R\")\n",
    "biocLite(\"biomaRt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Following [this](http://www.ensembl.org/info/data/biomart/biomart_r_package.html) tutorial, I want to extract coordinates of exonic and regulatory regions from the Ensembl database and then calculate the density of such regions in a defined window around each SNP.\n",
    "\n",
    "These densities will they be used as predictors in a linear model, predicting the Nea. ancestry at each site.\n",
    "\n",
    "Alternatively, I could just test if the distribution of densities for different regions differ based on frequency of Nea. alleles at each site."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading tidyverse: ggplot2\n",
      "Loading tidyverse: tibble\n",
      "Loading tidyverse: tidyr\n",
      "Loading tidyverse: readr\n",
      "Loading tidyverse: purrr\n",
      "Loading tidyverse: dplyr\n",
      "Conflicts with tidy packages ---------------------------------------------------\n",
      "filter(): dplyr, stats\n",
      "lag():    dplyr, stats\n",
      "\n",
      "Attaching package: ‘magrittr’\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    set_names\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    extract\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(tidyverse)\n",
    "library(stringr)\n",
    "library(magrittr)\n",
    "library(parallel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: GenomicRanges\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, cbind, colnames, do.call,\n",
      "    duplicated, eval, evalq, Filter, Find, get, grep, grepl, intersect,\n",
      "    is.unsorted, lapply, lengths, Map, mapply, match, mget, order,\n",
      "    paste, pmax, pmax.int, pmin, pmin.int, Position, rank, rbind,\n",
      "    Reduce, rownames, sapply, setdiff, sort, table, tapply, union,\n",
      "    unique, unsplit\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "Loading required package: stats4\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    expand\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    colMeans, colSums, expand.grid, rowMeans, rowSums\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "The following objects are masked from ‘package:purrr’:\n",
      "\n",
      "    reduce, simplify\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: BSgenome\n",
      "Loading required package: Biostrings\n",
      "Loading required package: XVector\n",
      "\n",
      "Attaching package: ‘XVector’\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    compact\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(biomaRt)\n",
    "library(rtracklayer)\n",
    "library(BSgenome.Hsapiens.UCSC.hg19)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get coordinates of admixture array sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "source(\"../R/utils.R\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Missing column names filled in: 'X12' [12], 'X13' [13], 'X14' [14], 'X15' [15], 'X16' [16], 'X17' [17], 'X18' [18], 'X19' [19], 'X20' [20], 'X21' [21], 'X22' [22], 'X23' [23], 'X24' [24]”Parsed with column specification:\n",
      "cols(\n",
      "  .default = col_character(),\n",
      "  Latitude = col_double(),\n",
      "  Longitude = col_double(),\n",
      "  Coverage = col_integer(),\n",
      "  HetRateAuto = col_double()\n",
      ")\n",
      "See spec(...) for full column specifications.\n",
      "Parsed with column specification:\n",
      "cols(\n",
      "  .default = col_integer(),\n",
      "  ref = col_character(),\n",
      "  alt = col_character()\n",
      ")\n",
      "See spec(...) for full column specifications.\n",
      "Parsed with column specification:\n",
      "cols(\n",
      "  .default = col_integer(),\n",
      "  ref = col_character(),\n",
      "  alt = col_character()\n",
      ")\n",
      "See spec(...) for full column specifications.\n",
      "Parsed with column specification:\n",
      "cols(\n",
      "  chrom = col_integer(),\n",
      "  pos = col_integer(),\n",
      "  ref = col_character(),\n",
      "  alt = col_character(),\n",
      "  Altai = col_integer(),\n",
      "  Vindija = col_integer(),\n",
      "  Denisovan = col_integer()\n",
      ")\n",
      "Joining, by = c(\"chrom\", \"pos\", \"ref\", \"alt\")\n",
      "Joining, by = c(\"chrom\", \"pos\", \"ref\", \"alt\")\n"
     ]
    }
   ],
   "source": [
    "european_ids <- get_european_ids(\"../raw_data/10_24_2014_SGDP_metainformation_update.txt\")\n",
    "\n",
    "europeans <-\n",
    "    load_dataset(\"../clean_data/ice_age.tsv\", \"../clean_data/sgdp.tsv\", \"../clean_data/archaics.tsv\") %>%\n",
    "    dplyr::select(c(\"chrom\", \"pos\", european_ids)) %>% \n",
    "    dplyr::mutate(chrom=paste0(\"chr\", chrom), start=pos, end=pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sites <- makeGRangesFromDataFrame(europeans)\n",
    "seqinfo(sites) <- seqinfo(BSgenome.Hsapiens.UCSC.hg19)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "484016"
      ],
      "text/latex": [
       "484016"
      ],
      "text/markdown": [
       "484016"
      ],
      "text/plain": [
       "[1] 484016"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(sites)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 6 ranges and 0 metadata columns:\n",
       "      seqnames           ranges strand\n",
       "         <Rle>        <IRanges>  <Rle>\n",
       "  [1]     chr1 [847983, 847983]      *\n",
       "  [2]     chr1 [853089, 853089]      *\n",
       "  [3]     chr1 [853596, 853596]      *\n",
       "  [4]     chr1 [854793, 854793]      *\n",
       "  [5]     chr1 [867552, 867552]      *\n",
       "  [6]     chr1 [871401, 871401]      *\n",
       "  -------\n",
       "  seqinfo: 93 sequences (1 circular) from hg19 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(sites)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculate the overlap of each window with each annotation feature"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Those black magic GRanges overlap tricks are from [here](https://support.bioconductor.org/p/56880/)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 382,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "WINDOW_SIZE <- 10"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Testing sites, testing windows and testing exons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Developing the code on a simple test example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 383,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 4 ranges and 0 metadata columns:\n",
       "      seqnames     ranges strand\n",
       "         <Rle>  <IRanges>  <Rle>\n",
       "  [1]     chr1 [ 10,  10]      *\n",
       "  [2]     chr1 [ 16,  16]      *\n",
       "  [3]     chr1 [ 22,  22]      *\n",
       "  [4]     chr1 [100, 100]      *\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "GRanges object with 4 ranges and 0 metadata columns:\n",
       "      seqnames    ranges strand\n",
       "         <Rle> <IRanges>  <Rle>\n",
       "  [1]     chr1 [ 5,  14]      *\n",
       "  [2]     chr1 [11,  20]      *\n",
       "  [3]     chr1 [17,  26]      *\n",
       "  [4]     chr1 [95, 104]      *\n",
       "  -------\n",
       "  seqinfo: 93 sequences (1 circular) from hg19 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "GRanges object with 4 ranges and 0 metadata columns:\n",
       "      seqnames     ranges strand\n",
       "         <Rle>  <IRanges>  <Rle>\n",
       "  [1]     chr1 [  7,   9]      *\n",
       "  [2]     chr1 [ 12,  13]      *\n",
       "  [3]     chr1 [ 16,  21]      *\n",
       "  [4]     chr1 [200, 220]      *\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# sites\n",
    "(s <- GRanges(seqnames=\"chr1\", ranges=IRanges(c(10, 16, 22, 100), width=1)))\n",
    "seqinfo(s) <- seqinfo(BSgenome.Hsapiens.UCSC.hg19)\n",
    "\n",
    "# windows around sites\n",
    "(w <- flank(s, as.integer(WINDOW_SIZE / 2), both=TRUE))\n",
    "seqinfo(w) <- seqinfo(BSgenome.Hsapiens.UCSC.hg19)\n",
    "\n",
    "# regions of interest (like exons)\n",
    "(r <- GRanges(seqnames=\"chr1\", ranges=IRanges(start=c(7, 12, 16, 200), end=c(9, 13, 21, 220))))\n",
    "seqinfo(r) <- seqinfo(BSgenome.Hsapiens.UCSC.hg19)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "WINDOW_SIZE <- 10"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate the # of overlapping bases for each window:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 384,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Hits object with 5 hits and 0 metadata columns:\n",
       "      queryHits subjectHits\n",
       "      <integer>   <integer>\n",
       "  [1]         1           1\n",
       "  [2]         1           2\n",
       "  [3]         2           2\n",
       "  [4]         2           3\n",
       "  [5]         3           3\n",
       "  -------\n",
       "  queryLength: 4 / subjectLength: 4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# get the hit table between windows and regions they overlap with\n",
    "(hits <- findOverlaps(w, r))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 385,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 5 ranges and 1 metadata column:\n",
       "      seqnames    ranges strand |       hit\n",
       "         <Rle> <IRanges>  <Rle> | <logical>\n",
       "  [1]     chr1  [ 7,  9]      * |      TRUE\n",
       "  [2]     chr1  [12, 13]      * |      TRUE\n",
       "  [3]     chr1  [12, 13]      * |      TRUE\n",
       "  [4]     chr1  [16, 20]      * |      TRUE\n",
       "  [5]     chr1  [17, 21]      * |      TRUE\n",
       "  -------\n",
       "  seqinfo: 93 sequences (1 circular) from hg19 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# get all overlapping regions (duplicates for regions\n",
    "# overlapping several windows/sites)\n",
    "(w_over <- pintersect(w[queryHits(hits)], r[subjectHits(hits)]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 386,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl class=dl-horizontal>\n",
       "\t<dt>1</dt>\n",
       "\t\t<dd>5</dd>\n",
       "\t<dt>2</dt>\n",
       "\t\t<dd>7</dd>\n",
       "\t<dt>3</dt>\n",
       "\t\t<dd>5</dd>\n",
       "</dl>\n"
      ],
      "text/plain": [
       "1 2 3 \n",
       "5 7 5 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# for each window count the number of overlapping bases in regions\n",
    "# as an integer vector\n",
    "(w_counts <- tapply(w_over, queryHits(hits), FUN=function(x) sum(width(x))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 387,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 4 ranges and 1 metadata column:\n",
       "      seqnames     ranges strand |   overlap\n",
       "         <Rle>  <IRanges>  <Rle> | <numeric>\n",
       "  [1]     chr1 [ 10,  10]      * |       0.5\n",
       "  [2]     chr1 [ 16,  16]      * |       0.7\n",
       "  [3]     chr1 [ 22,  22]      * |       0.5\n",
       "  [4]     chr1 [100, 100]      * |         0\n",
       "  -------\n",
       "  seqinfo: 93 sequences (1 circular) from hg19 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# add the column with % overlap to the original sites object\n",
    "s$overlap <- 0\n",
    "s$overlap[as.numeric(names(w_counts))] <- unname(w_counts) / WINDOW_SIZE\n",
    "s"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The code seems to work, let's construct a function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 388,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "calc_window_overlaps <- function(sites, regions, window_size) {\n",
    "    # construct windows around each site\n",
    "    windows <- flank(sites, as.integer(window_size / 2), both=TRUE)\n",
    "    seqinfo(windows) <- seqinfo(sites)\n",
    "\n",
    "    # get the hit table between windows and regions they overlap with\n",
    "    hits <- findOverlaps(windows, regions)\n",
    "\n",
    "    # get the all overlapping regions (duplicates for regions\n",
    "    # overlapping several windows/sites)\n",
    "    w_over <- pintersect(windows[queryHits(hits)], regions[subjectHits(hits)])\n",
    "\n",
    "    # for each window count the number of overlapping bases in regions\n",
    "    # as an integer vector\n",
    "    w_counts <- tapply(w_over, queryHits(hits), FUN=function(x) { sum(width(x)) })\n",
    "\n",
    "    # add the column with % overlap to the original sites object\n",
    "    sites$overlap <- 0 # by default zero overlap for each site\n",
    "    sites$overlap[as.numeric(names(w_counts))] <- unname(w_counts) / window_size\n",
    "\n",
    "    sites\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Test of this function on the toy dataset from above:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 389,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 4 ranges and 1 metadata column:\n",
       "      seqnames     ranges strand |   overlap\n",
       "         <Rle>  <IRanges>  <Rle> | <numeric>\n",
       "  [1]     chr1 [ 10,  10]      * |       0.5\n",
       "  [2]     chr1 [ 16,  16]      * |       0.7\n",
       "  [3]     chr1 [ 22,  22]      * |       0.5\n",
       "  [4]     chr1 [100, 100]      * |         0\n",
       "  -------\n",
       "  seqinfo: 93 sequences (1 circular) from hg19 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "calc_window_overlaps(granges(s), r, WINDOW_SIZE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting exon coordinates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Show all the available biomarts for hg19:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>biomart</th><th scope=col>version</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>ENSEMBL_MART_ENSEMBL </td><td>Ensembl Genes 89     </td></tr>\n",
       "\t<tr><td>ENSEMBL_MART_SNP     </td><td>Ensembl Variation 89 </td></tr>\n",
       "\t<tr><td>ENSEMBL_MART_FUNCGEN </td><td>Ensembl Regulation 89</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ll}\n",
       " biomart & version\\\\\n",
       "\\hline\n",
       "\t ENSEMBL\\_MART\\_ENSEMBL & Ensembl Genes 89        \\\\\n",
       "\t ENSEMBL\\_MART\\_SNP     & Ensembl Variation 89    \\\\\n",
       "\t ENSEMBL\\_MART\\_FUNCGEN & Ensembl Regulation 89   \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "biomart | version | \n",
       "|---|---|---|\n",
       "| ENSEMBL_MART_ENSEMBL  | Ensembl Genes 89      | \n",
       "| ENSEMBL_MART_SNP      | Ensembl Variation 89  | \n",
       "| ENSEMBL_MART_FUNCGEN  | Ensembl Regulation 89 | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  biomart              version              \n",
       "1 ENSEMBL_MART_ENSEMBL Ensembl Genes 89     \n",
       "2 ENSEMBL_MART_SNP     Ensembl Variation 89 \n",
       "3 ENSEMBL_MART_FUNCGEN Ensembl Regulation 89"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "listMarts(host=\"grch37.ensembl.org\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Connect to the human gene Ensembl dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>dataset</th><th scope=col>description</th><th scope=col>version</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>hsapiens_gene_ensembl   </td><td>Human genes (GRCh37.p13)</td><td>GRCh37.p13              </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lll}\n",
       " dataset & description & version\\\\\n",
       "\\hline\n",
       "\t hsapiens\\_gene\\_ensembl & Human genes (GRCh37.p13)  & GRCh37.p13               \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "dataset | description | version | \n",
       "|---|\n",
       "| hsapiens_gene_ensembl    | Human genes (GRCh37.p13) | GRCh37.p13               | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  dataset               description              version   \n",
       "1 hsapiens_gene_ensembl Human genes (GRCh37.p13) GRCh37.p13"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ensembl_mart_genes <- useMart(\"ENSEMBL_MART_ENSEMBL\", host=\"grch37.ensembl.org\")\n",
    "listDatasets(ensembl_mart_genes) %>% filter(str_detect(dataset, \"sapiens\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genes <- useDataset(dataset=\"hsapiens_gene_ensembl\", mart=ensembl_mart_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get coordinates of exons from the gene annotation database:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "exons_df <-\n",
    "    getBM((listAttributes(genes) %>% filter(page==\"structure\"))$name,    # get all attributes\n",
    "          filters=c(\"chromosome_name\", \"biotype\"),                       # filter for chromosome and biotype\n",
    "          values=list(1:22, \"protein_coding\"),                           # chromosomes 1:22 and \"protein_coding\"\n",
    "          mart=genes)\n",
    "\n",
    "exons_gr <- \n",
    "    dplyr::select(exons_df, chromosome_name, exon_chrom_start, exon_chrom_end) %>%\n",
    "    setNames(c(\"chrom\", \"start\", \"end\")) %>%                             # rename the columns to BED names\n",
    "    mutate(chrom=paste0(\"chr\", chrom)) %>%                               # convert chromosome IDs to USCS format\n",
    "    makeGRangesFromDataFrame %>%                                         # convert to GRanges object\n",
    "    reduce                                                               # merge overlapping exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqinfo(exons_gr) <- seqinfo(BSgenome.Hsapiens.UCSC.hg19)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many exons did I get? What is the total length of the exonic sequence (in Mb) that we extracted?\n",
    "\n",
    "Bioinformatic databases are a mess - are the numbers I got reasonable?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "222587"
      ],
      "text/latex": [
       "222587"
      ],
      "text/markdown": [
       "222587"
      ],
      "text/plain": [
       "[1] 222587"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "89.81306"
      ],
      "text/latex": [
       "89.81306"
      ],
      "text/markdown": [
       "89.81306"
      ],
      "text/plain": [
       "[1] 89.81306"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(exons_gr)\n",
    "sum(width(exons_gr)) / 1e6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "WINDOW_SIZE <- 100000\n",
    "\n",
    "exon_densities <-\n",
    "    mclapply(paste0(\"chr\", 1:22),\n",
    "             function(chr) {\n",
    "                 calc_window_overlaps(sites[seqnames(sites) == chr], exons, WINDOW_SIZE)\n",
    "             }, mc.cores=22) %>%\n",
    "    do.call(c, .)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 484016 ranges and 1 metadata column:\n",
       "           seqnames               ranges strand |   overlap\n",
       "              <Rle>            <IRanges>  <Rle> | <numeric>\n",
       "       [1]     chr1     [847983, 847983]      * |   0.09822\n",
       "       [2]     chr1     [853089, 853089]      * |    0.1211\n",
       "       [3]     chr1     [853596, 853596]      * |    0.1211\n",
       "       [4]     chr1     [854793, 854793]      * |    0.1211\n",
       "       [5]     chr1     [867552, 867552]      * |   0.17652\n",
       "       ...      ...                  ...    ... .       ...\n",
       "  [484012]    chr22 [51142979, 51142979]      * |   0.09574\n",
       "  [484013]    chr22 [51145136, 51145136]      * |   0.09574\n",
       "  [484014]    chr22 [51154908, 51154908]      * |   0.09574\n",
       "  [484015]    chr22 [51174091, 51174091]      * |   0.14187\n",
       "  [484016]    chr22 [51177248, 51177248]      * |   0.14187\n",
       "  -------\n",
       "  seqinfo: 93 sequences (1 circular) from hg19 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "exon_densities"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculate the overlap of each window with each annotation feature"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Those black magic GRanges overlap tricks are from [here](https://support.bioconductor.org/p/56880/)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Developing the code on a simple test example"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Testing sites, testing windows and testing exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "WINDOW_SIZE <- 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 4 ranges and 0 metadata columns:\n",
       "      seqnames     ranges strand\n",
       "         <Rle>  <IRanges>  <Rle>\n",
       "  [1]     chr1 [ 10,  10]      *\n",
       "  [2]     chr1 [ 16,  16]      *\n",
       "  [3]     chr1 [ 22,  22]      *\n",
       "  [4]     chr1 [100, 100]      *\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "GRanges object with 4 ranges and 0 metadata columns:\n",
       "      seqnames    ranges strand\n",
       "         <Rle> <IRanges>  <Rle>\n",
       "  [1]     chr1 [ 5,  14]      *\n",
       "  [2]     chr1 [11,  20]      *\n",
       "  [3]     chr1 [17,  26]      *\n",
       "  [4]     chr1 [95, 104]      *\n",
       "  -------\n",
       "  seqinfo: 93 sequences (1 circular) from hg19 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "GRanges object with 4 ranges and 0 metadata columns:\n",
       "      seqnames     ranges strand\n",
       "         <Rle>  <IRanges>  <Rle>\n",
       "  [1]     chr1 [  7,   9]      *\n",
       "  [2]     chr1 [ 12,  13]      *\n",
       "  [3]     chr1 [ 16,  21]      *\n",
       "  [4]     chr1 [200, 220]      *\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# sites\n",
    "(s <- GRanges(seqnames=\"chr1\", ranges=IRanges(c(10, 16, 22, 100), width=1)))\n",
    "seqinfo(s) <- seqinfo(BSgenome.Hsapiens.UCSC.hg19)\n",
    "\n",
    "# windows around sites\n",
    "(w <- flank(s, as.integer(WINDOW_SIZE / 2), both=TRUE))\n",
    "seqinfo(w) <- seqinfo(BSgenome.Hsapiens.UCSC.hg19)\n",
    "\n",
    "# regions of interest (like exons)\n",
    "(r <- GRanges(seqnames=\"chr1\", ranges=IRanges(start=c(7, 12, 16, 200), end=c(9, 13, 21, 220))))\n",
    "seqinfo(r) <- seqinfo(BSgenome.Hsapiens.UCSC.hg19)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate the # of overlapping bases for each window:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Hits object with 5 hits and 0 metadata columns:\n",
       "      queryHits subjectHits\n",
       "      <integer>   <integer>\n",
       "  [1]         1           1\n",
       "  [2]         1           2\n",
       "  [3]         2           2\n",
       "  [4]         2           3\n",
       "  [5]         3           3\n",
       "  -------\n",
       "  queryLength: 4 / subjectLength: 4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# get the hit table between windows and regions they overlap with\n",
    "(hits <- findOverlaps(w, r))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 5 ranges and 1 metadata column:\n",
       "      seqnames    ranges strand |       hit\n",
       "         <Rle> <IRanges>  <Rle> | <logical>\n",
       "  [1]     chr1  [ 7,  9]      * |      TRUE\n",
       "  [2]     chr1  [12, 13]      * |      TRUE\n",
       "  [3]     chr1  [12, 13]      * |      TRUE\n",
       "  [4]     chr1  [16, 20]      * |      TRUE\n",
       "  [5]     chr1  [17, 21]      * |      TRUE\n",
       "  -------\n",
       "  seqinfo: 93 sequences (1 circular) from hg19 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# get all overlapping regions (duplicates for regions\n",
    "# overlapping several windows/sites)\n",
    "(w_over <- pintersect(w[queryHits(hits)], r[subjectHits(hits)]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<dl class=dl-horizontal>\n",
       "\t<dt>1</dt>\n",
       "\t\t<dd>5</dd>\n",
       "\t<dt>2</dt>\n",
       "\t\t<dd>7</dd>\n",
       "\t<dt>3</dt>\n",
       "\t\t<dd>5</dd>\n",
       "</dl>\n"
      ],
      "text/plain": [
       "1 2 3 \n",
       "5 7 5 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# for each window count the number of overlapping bases in regions\n",
    "# as an integer vector\n",
    "(w_counts <- tapply(w_over, queryHits(hits), FUN=function(x) sum(width(x))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 4 ranges and 1 metadata column:\n",
       "      seqnames     ranges strand |   overlap\n",
       "         <Rle>  <IRanges>  <Rle> | <numeric>\n",
       "  [1]     chr1 [ 10,  10]      * |       0.5\n",
       "  [2]     chr1 [ 16,  16]      * |       0.7\n",
       "  [3]     chr1 [ 22,  22]      * |       0.5\n",
       "  [4]     chr1 [100, 100]      * |         0\n",
       "  -------\n",
       "  seqinfo: 93 sequences (1 circular) from hg19 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# add the column with % overlap to the original sites object\n",
    "s$overlap <- 0\n",
    "s$overlap[as.numeric(names(w_counts))] <- unname(w_counts) / WINDOW_SIZE\n",
    "s"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The code seems to work, let's construct a function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "calc_window_overlaps <- function(sites, regions, window_size) {\n",
    "    # construct windows around each site\n",
    "    windows <- flank(sites, as.integer(window_size / 2), both=TRUE)\n",
    "    seqinfo(windows) <- seqinfo(sites)\n",
    "\n",
    "    # get the hit table between windows and regions they overlap with\n",
    "    hits <- findOverlaps(windows, regions)\n",
    "\n",
    "    # get the all overlapping regions (duplicates for regions\n",
    "    # overlapping several windows/sites)\n",
    "    w_over <- pintersect(windows[queryHits(hits)], regions[subjectHits(hits)])\n",
    "\n",
    "    # for each window count the number of overlapping bases in regions\n",
    "    # as an integer vector\n",
    "    w_counts <- tapply(w_over, queryHits(hits), FUN=function(x) { sum(width(x)) })\n",
    "\n",
    "    # add the column with % overlap to the original sites object\n",
    "    sites$overlap <- 0 # by default zero overlap for each site\n",
    "    sites$overlap[as.numeric(names(w_counts))] <- unname(w_counts) / window_size\n",
    "\n",
    "    sites\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Test of this function on the toy dataset from above:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 4 ranges and 1 metadata column:\n",
       "      seqnames     ranges strand |   overlap\n",
       "         <Rle>  <IRanges>  <Rle> | <numeric>\n",
       "  [1]     chr1 [ 10,  10]      * |       0.5\n",
       "  [2]     chr1 [ 16,  16]      * |       0.7\n",
       "  [3]     chr1 [ 22,  22]      * |       0.5\n",
       "  [4]     chr1 [100, 100]      * |         0\n",
       "  -------\n",
       "  seqinfo: 93 sequences (1 circular) from hg19 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "calc_window_overlaps(granges(s), r, WINDOW_SIZE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "WINDOW_SIZE <- 100000\n",
    "\n",
    "exon_densities <-\n",
    "    mclapply(paste0(\"chr\", 1:22),\n",
    "             function(chr) {\n",
    "                 calc_window_overlaps(sites[seqnames(sites) == chr], exons_gr, WINDOW_SIZE)\n",
    "             }, mc.cores=22) %>%\n",
    "    do.call(c, .)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 484016 ranges and 1 metadata column:\n",
       "           seqnames               ranges strand |   overlap\n",
       "              <Rle>            <IRanges>  <Rle> | <numeric>\n",
       "       [1]     chr1     [847983, 847983]      * |   0.09822\n",
       "       [2]     chr1     [853089, 853089]      * |    0.1211\n",
       "       [3]     chr1     [853596, 853596]      * |    0.1211\n",
       "       [4]     chr1     [854793, 854793]      * |    0.1211\n",
       "       [5]     chr1     [867552, 867552]      * |   0.17652\n",
       "       ...      ...                  ...    ... .       ...\n",
       "  [484012]    chr22 [51142979, 51142979]      * |   0.09574\n",
       "  [484013]    chr22 [51145136, 51145136]      * |   0.09574\n",
       "  [484014]    chr22 [51154908, 51154908]      * |   0.09574\n",
       "  [484015]    chr22 [51174091, 51174091]      * |   0.14187\n",
       "  [484016]    chr22 [51177248, 51177248]      * |   0.14187\n",
       "  -------\n",
       "  seqinfo: 93 sequences (1 circular) from hg19 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "exon_densities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Parsed with column specification:\n",
      "cols(\n",
      "  chrom = col_integer(),\n",
      "  start = col_integer(),\n",
      "  end = col_integer(),\n",
      "  overlap = col_double()\n",
      ")\n",
      "Joining, by = c(\"chrom\", \"start\", \"end\")\n"
     ]
    }
   ],
   "source": [
    "# old exon densities\n",
    "x <- read_tsv(\"../clean_data/annotations/exon_density__100000bp.bed\",\n",
    "              col_names=c(\"chrom\", \"start\", \"end\", \"overlap\")) %>% mutate(chrom=paste0(\"chr\", chrom))\n",
    "e <- mutate(europeans, start=pos-1, end=pos) %>% dplyr::select(chrom, start, end)\n",
    "\n",
    "exon_densities_bedops <- inner_join(x, e)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How does the new exon density metric compare to the old one?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeAAAAHgCAIAAADytinCAAAABmJLR0QA/wD/AP+gvaeTAAAg\nAElEQVR4nOzdeVxN6f8A8Oecu7ZpT6KFUJSyFM3IoFHWKSSRfV8zhLKPGbIvYwszQ2PsFRLZ\nkyzJTmkhQrKUSvtyb53z++P4Xv3q3tu5dZdTfd5/eHXPdj9d3c997nOe5/NgJEkiAAAAzIOr\nOgAAAADiQYIGAACGggQNAAAMBQkaAAAYChI0AAAwFCRoAABgKEjQAADAUJCgAQCAodiqDkAO\nSktLy8rKZDqFzWZraGgghAoKChg+VUdDQ0MoFAoEAlUHIg2Px+Pz+QRBFBYWqjqWWjRr1qy4\nuLiyslLVgUijpqbG5XIrKiqKi4tVHYs0OI5raWkx/02kqanJYrHKy8tlTRTKoaurK2lXY0jQ\nJEnK+n7DcRzHcYQQQRAEQSgmLvnAMKwOv6CSkSRJvZ4MjxMhhOM4819PDMNwHMcwjOFxIoRw\nHK+srGR4gqZeT9QQ/j6rgS4OAABgKEjQAADAUJCgAQCAoSBBAwAAQ0GCBgAAhoIEDQAADAUJ\nGgAAGAoSNAAAMBQkaAAAYChI0AAAwFCQoAEAgKEgQQMAAEM1hmJJAACgQmlpadHR0RkZGa1b\nt3ZzczM2NpbXlSFBAwBAHZEkuWHDhp07d1ZUVFBb1NTUVq9ePXnyZLlcH7o4AACgjg4cOLBt\n2zZRdkYIlZWVBQQEXLlyRS7XhwQNAAB1tHv3bgzDqm6haqPv2bNHLteHLg4AAKiL/Pz8Dx8+\n1NxOEER8fLxcngJa0AAAUBfUKi1iVWtW1/0p5HIVAABoarS0tFq3bl0zTeM43q1bN7k8BSRo\nAACoIz8/P4IgqraXqUVEf/31V7lcHxI0AADU0ahRo9asWcPn80VbdHR09u3b5+zsLJfrw01C\nAACoXUpKyq5du+Lj49lsdpcuXRYsWGBqaooQmjlz5rBhw+7cufP+/fs2bdr89NNP2tra8npS\nSNAAAFCLI0eOLFq0iCRJqkMjMTExJCTkn3/+GTBgAEKoefPmw4cPV8TzQhcHAABIk5GRERAQ\nQBAEQRAIIZIkSZIUCoVz584tKChQ6FNDggYAAGnOnTsnEAhIkqy6kSCI/Pz8qKgohT41JGgA\nAJDm/fv3knalp6cr9KkhQQMAgDQ6Ojp12CUXkKABAEAaFxcXsdtxHO/bt69CnxoSNAAASOPg\n4DBy5EhUZQI3NXvQ19fXzMxMoU8NCRoAAGqxY8eOwMBA0QBnIyOjXbt2LV++XNHPC+OgAQCg\nFmw2e/r06dOnT8/IyOBwOM2bN1fS8yrnaQAAoBFo1aqVMp8OEjQAACChUHj9+vXk5GRNTU1H\nR0d7e3tVR4QQJGgAALh8+fLkyZMFAoFoy9ChQ7dv366pqanCqBDcJAQANHGBgYFjx46tmp0R\nQuHh4fIqGVofkKABAE1XZmbmn3/+KXZXRETEu3fvlBxPNZCgAQBNl7+/v5S9z549U1okYkEf\nNACgKcrPzx8/fnxcXJyUY6oVSFI+aEEDAJqcEydOtGvXLjY2lqogKonKx3JAggYANC3FxcXz\n5s2rtXXcqlUrCwsLpUQkESRoAEDTMn/+/FqzM4Zh169fV048UkCCBgA0FZ8+fZo0aVJ4eLj0\nwzQ1Ne/cuaOrq6ucqKSAm4QAgCahtLR06NChb968kXIMl8sNDg52c3NTWlTSQYIGADQJR48e\nTUtLk35M7969mZOdEXRxAACaiDt37lB1nCXBcXz//v1Ki4cOlbSgybvH94Tdevy+kN3Rttv0\n+VNNeKyaB72KOXns0t3k1E86ph28pv7qYqP6/iAAQMNVWloqqrhfE5/PDw0N1dLSUmZItVJB\nCzoheNn641f4rR3ce3fKeHDJb97mUqL6HdWsu/v9th79wG05wN1Fqzhl58r5jwoFYq8GAAB0\nWFlZVVZWit21YMGC9PR0JycnJYdUK2UnaEKYFRiRZNBtXuDi2WOn+m5a5VbyKXZnQu7/O4is\n3LrjSjOL0UGrF08YP2Pdn79hlXn/nnir5FABAI3J2LFj2Wx2tUY0juPdunVbtmyZlMa1Cik7\nQRd/PF5SSdr4OFIPdW3Gq7OwlJD/d1+1NOdsconQZnp/FoYQQmz1DidPntw83lLJoQIAGpN2\n7drt3r2bz+cjhFgsFtUf3bZt27///lvVoUmk7D7oghcfEUKWejzqIcbSMOex331ORshBdExh\n2mOEkN6X6ysXxaS+/aJvZjnIZ/Zgh5aiAwQCwfnz50UP27Vr17p1a5nCYLG+9XrzeDyVT7eX\nDsdxDoej6ihqwWazEUIYhlF//QzH5XJFfwDMRIWH4zjDX08qx/H5fOa/iRBCbDZ7zJgxLi4u\nwcHBiYmJWlpaP/74o4+PD/XXqyrSXzplRyYoECKEdNjfv03osvE3FV+rHlP6oRQhdGFHiKNL\nP4+uvLQH1/5aM7t0638j2n5bsbG4uHjdunWi46dPn96pU6e6xaOhoVG3E5WJxWLxeDxVR1E7\nDMNUXuCcDjU1NVWHQAuLxWoQr2eDeBMhhDgcDofDadeuXdXsoXKSusUpyk7QHC02Qii/4vuH\nxtcKAuc2q3oMxsIQQlZTNq34xQIhRHoPmzd6QtiG8yP+GfPtAAxr1uz7KXVrBVNdTgz/5EcI\nYRjG/CARvJ5yJeoPbRChNoggqR8YGCqzWtDa1iYIpbwpECB9PkKIJMsyyiv45h2rHqNpqY0Q\n6tDdiHqIsZo5avLOFieKDtDR0ak6Tb6kpCQnJ0emMDgcDrWC+tevX6WXs1I5bW3t8vLysrIy\nVQcijZqamoaGBkEQubm5tR+tUvr6+gUFBUKhUNWBSKOlpcXj8YRCYX5+vqpjkYbFYunq6ubm\n5jIw8VWlo6PDZrNLS0uLi4tVHYsYBgYGknYp+yahRotRajiWFPIt2xa+CSmsJK2821Q9plnr\n4TiGxd/IoB4Swi+xBeVqBj8oOVQAAFAtZbegca7xMnfrVWc3bD0w2lq3MirkrLrxD/M76yOE\ngv3n3y8s37onSF3dZkHvVtuOLf3jyyBrA07SrYufKrh+K/spOVQAAFAtFdy+tJ+8wZ+389Tt\nS3GF7A5d+m+bP0UNxxBCXz99/JBfRn1V6r1gJ2Gw89zdOyG5QjNLO3+/2T2NGsaNHQAAkJcG\n0MFfq5KSkpKSEplOEfVB5+bmQh90/UEftHw1rD7onJwchqcR6IMGAAAgZ5CgAQCAoSBBAwAA\nQ0GCBgAAhoIEDQAADAUJGgAAGAoSNAAAMBQkaAAAYChI0AAAwFCQoAEAgKEgQQMAAENBggYA\nAIaCBA0AAAwFCRoAABgKEjQAADAUJGgAAGAoSNAAAMBQkKABAAxSWlr69u1bhi95ozSQoAEA\njJCUlDR06FBzc3NHR0czM7O5c+dmZmaqOigVgwQNAFC9R48eubq63r17l1resKKiIiQkxNXV\nNScnR9WhqRIkaACA6q1YsaKioqLqCs4kSX7+/Hn79u0qjErlIEEDAFSsoKDg0aNHVbOzyNWr\nV5UfD3NAggYAqFh+fj7Vs1ENSZJfv35VfjzMAQkaAKBiRkZGXC635nYcx83NzZUfD3NAggYA\nqBiPx3N3d8cwrNp2giBGjhypkpAYAhI0AED1/vjjD0tLS4QQjuOif93c3CZOnKjawFSLreoA\nAABNiFAojI2NTU1NNTAwcHR0bNmyJbXd0NAwJiZm//79165dy8jIaNeu3ciRI4cNG1azWd2k\nQIIGACjJgwcP5s2b9+rVK+ohh8OZM2fOkiVLWCwWQojL5fr6+vr6+qo0RmaBBA0AUIb379+P\nGDGirKxMtEUoFP75558sFmvJkiUqDIzJoA8aAKAMf/31V2lpabXBzhiG7dmzp7S0VFVRMRwk\naACAMjx+/LjmRpIky8rKkpKSlB9PgwAJGgCgDJWVlXXY1cRBggYAKIONjY3Y7Ww228rKSsnB\nNBSQoAEAilVeXo4Qmjp1KovFogY4VzV69GhtbW1VxNUAQIIGAChEaWnpxo0b7e3tW7VqZWlp\nuXPnzvXr12tpaVU9ZujQoevWrVNVhMwHw+wAAPJXUlIyaNCgxMREaqZJQUHBqVOnLl68ePLk\nyZSUlBcvXhgaGjo7Ozs6Oqo6UkaDBA0AkL99+/YlJiYihERl6kiSLC0tDQwMjIiIUGloDQl0\ncQAA5C8yMlJs8aO4uLgmXkFUJpCgAQDyl5WVJanE85cvX5QfTwMFCRoAIH/GxsY1B2wghDAM\nMzIyUn48DRQkaACA/Hl4eNRcwgrH8Z9++klHR0clITVEkKABAPKRmZkp6l+eNm2ag4MDQkjU\nE41hmLa29oYNG1QWXwMECRoAUC9CoXDv3r3t27e3tbVt3769o6NjREQEj8eLiIj4/fffrays\nOByOqanppEmT7t6927ZtW1XH25BgYjvyG5aSkpKSkhKZTuFwONTkpdzcXLFrCTOHtrZ2eXl5\n1SKNDKSmpqahoUEQRG5urqpjqYW+vn5BQYFQKFR1INJoaWnxeDyhUJifn6/qWKRhsVi6uroe\nHh4REREY9i2Z4DhOEMTKlSvnzZun6gC/0dHRYbPZpaWlxcXFqo5FDAMDA0m7oAUNAKi769ev\nU+OaRU09giAwDNuwYQOM1qg/mKgCQHXx8fHUcF0rK6v+/furqampOiLmunjxYs2NJEkKhcKb\nN296enoqP6TGBBI0AN+VlZUtXLgwNDRU1B7k8/lt2rRxcnLy8/Nr3ry5asNjoNzcXKpPo+Yu\nmJBSf9DFAcB3S5YsCQkJqXpjhionf/DgQTs7u40bNxYXF9+9e/fs2bMJCQkMv3uhaOXl5eXl\n5WZmZpJeB1NTUyWH1PhACxqAb3Jyco4fPy5pL0EQW7Zs2bt3r+hGk52d3bZt2+zt7ZUVIFOE\nhYVt2LAhPT0dx3FLS0tqIF3VTzUcx3V1dX/66SfVxdhIQIIG4JukpKRaG8VVhwE8f/586NCh\nN2/ebDpNRZIkvby8YmJiqIeVlZWpqakkSWIYJurowDCMy+Xu3bsX+u7rrzEMs6PKgcsEx3EO\nh4MQEggEDH8FOBwOQRAMXxOIxWKx2WySJAUCgapjqQU1fK1mIv748aOHh0dCQoJMV8MwbNq0\naTt37pRfgAghxGazWSwWQRDMGQ5YXl7+7NmzP/7449q1azX3slgsNze3lJQUdXX1Hj16LF26\nlFEfWlwuF8OwysrKiooKVcdSHUmSfD5f0t5G0oKuc28gQRAMT9AIIZIkGd7dKaq6wPA4KQRB\nVIuTIAgvL6/nz5/LeimSJG/fvq2435ohr+e1a9d8fX3fvXsn6YDKysoBAwacOnVKtIUhkVOo\nNj4z30fSQ2oMCbqysrIOE1V4PB5CqKSkhIH/Z1Wx2ewGMVGFw+GQJMnMiQBV8fn8srKyai3T\n2NjYR48e1e2CxcXFcv+tcRxnsViVlZVMeD0fPHgwfPjwWr/DZWRkMCFasTgcDo7jQqGQmRFq\naGhI2gWjOABAdWg7U3Ac79ixo3yDYZotW7ZUVlbW+kXTxMREOfE0KTIlaPLR2aDRwwY62Laz\naGvTd+CwwP9uVDC9ewCA2tUsLU/zGIIgJk2apICIGCQuLq7W7Mxms93c3JQTT5NCt4uDEGb6\n/tw96FY6QoitrsMTFrx7nXTjUviePZMf3vrbhAstcdCASRkq5+TkxOfz7ezspk6d6ufnd+3a\nNQzDMAwjCILNZi9fvrxv377KDFX56Nz49fPzgyrPikA3Qd9Z1C/oVrpXwM7V88Z2MNHFSGFG\n4s3NS6fuPH/QbeGE57tgwCNowBwdHZ2cnO7du1dtMK+5ufmpU6e4XC615fjx49euXbtx40ZW\nVpalpaWXl1ebNm1UFLLytGvXLjk5WdJeHMfnz5+/ePFiZYbUdNAdZuesw88aeuTlvyOqbiQr\n8ka0aBkp7FKWd1sx4dEC1exUrhFUs8vJyfH19b169apoi6OjY1BQkIWFhVLjQwgxrJrdf//9\nt3DhQrG7jI2Nb9++raOjw/ChUA23mh3dFvTzYqHr4upf5TC2zlzn5uHnZRs6CgAD6evrHzt2\n7OHDh0+fPhUIBHZ2dj179qTTN93ojRs37ty5czdu3Ki5a/PmzZaWljk5OUoPqqmgm6A9DdSe\nX/mAbPSrbX/0Ip+vO0jeUQGgGg4ODtQ6IEAEw7DQ0NClS5cGBweLBtupqamtWrVq8ODBqo2t\n0aOboH/bP6H1iL5bW1+cP7Q7CyGEEFlZeHbnjMXJud7/rVFcfAAAJli/fv3s2bOvX7+ekZFh\nbm7u6uoKtf2UgG6C3nQFc9AqWTSsxx+m7W1am/Er89JSnr/LKWNx9ItOzh9y8vuR58+fV0ik\nAACVMjU1nTBhgqqjaFroJug9e/ZQPxS8f3n3/UvR9kphTmRkpPzjAgCAJo9ugv706ZPY7UVv\nkt6Wt7S11pZfSAAAABCin6CNjY3FbCUrjk+cODNpTH76enkGBQBjlJeXnzp1KiEhgc1md+7c\n2cPDg81uDBVsQINAeyZhRc7asV57I29/Lqo+gFS3vcRKHwA0aE+ePJkyZcr79+9FNen//PPP\nf//919LSUtWhgSaBboJ+/Hu/304+1Wplba2ZnvK5xNy+i64wJz75PUvdNvTaAoWGCIBKFBUV\n+fj4UFNvRBMxXr58OX78+JiYmMbRji4oKPjnn3+ePn0qFAo7deo0bdo0Q0NDVQcFvqNbQ2Nl\nUIq60eiPb5OSMj53UOc4Bl14kvjuw+097LLnl9OLFBoiACoRERGRnZ1ds3L0y5cvb926paqo\n5OjBgweOjo7r16+/cuXK9evXt2/f3r179ytXrqg6LvAd3QQdk19u1GOqJgvDWFrjm6u/uvAR\nIWT846wgx+Z/T/xHkRECoBpSClAkJiYqMxJFKCsrmzRpUl5eHkKosrKS+hwqKSmZMWMGzAxk\nDroJWhPHK4q/FbXq1rbZx4vfRtrZu7UofB+kkNAAUClqUTRZdzUUUVFRmZmZNb8fFBUVnT17\nVlVRgWroJuixzdUz7/nfTC9ECJkMMcl7ubmYIBFCXxLzEcno5fIAqBspc74bwXTwtLQ0Sbte\nvXqlzEiAFHQT9K+7R1eUPO9joXMkq8Ri5AxB0WMbN5/5M8d4hr/Vtw9QaIgAqET//v1tbGxq\nbm/RokXnzp2VH498qaurS9olZQUmoGR0E7S5x/6nx/4Y0qcrG8M0jCcdn9vzXdSJHfuPVWh3\n3ntmukJDBEAlWCyWi4tLze2fPn06cOCA8uORLym1+pydnZUcDJBEhpVQ7EatiLj+YJShGkJo\n1K7b7x7dPHf1Ttr7e8NbwuctaJxOnTpVM4thGPbvv/+qIhx5sra2HjVqFKqxlJeLi0vv3r1V\nFBSoru5jOc269jKTYyAAMExZWdnHjx9rbidJMi0tjSAIHG/YK71t2bLFzMxsx44d1HIQbDZ7\n2rRpAQHQY8kg0hK0pqYmzasUFcFQaNDYcDgcFoslqoBcFZfLbejZGSHE5XIXLVo0a9asFy9e\nCIXCDh06NGvWTNVBgf9HWoJu1aqV0uIAgGlYLFaPHj3i4uKqjUXDcbxXr16qikruNDQ0unbt\nquoogHjSEnRKSgr1A1mR++JVFt+wtYU+TylRAcAIy5YtGzp0KEJIlKNxHGez2dAPAJSD1te0\nwoxtHTp0GLz6iaKjAYBRevToERoaWnXpbisrqzNnztjZ2akwKtB00LpJqG7k05y74X34kcpd\nTixFRwQAkzg7O9++fTs1NTU9Pb1169Zt2rRhseBNAJSEVguard7x0ZWdRjn7+83fGv8us6yC\n0UusAyBfLBbL2trazc2tXbt2kJ2BMtG9E917yrYyDn5jxyJ7C2N1LrsahYYIAABNE93cqqmp\nqanZwUChsQAAAKiCboJ++vSpQuMAjUZqaury5csTEhIIgrC3t/fz83NyclJ1UAA0SDINtidj\nTwbNHe89eIDrwvtZZbnnQ288J2o/CzQhv/32248//hgdHZ2dnZ2bmxsdHe3u7r5161ZVxwVA\ng0Q7QZMVG4fb9Bw1Z8/hkAuXryV8LS/LvTiyb6duE7eXQZIGCCGE9uzZExRUvTg4SZIbN25M\nSkpSSUgANGh0E/SrI0OXnEluN2TBpVvf+jo0Wsz097Z9eshv1InXCgsPNBhCoXDRokVid5Ek\nGR4eruR4ZCUQCPbu3Tty5MjevXtPmTIlOjpa1REBQLsPOnDxDTW9wY/Ct2qxvtW+4mh02nji\nacFt3cMLViOfwwqLEDQMsbGxAoFA0l6G14D//Pnz0KFDX79+jeM4QRApKSkRERE+Pj5//vmn\npJqcACgB3RZ0aHapcS8/UXb+H9boXs1Lc2CBHIAyMjKk7JVvFZ7r168PGTKkc+fOLi4uNDu4\nCwoKpk+f7uTk1LZt2zFjxnz69Knq3iVLllArjFBTuql/jx07durUKTmGDYCs6CZoMx6r8PX7\nmts/pxezuC3kGhJokKQ0nxFCVEULufD29vb29r53796HDx8SEhI2bNjQoUOH3NxcKadERERQ\nU7Rfv36dlpYWGRnZuXPnv/76i9qbn59/6dIlkqw+/QrH8ZMnT8orbADqgG6CXtrNMDd57pH4\n//c2yH7635R7mXp2/goIDDQwp0+flrSrVatWffr0ET0sKCi4detWeHh4YmJi1bQoEAjOnDkT\nGhoqJddv2bLl+vXr1TZmZ2cPGzZM0ikfPnyYNm1aRUVF1Y0EQaxYseLdu3cIoY8fP4qtKUoQ\nhJSF+wBQAroJ2jNklzG7dEI3y4FekxFCqQd+H+vpatptYimutzXEW5ERggagrKwsMjJS7C4M\nw+7cuSN6+Pfff9vb2w8fPnzatGl9+vQZNGhQcnJycXFxnz59WrZsOX369NmzZ7ds2dLZ2bmw\nsLDm1Q4ePCj2WZKSkvLy8mpu//Tpk7Ozc7V6oRSSJDdt2vT27Vux2ZmKXEdHR+wuAJSDboJW\nNx7+/OFZb0eNS2HBCKG3oQeOnr6m390r9G7SGDO6df1Bo3T//n07O7uaXQQUDMPU1NSon4OC\ngpYtW1Z1eYdHjx4NGTKkQ4cOiYmJVc968eKFvb19zat9/fpVShg1N65bt07KahIhISGOjo59\n+/bFMKzmzUCSJMWuSQiA0tAdxfH8c4mt7eBjselbXicnJiYWsg2sra2t25g0+FUlQP08f/58\n+PDhUjol9PX1qdwnEAgCAwOr7SVJsqCgQOyJhYWFW7ZsqTZ0T8o6JlpaWjU3nj9/XkrwVcNA\nCGEYJvqYwTDMxMRk1qxZdE4HQEHoJli7lnpOQyb+dfp2M4uOru4jhw9y6QjZGSC0efNmoVAo\nqfmMEBLN87537570G4k1hYaGVttiYWEh9kgcx7t3715tY3l5uUyLsYl+CxzHPT09L126pKen\nR/90AOSObo61bcG+F3lohmdvA6P24/3W3Uj4VPs5oAm4ffu22B5ekRkzZlA/1KGcS3FxMUJI\nIBDs2LGjb9++ZmZmnz9/FnskQRAtWrSglqkW4fF4sg7vmzFjRkxMzNu3b/fu3WtsbCxrwADI\nF90EHf8+LyHmzPKZI1uh94e3L+9rZ9Km+6C1+8LeFwkVGh9guPLyckm7MAxr3bq1qGFbh9ao\niYlJYWGhm5vb2rVrExMTS0tLJfWHIIRIkoyKijIyMvry5Ytoo7u7u0wzTfLy8jp27CjqNAdA\ntWj3UmBs25+Grt17MvXL13sXjswfN7js+dWVs7ws9IwHjJ2vyAgBo7Vt21ZSvzCfz9+zZ48o\nP3bu3FnWi69evXr69OnU/UMpvShVkSRpY2Mjerhs2TKZ1j62srKSNUgAFEfmbmQMV+s+cMz2\n/87fvRXibqtHCHMvH92hiMhAgzBhwgSxXRydOnW6d++eo6OjaMuXL194PNkWHfb09Lx27Zqs\nIZEkGRsbS/1saGh48+ZNX19fc3NzNpvdvHlz6ee2bt1a1qcDQHFkTdDk28dRW5bNcmpvaOEw\nPOJ5LlutxZDxCxQSGmgIJkyYMHHiRGqYGovFotrLHh4ely5datHi+xTTO3fujBo1Siis3iEm\nOkWsarNL6FuxYoXoZ01NzVWrVj18+PDTp0+HD0srGoNh2Nq1a+v2jAAoAt1hdq/uXw4LCws7\ndepR2leEEM7W7j1sqo/PaC+PProcGM3RdOE4vnnz5nHjxkVERKSmphobGw8cOLDqvEFKYGAg\nSZI129qSJonUU2JiYnBw8KRJk6ptb9u2rYGBQU5OjtgOE5IkX79+/ebNG2hHA4agm6Db9RiA\nEMJwbtd+3j4+Pt5eg1ppwlKE4JsePXq4uLgQBCG2JoZQKHz48CHNTmS5IAjC39/f1NS0X79+\noo1Xrlzx8/PLzs6Wfm5OTg4kaMAQ9BP0IB8fn9GjhlsZwQ1uIJvy8nJlZmeRpUuXihJ0fHy8\npO7yqjAMk+mmIgAKRTdBv4yLRAjlvYza+veNly9ffhFqdujQoZfn5AEd6zCSn7x7fE/Yrcfv\nC9kdbbtNnz/VhCdxKfuynLtTp24089u3rhcMSm2oNDU1qdFvSk7Tb9++LSsr4/P5CKHdu3cT\nBFFrguZyucHBwb/++qu6urpSYhQvLy+voKAAPioA/e5j4t9FniYdXRetWPvXfyFnjh9ct2rx\noE4tPRcfkvU9lxC8bP3xK/zWDu69O2U8uOQ3b3MpIf4aJFGyO2BHQSWsqdXgTZgwQSWN6JKS\nEuqHx48f15qdEULl5eXbtm1zdXWVMuBaoe7cudOnTx9jY+P27dvr6OgsXbpUbBEo0ETQTdCv\nj42ctPU0r53LhoMhN2IfPYm7GRq8qW9b3uktE0edlKEkIyHMCoxIMug2L3Dx7LFTfTetciv5\nFLszQXwx3wd/L7mZVUL/4oAJwsPDx44d279/f19fX6qeJ0Jo/vz57u7uSo5EU1OTy+UePHhw\n4cKFOTk59E98+fLln3/+qbjAJDl//vywYcOSk5OphwKB4J9//nF3dy8rK1N+MIAJMJrtmlkm\nWsHFNs8+3bZS/94rUlHy4qcW9gmaUws/7Kb5fIXvdozxjeq99fDCdtoIIbtBMXoAACAASURB\nVLKyePQIH7WOK4MDHaod+TXx+MRlJ3pPH3Fjf6jt4r+qdnEQBFF1RQwul8tmy3bHksPhaGpq\nIoTy8/PpNKxUSEtLSyAQSJmwxwR8Pl9NTa2wsLBr165VV7fCMGzWrFmisWseHh63bt1SWlSu\nrq7x8fGZmZlVqyDRgWGYmZnZkydPFBdbTQRB2NraZmVl1fyDDAwMZGbZJhzHtbW18/LyVPL1\niL5mzZqxWKyysrLS0lJVx1IdSZJSJtnSzWsh2SXGrmuqZmeEEFvdaq2LiWvkcYToJuiCFx8R\nQpZ63yYsYCwNcx773edkhP5fgq4sS1v9e2jzH+fM/hG7sb/6RfLz8z08PEQPp0+fPn36dJoB\nVKOtrV23E5WJzWartkuUJhcXl2prD5IkGRQU1LNnTx8fH4RQ7969lZmgb968SZVnkjV9kCSZ\nlZWlq6urmLjEi4+PF1tpBMfxmJiYZcuWKTMYmTSUqtl8Pp+6IcEo0kea0u3i0MBxQb6YAupF\neQKcJUM9aEGBECGkw/4+N0GXjRMV1Yr8kiG///6Ba7NuYT8EGojs7OyHDx+K3bVq1SrqB29v\n5a3tgGFYnUePYBim/EpJkvqaCYKQqX8GNCZ0W9Dz2jRbcn/60RSXMdbfPy3zU09Oi/3czHIj\n/efjaLERQvkV3982XysInPv/So69iww8kVQ8c3eAARsX2/emqakZFBQkemhgYJCfn08/BoQQ\nm83W0NBACBUUFDD825mGhoZQKJS1UKeS8Xi86OhoSXszMjKo/6BWrVrp6upKKbrPECRJ/vLL\nL7L+UdWTpO+5OI6bmpoqORiaWCyWpqYm899EmpqaLBarvLycmb35Ur7H003Qk0OWr7BfNN7W\n9LDPOCfrNpqo6M3L+4ePXC4iOZtPVp+vJS0UaxOEUt4UCJA+HyFEkmUZ5RV8845Vj0mMSCJJ\nwd45Pnv/t+X55unum1HY2QguhhBCHA6navHfkpIS0c16WVVUVDC8D5okycrKypqTpBmFzWaz\nWBLHSmIYRsWfkJCgtOxcn5RhaWk5b948ub/mubm5UVFRaWlpWlpaNjY2zs7OQqHw/PnzKSkp\nzZo169GjR48ePR48eFDtD5IgCE9PT2b+AVChSi8IzgRUeARBMPNllIJugtaz9Xt+jj95zpLL\nh/de/t9GLXPnXbsPzukkw1BojRaj1PDopJBEFPADQqjwTUhhJfmDd5uqx7T38BpT+K3BWFn2\n+sSpOCNnD1dzTYkJAMhJbm7uiRMnRPliyJAh9Gt19uvXT9K9OFNTU+qHt2/fyitUhXr9+nWP\nHj2GDRvm7+8vl57onJyclStXVlt/wMLCori4uGpx1F69ehkYGGRlZeE4ThAE9e+UKVMGDBhQ\n/xhAQyTb3W2ysvDp/YcpKSl5SMfKysqhR5dmLBmK7VKeHQxYdTblJ/fR1rqVUSGnPzZzCN63\nRA3Hgv3n3y8s37onSB3/fs2yr1dHTthVbRRHNXVoQXM4HOprRW5uLsNb0Nra2sr5ahYZGTlv\n3ryCggIcx0mSJEmya9euhw8fNjIyqvVcNTU1DQ0NFxcXsR0dx48fp2b0RUdHjxw5UtJFNDQ0\nqAr9zGFiYnL16lXRK1BeXi5rQT6EkEAg6Nevn2jwnBQYhg0YMMDGxubu3btfvnzp2LHj2LFj\ne/fuLXPcysJisXR1dSXVNmEOHR0dNptdWlrKtD8wioGBgaRdtd8kTH98K+TYt0YzxtLq1Nni\nfWryueDNS5YuW+S//W2JzPXG7Cdv8B/p8uHhpX9PR2t26b9tl78ajiGEvn76+OHDB2b/Rzda\nb9++nTZtGrVAFEEQ1PvtyZMns2fPpn+RS5cuVVt3isVirVq1SjTf2sHBgc/ni22V4zjOwDfP\np0+fNm7cmJOTs2jRImtr61atWnXs2HH58uUy9QiHhobSyc4IIZIkL1686OPjExUV9eLFi5CQ\nECZnZ6AEUlvQpDBwQp8Vh2PVDTyLv4QhhEiiZEqHlsEv83CWmhZXmF9awdd3fJQe21FdlYWT\noAVdf4GBgZKmZty9e7dt27bST6da0FSxpPv374eGhn769KlTp06zZs2qtujU7t27f//9d1kH\nJquQjo4Oh8PJzs6uGnCrVq2uXr0qpeFT1ezZs8PCwuj/vocOHfL29ubxeEKhkJn3BkWgBS0X\ndWxBP98xZMXh2N7jlvxzIIDa8vH6tOCXeWa/bMwsLcorKX92boda/qMhY8PlHC9QupSUFEl3\n+VJSUmS6VPfu3Tdv3nzkyJGAgICaSwJ6e3ubmJgw/P1cVV5eXs0qIh8+fNi0aRPNKxQXF8u0\n7JZMB4PGTVqCXh14u0WvvTf+Wz/a/du6GOf9r2IY5/iR+QYcHCHcbsi8s1Ot3l/wY3QTFNSQ\nlpbWp08fExMTIyMjc3PzmTNnslgsSUmTy+XKen2CIJ48eXLixIlr165VG8O7cOHCqrNAlUBT\nU1PSolx1RpLk+fPnaR7cpk0b+h9IGIbVYW0w0FhJ7JpwcHCIzynlvtnl4PCPaOOL+BwMV5/n\n8qNoi7DwY0V5fpduDhwMSZqnABjl8uXL48aNE6WMkpKSU6dOqampie3qYbFY3bp1o39xoVC4\nadOmf/75h+rORgjx+fyFCxf6+PjMnj377t27yh/QPW7cuOjoaFm/B9SK6vSQ0tq9d+9eaGho\nWlqahoYG/S6dUaNGVV2JBjRxEhN0nz590uOfEnr2ffqYUFsEhXGPHhGGjkP7/PR9Ybe8F1nx\nL/N79e7Dh2VVGoJq2VmktLSUz+cLBAJRmqZyyvTp0/X19Wle/P379927d8/Kyqq6saysLDAw\nMDAwsP7B1015eXlwcHBMTExGRsZff/0lr08IIyMjSdn56tWr69evT0hIoFYCQ/8bLywdhmFj\nxoxZv369XMIDjYO0D/bAdnobit0zM4KpcW/n5nR0D0r2S8zZ+r0GNLHewXj1C9PywkdKiVY8\nuElIU3x8fL9+/ST9j2toaHTr1u3mzZvUQy6XO3fu3MWLF9MpREXdJDQzM3v//n3946yKGgtc\nnyuw2WxqbUMWiyXHFbamT59e81OnoKBg8uTJMTExMl2KxWJt2bKlR48e7dq1o7ZoaWnBTUI5\narg3CaW996Ydnr/yx9WtHXLGuHUqeHXj4OkUjRYj1nfQQwiVf024Fp14K3zLxkdffvwtQv4h\nA3l79eqVm5ublDdSeXn5qVOnkpKSkpOTtbW1O3fuLH2UQk5OTnp6uoWFha6u7rVr13bs2CHf\n7Kyvr79s2bKFCxfW8zqilWflmJ3btWu3ePHimtv9/f1Fn3D0qampjR07Vh5xgcZGWoI2clp1\nfVeej9/O7U/OI4Sa23lFXP+PmmyddX/hEM+rCKHOwxZfXtlDKaGCuhMIBIMGDZKeoagpGB07\nduzYsaOUwxBCT58+DQgIePz4MfXQwMCg1oX+pMAwTFNT08zM7NWrV0KhkCAIqnclJydn+fLl\ndb6s4nTr1u3MmTNqatXXfsvNzT1z5kwd2pI//PCDnEIDjU0t3177zNmWMe23l4nJmKG1Vavv\nZZKatfHZsMPLrnOvgT9ZKzhCUF+VlZXe3t61FsGws7Ojc7UnT5788ssvVWsa1Cc7I4R69eq1\nYcOGNWvWJCYmUltEOY4JpW2q3d8zMzM7duxY1excWVkZFhb24MGDN2/e1KE3Bsfxffv2ySdW\n0OjU3r2Ic7WtuzghhIiKr09i73bqNYiLIZxl8vpe0P3H8Tn8lWO71z4VGKjQvn37bt++Xeth\ndKZ0CwSCX3/9VSAQyKvPsU2bNmFhYbNmzbp48aJcLihfdnZ2bdu2vXz5cnFxsba29siRI/39\n/auWP/748ePo0aOTkpJE9wNldf/+/ZqjxQGg0J0BWJ5/e2i3wZdeF3wVEhyUN7Sbx/W8MoRQ\n+LFjZalvp5prKTJIUC///PNP7Qch9Pr1a+kHPHjwYObMmenp6fIICiGE2Gz2oUOHMjIyTp8+\nLa9rypG6uvqBAwcsLCwQQnl5eWLL0s+YMYMawEdVL5H1KcLDw83NzesdKWi06A6OOzth3KXX\nBc4jp/Bw7MvThdfzyty3hT65edyMVfDb+DCFhgjqo6ysLCMjg86RHA5Hyt7s7OxRo0bRvBQd\nurq6jx8/tra2fvbsGQPHALBYrKioKCo7IwmLhrx48SIuLq7Og0zCwsJ69uxZ5whBU0A3Qf92\n9YN+x223Tv6jhqPnG2Jwtu4hX8/OvUb96Wj05eFmhYYI6oPNZtP86u3s7Cxl75EjRwoKCuo5\n4o3P52toaNjY2Bw6dOjly5fUjIznz5/X55oKMn78+ForkKSmptZ6HU1NzZrLlbFYrJs3b0Ih\nJFArul0c78oqjK2/zUA9didTq+ViatkqPXONijsyrOoNlKmgoGDzZlofn2pqakuWLJFyQEJC\nQj2HJFtYWLRt27ZZs2ZOTk5Ufbvs7OwpU6bExsbW+ZqKgGEYjuNTpkwRu7eysvL48ePR0dGf\nP3+ms1BkUVFR3759jY2NIyMjCwoKdHR0PDw8fvvtNy0t6BUEtaOboJ21efceXSJRX0FezL+Z\nJfa/f1uz9cHjXLZae4WFB+pOKBR6eno+ffq01iNbtGgREhIiqeZGVlbWq1ev6j/C/927d1T/\n9enTp/ft2xcWFjZq1KiXL1/W87JyRM1k4XK5W7ZssbKyqnlAbm6ul5dXfHw8i8WihgPSuWx0\ndPSVK1d27txZVlbGwEVLAZPRTdAb/Oy6rdxk0+8lJ+EygfAFM9oLCuOW+C7ZnpTTwnmrQkME\ndRMWFiY9O2MYZmNjs2zZMldXV7EHvHv3bunSpVevXpVLPFVvo719+9bT0/PNmzdyubK89OzZ\ns2vXrhMmTGjVqpXYA5YvX56QkID+N+dF9OvUWmrj/v37Xbp0gewMZEU3QXcOOLc49sfNF8MR\nQr3nho01Ui94e377oRiNFs4HT0lcIwOoUHR0tKROCWr5u19++UVK4bTs7OyBAwcqaD1pgiCY\nlp0RQhs2bBBNtq6puLg4PDy8ZiKmUwiJCQO6QUNEN0HjHKNNF14FvH3+iTCwbWOMEOLr9f83\n1LnvYFczNVgskIkKCwsl7Zo5c+aQIUOknx4UFFR1uTwpbGxsRHNMGi4+n29mZiblgA8fPohm\njVdFZwhKhw4d6h4ZaMJkqkFHvrh3c9/qXwcPcF14P4uoyFc3aNUKsjNTmZubS8odc+bMCQ8X\nv8xCdnZ2ZGTkwYMHL1y4QKePdenSpTdu3JC0sqqhoWGHDh3kXo5ZEcaNGyd9vUENDY06XBbH\ncXNz8z59+tQxLNC00V6qiqzY6Gm35My3pdWEC8rLci+O7BvUecK2uwcXQK1RBvLy8jp48KDY\nXSUlJdOmTVNTU+vfv3/V7bt37960aVNpaSn9Zzl79uybN2/GjRu3e/fuqr3MGIYZGRn169fv\n3LlzjKoXyOVyZ8yYsW/fPqFQKOoCGjJkyKpVq6gD7t+/f+fOnS9fvpibm7dv357P53fo0EFH\nR6dly5atW7d+9+6dTL+OmZnZf//9V4dFDwBA9Ff1fnV4SLvxke2GLNgVMGFAr86ul9IjnfNW\nTPHZdPK5x9FX4T6Wig5UCig3Ksn27dvXr18v6b+4RYsW8fHxood//fXX8uXLZV0tkGodEwRh\nZWVFkiQ1KgPDMFdX18ePH9ezTId0HA6nakkQmejo6Pz444/q6uotWrTo27dvr169EEKlpaVz\n586NiKhenZHFYk2ZMmXp0qW3b98eP348hmE0/2Z69Ohx+vTpOmRnKDcqXw233Cjdd+MkY82T\nwj6ZWee0WBiGYa6X0q/0N0WoclYr3cNCj6LMw/KLVmaQoBFCmZmZr1+/bt68uYWFRdXVBR8+\nfDhw4EBJZ4l6mSsrKzt27Pj169c6v9MwDBsxYsSqVasyMzMtLS0XL1586tQpxr5vqbbz77//\nXnXZ8vnz5x89elTSKbq6ura2ttra2vHx8XTmu2MYFh8fb2xsXIfwIEHLV8NN0HT7JkKzS417\n+WmxqnVKskb3al6ac7YesYH6SktL8/b2trW19fDwcHJycnR0vHz5smhv69atpZwr+nBKT0/P\nzc2V9DajUwmIJMlTp07xeDx7e3sNDY3IyEgmv2mpUczr168XLc2Vl5d38uRJKad8/fr19u3b\nkZGRNOe7b9q0qW7ZGQARun3QZjzWl9diyrF/Ti9mcWEJNZXJzMwcOHBgXl6eaMuHDx/GjRsX\nHBw8ePBghJCmpqaU00VpV+z4BIqmpqadnZ2hoSGLxbpz505mZqakbhCCIF69evXs2bPIyEiZ\nOrJVgiTJsrKyR48eUVOuU1NTpbwIolMQvWEbO3bs8PHxkUucoCmjm6CXdjOcGDv3SPwvY+1E\n612h7Kf/TbmXqeewRjGxgdrt2rWrWr8EQRA4jq9atYpK0FJavjiOi/ZKGcDQsmXLs2fPIoSi\no6OfP3+emZkpJUP9+uuvqampDWLYBkU0GFFeMWMYtmbNGsjOQC7o/lF6huwyZpdO6GY50Gsy\nQij1wO9jPV1Nu00sxfW2hngrMkIgza1bt2puJAgiPT2dWoCKy+VKGoTr6Ogo+lnKotcaGho5\nOTm//fabt7f3q1evpASDYRhVP4jh3fpVWVp+u7/doUMHPp9ft5rOIubm5s+fP58xY4Y8QgOA\ndoJWNx7+/OFZb0eNS2HBCKG3oQeOnr6m390r9G7SGDNpX6KBQhUVFUlqz4p6V/39/dH/b0pT\nfcqLFi0SbZHSKI6Pj7e2tg4KCiJJUnrmZXKnc004jtvb27NYLKoRra6uPmvWrPr8Cjwe78SJ\nE3TWPQCAJhm+1unaDj4Wm/7h1fMrZ0+eioxKfP0h/e7JYd0MFRccqFXbtm3FfjfncDiieXFD\nhgz5888/q5ZP09XV3bdvX9XZE/r6+pKeotae2QaKIIj4+PiePXu2adNmyJAh8fHxAQEB8+fP\nr1tfB5vNPnbsWK0VSgGQSS3D7B4+fIgQsuvmwK3XNz/FasrD7CIiImoWxsQwzMvLa8+ePVU3\nfv36NTY2NiMjw9zc/Mcff6y2zFJSUlJTLk+M4ziLxTp79qyjo6ORkVEd2tF///330KFD5RUP\nDLOTr4Y7zK6Wm4RUN2VaaUU7TWmzYBtrI4tpCIIIDw+PjY3Ny8tr3779mDFj3N3dZ82atW/f\nPmpkBTWHonPnzoGBgdXO1dXVpW4bimViYlLPcs8NGvWLr1ixYsyYMbLmGgzDAgIC5JidARCh\nO4rD1tZWoXGAWmVkZAwcOPDz58+iLTt27Ni0adMff/zh7u5+/PjxV69eGRsb9+3bd+TIkbJ+\nT9fR0XFzc7t8+TLDm0KKQxDEkydPZC24gWFYZGRk1dutAMhRLQla9HalU/cdKE5hYWGvXr1E\n9/0oAoHAz8/P3t7ewcHBwcFB+hWEQmFycvL169fj4uI+f/5saGjYs2fPGTNmqKmpUQd4eXld\nu3atKX8ZIklS7KgYKQICAiA7A8Wh285q1arV6Bsfa25/tta1tZXEmcRAXtatW1ctO1MIgvjv\nv/9qPT0qKsrJyennn38ODAyMiopKTEyMiYkJDAx0cnJ6+/YtQujNmzfTp09vytlZVurq6tu3\nb1+4cKGqAwGNWS0t6GvXrlE/fPjw4eXdG9cq/v8QIrLiXtyL9LQCBQUHRC5cuCBpV3JysvRz\nb9++7ePjU63vgnr4+fNnX1/fc+fOjRs3jlolBNQKwzBLS8tLly5R95kBUJxaEnTVxZAeLxsj\ndmWkZuZz5RoSEKPqZO5qqpZGEmv9+vVIwiBlgiDi4uLevXvHqLUBmYwaTr527VrIzkAJaknQ\ny5cvp34IDAxsPcrXx7JZtQNYPN1BM+coJDQm+fnnnxMSEqgch2FYSkqKnp5erWfJkbGxcVqa\n+NXTnZ2dpZxYUVHx4MED6bf+oqKimuy9QVlpaWmtXbv2559/VnUgoEmoJUGvXbuW+uHEiRM9\nZy1d+1NTrItUbWAsSZJWVlZ2dnZRUVFKi8HDw2P79u01t/N4vGnTpkk58ejRo7Um3wZUOkOF\nMAzbv39/7969lfzZDJqyWt6ZDx8+fPjwoYBEr169OtQks3OrVq3EJriqpe6VwNfX19rautpG\nDodz4sQJKd+1SZLctm2blMviOK6np/fLL7/IJ8rGi1oBfdiwYZCdgTLVkqAdHR0dHR0/lFey\npVJOrMpXVFRUXl4uaW/79u2VFomWltbVq1f9/f1NTU0xDNPV1R0+fPjz58+l9298/Pjx40cx\nY28o1NyWwMBAfX39Tp06yRpSk2p3kyQ5dy7cawHKBhNVpLlx44aUvVJu3CkCn89fvHjx4sWL\nKysra70xSJE+bM7c3HzdunXUfeCwsDBnZ2exy3jXXFmKqrXUpKYd+vr6enp6qjoK0OTARBVp\npDQ/VYhmdkYItWjRQkNDo2b9ARzHnZ2dT506JdoyYcKEmtkZx/GRI0euW7fOz8+v6irgVReH\nbQpcXFxES8oCoEwyfEslKr4+unlBQCKEUGHaleljhnpO9D1yP0tRoTGArq6ulL0uLi5Ki6Ru\nuFzuxIkTq22kGr9V1+KbPXt2XFxczdN5PN727dtPnDhB1czCcVxdXb2eFZMbHEdHR+lLYQGg\nOHQXjS3Pvz202+BLrwu+CgltlNfP0OR6XhlCCOfo7U99O9Vcq9YrKI7iqtllZWV17txZ0tLR\nYjsE5K4+q3ojhAQCga+v7+nTp9H/VkrlcrkrVqyYNWuW6Bgp9dssLS1fv34t61LfDZqo7BRJ\nkl5eXkFBQcqPAarZyVejrWYncnbCuEuvC5xHTuHh2JfHC6/nlblvC/3doWKY27jfxodNjZkk\np1CZxcjIaOnSpX/88Ue17RiGXb9+XSUhyYrL5e7fv3/SpElXr15NSEjg8/k//fRTtXW+pby7\nXr9+Lf2ARobP5//www9fvnyxsrKaOHGik5OTqiMCTRrdllEHDe4Xi43ZiQsQQtdHWLqe/ZpT\nmqPDxs7+1NLrkbagOEnBcUqj6HrQly5dWrNmTWpqKtWwMjU1ffDggdLGMNSzBU15+vTpvHnz\nRJPC2Wz27Nmzly1bRnVnGxrCqgvfmJubU/05qgUtaPlq/C3od2UVxtadqZ+P3cnUarlYh40h\nhPTMNSruiJ/h1mgMGDBgwIABxcXFAoFAeq80M2VmZnp6elattVRRUbFz506E0MqVK1UXFxOZ\nmJioOgQAvqPbDHTW5uU8ukQiVJ4X829mieUUD2r7g8e5bDXlDQdWIQ0NjYaYnRFC//zzT0FB\nQc0vCnv37qWW42vRoilOQaqJqoKk6igA+I5ugt7gZ1fwbpNNv2HdrQYSCF8wo72gMM5vYp+F\nSTkGXaDiIqM9fvxYbIeMUCh89OgRUvqsSMYiSXLAgAGqjgKA7+gm6M4B5xYPtEyOCo/PKv1p\nzsmxRuplOee3H4rRaOF88NRIhYYI6qmiokJSF+HIkSP9/f3z8vJevHhBf3h1YzVo0CA3NzdV\nRwHAd3QTNM4x2nThVfabhITXn27s8kQI8fX6/xt6Men1jQFGaoqMENTdhw8fPD094+LiJCVo\nkiSDg4OHDh2qpaXVv39/JYfHHK1atVq/fv2BAwea2ihvwHAylNEghPkfP2V9yM/I+F/p4Oaa\nKCnmahJC7fu4tuGrrP2F4ziPJ21N25pE9UO4XC79G9ClpaVcLlfJLU0cx9lstpRfMD09/fff\nf79582Zubq6VldW0adMmTJiA43hUVJSXlxedRVISExNPnz6tzOJ8TGBnZ2djY7NgwYKWLVtq\naalyIH9NVJdUHf6wlYyKk8fjMXwUBxUni8Vi4Osp/aWjO8yuJDOyf+eRtz+LH80WmVM6SI9f\nl+jkQSAQyDroLTMz09HRMTMzkyRJDoczbty4ffv2STpYKBTu3bt3x44d6enpPB7vxx9/3LBh\nQ9euXesdOC0sFoskSUljAePi4tzc3MrLy6kDqKkoHTp0sLa2joiIoL9IyogRI06dOsXwt5l8\nxcbGMnY5QRaLRc2UYfgyNxiGsVgs5q+URr2eBEEwsIAMNXdM0l66CXrHD8Z+D4p/GTfetqWY\nkQzL/lijjqvsu6Gs46CPHj06f/78ahu1tLTEVsQnCMLHxycqKko0m476MPj333+rTfdQECnj\noBMTE0eNGkV9zFTbJevcP3t7+2fPntUr0Ibm4cOH5ubmqo5CPBgHLV8Ndxw03bexEZdts/d5\n9JTqJYmZQNYELWlmc9++fUNCQqptDA8Pr1kRH8dxfX39Z8+ecTgcWaOVldgEnZmZuXDhwsuX\nL8vrWWqWrGvcDA0Nk5JUObtKOkjQ8tVwEzTdngE1Fvaza0s5xaNKp0+flvTHdPPmzZobL126\nVLP/hCCIL1++qKrCX2Vl5ejRo69cuSLHazL/W6p87dq1S9UhAFA7ugl6eVeDyJAXCg1FOS5d\nuiRpl9j+vpycHEnHZ2WpppLftWvXRAskykWTqoWEEJo8eTIsKggaBLoJesL5UHJz/7kb/kt9\nnyVsyO9lKXUnxA6xMjY2lnS8qqYFU7NL5KiJZGcMwwwNDc+cOSN9GTAAmINugrbqMvYDVrpn\n6YT2Zs15DXnJq6VLl0rapaOjU3Oju7t7zTu/OI6bmpra29vLOTh6mlp3hLzs3LkzKSnJw8ND\n1YEAQBfd3Kqjo4OQjr7E1mSDoamp2a5du9TU1Jq7at4hRAi5urqOGjXqxIkT1Ag2hBCGYVwu\nd8+ePapalK9jx45yvBqfz69nqbyGQiAQqDoEAGRDN0E3piWvYmNjBw8efP/+fdEWNpu9d+/e\nzp07iz1+165drq6u+/btS0pK0tXV7dmzZ0BAgKmpqbLirW7w4MEtWrT49OmTXK72ww8/REdH\ny+VSDNehQwdVhwCAbGTqnSBjT+49FhnzJivX+o+jgW3vn4u38Oxj2xDXdr5y5QqPxwsKCrp3\n797kyZO7desm/Xh3d3d3d3flxFYrNTW148ePDxs27OvXr/W8FIfDEURMpAAAIABJREFUsbCw\nkEdQjIbjeKdOnWr9XwaAaWgnaLJio6fdkjPfKr4LF5SX5V4c2Teo84Rtdw8u4DfAJM3n8/38\n/OgU7GeaoqKiM2fOGBkZ1T9Bz5gxY/fu3XKJioGou74kSXbs2DE4OFhVXVIA1BndBP3qyNAl\nZ5LbDVmwK2DCgF6dEUIaLWb6e9/cdMhvlJt7uE+TqKJbWlqqpqbiylCpqalubm5Vq+/XGY7j\nTFg9RBHU1dVXrlyZnJyspqbm4ODg7u4O2Rk0RHQTdODiG2p6gx+Fb9VifRuLxtHotPHE04Lb\nuocXrEY+hxUWoeoVFhauWLEiIiKiqKhIXV198ODBq1atkjL8TnGysrL69u1bXl4ul6sRBCF2\nMe9GYMyYMVOnTlV1FADUF91mRWh2qXEvP1F2/h/W6F7NS3POyj0s5sjOzu7cufOxY8eoRmtJ\nSUloaKiTkxO1mqoyffr0ydHRUV7ZuRFzcXGBpbxA40A3QZvxWIWv39fc/jm9mMVtzAsm+fj4\nFBQUVNtYXFy8aNEiZYZx5coVe3t7WdfGbWqsrKyOHDly8uRJlfdEASAXdBP00m6Guclzj8Tn\nVt2Y/fS/Kfcy9ez8FRAYIxQXF0saXxgbG6u0xuyLFy/GjBnTROb71YeLi0tTXnYAND50E7Rn\nyC5jdumEbpYDvSYjhFIP/D7W09W028RSXG9riLciI1Sl5ORkSWmRIIj6D6KoVW5u7s8//+zg\n4FDrkWpqak18NRAMw5hcoA6AOqB7k1DdePjzh2fnTJ9xPCwYIfQ29MBbhFo6jdy1e/cwM00F\nBqgw7969e//+vaGhob6+vqRj3rx5I+UKenp6Cojru/fv3zs5OcH8N5qa+OcTaJRkmKiiazv4\nWGz6ltfJiYmJhWwDa2tr6zYmDXHsUmxsbEBAQEpKCvWwV69eGzZsaN++fc0jpVRWYrPZUtZB\nkIupU6fSz86lpaUKDYb5CIJQVXUUABREWoKWOoUh62Na0nWEEEI8nd7TxnaSb1iKc/v2bS8v\nr6qTU+7cuTNw4MDo6GgzM7NqB0u516SEVezi4+MV/RSNBoZhampqkydPVnUgAMiTtATt6+tL\n5xK6bfc0oAS9evXqakuTEQRRWFi4ZcuWnTt3VjtYyn05bW1tRYX4P1C1jj5TU9M9e/a0bNkY\n1pQAQERagj58+Pv0k8ry90tnr8wktAZNmNzbrr02Xvwq+e7Bv88U8OyCjw9TfJzyUVhYGB8f\nXzPtkiQZExNT83hJs1EwDBPbJSJfLBZLjmuGNtaq/Hp6etu2bXN1dVV0jxMAyictQY8dO1b0\n89XJ1lmk3n+JqWPaf285rgw4Y9Xea9GfTz2ONIyh0MXFxZKSVGFhYc2NFhYWNjY2ycnJ1ep1\nkCT5yy+/KCTEKmxsbOTVy4FhWKNM0DiOu7q6Dh48WNWBAKAQdG/yLQl9Y+Swt2p2Rghpmg3b\n+4Pxu/AABQSmEAYGBhoaGjW34zhuaSm+nMj27dv5fL5ohAD1g6urq5eXl+LipBw4cEBeiyGQ\nJNngakLVCsdxHMdnz56t6kAAUBS6Cfp1WQXGFrOCNQvHKsvFzDBkJjabPWrUqJrbCYIYM2aM\n2FO6dOly9+5db2/vFi1acDgcW1vbbdu2HTlyhMViKThYZGFh8fTpUykr/jZxWlpahw8flu/y\nBQAwCt2vvVNbaB7KM7jwLtnV6PvAhrLsaBtTtyzdqYUf9yoswtqVlJTQnwNdVFQ0evTouLg4\nHMdJksQwjCAIb2/vnTt3MrDgWXl5eVpa2q5du8LCwhpfB0U9vXv3Tl1dXdaz9PX1CwoKhEKh\nIkKSFy0tLR6PJxQK8/PzVR2LNCwWS1dXNycnh+F/nDo6Omw2u7S0tLi4WNWxiCGlEUb3G/Ty\nHR4HvI/9YtVlhr/vT7btmmFFqYm3dm3cm1ZWMW5Xg+niQAhpampGREScOXPmypUraWlp7dq1\n8/Ly6tOnjwpDEggEL1++rKiosLKyqjqw7/Lly/7+/h8/flRhbIxla2tbh+wMQMNCN0G3Hnn0\nzEv+1D8O7Vw2VzQYjcVtPi3wr788LRQTm6IkJycHBQU9e/YMIfTs2bO8vLz27durZInuioqK\noKCgjRs3UhNScBz39PTcunWrmprazZs3x48fr/yQGgRbW9smskwXaOJku7Nf8vHB8TM3U1JS\nijmG7du3dxkxys6AL9rLZrMzSgTGXGV3FMjUxfHw4cMhQ4ZUG76mqakZFxfXvHlzBUQnzYIF\nC44cOVJto6Gh4Zo1a5YsWZKXl6fkeJgPx/GrV6/a2dnV+QrQxSFH0MUhF1K6OOQ59ArDsE/l\nlQxP0F27dn3/XsxdzZ9//vnEiRNyjasWNjY2WVlZynzGRmDRokUBAfXqUoMELUeQoOVCDn3Q\njUNhYaHY7IwQun37tjIjMTIyYvjfNAPNnDmzntkZgIaFceMWFCo9PV3SLmUWjXNwcIDsLCsM\nw9asWaPqKABQqqaVoI2MjCTtUsK4ZpF3794p7bkaBxzHoTsINEFNK0EbGhpKqkKnhNoaoA5w\nHN+/f39mZqaqAwFABZpWgkYIrVq1quZGDMO2bt2q/GBArdauXTt8+HBVRwGAajS5BD1x4sQt\nW7bw+d9HB+rp6Z04cYLOslLyooRa0o3DxIkTp02bpuooAFCZpjWKgzJhwoQRI0YkJSWlp6cb\nGxs7OjoquVLlL7/8cuzYMWU+Y0M0bdq0devWqToKAFSpKSZohJCOjs7AgQMRQrm5ucov8xYZ\nGankZ2xYMAybOXPmH3/8oepAAFCxJpqgVaisrIzhsw9Uq3v37hs3brS1tVV1IACongwJmhDm\nJz589CFfzHjh9n1c2/BZAQEBGixYWbkWlZWV8l0qpZEJDg6WMhoSgCaFboIuyYzs33nk7c/i\nZ1RH5pS24bM2bNggv8AaJ4IgunXrJjY7N8oVT2Q1fPhwyM4AiNBN0H8PnRL7BfeYONu2pW7N\nvX10eHKNqtFau3ZtTk6OqqNgKC6Xu2fPHlVHAQCD0E3QgY+yf9r/PHyKtUKjYayCgoJjx44l\nJSVxuVwHBwdPT08OR8z6MiIZGRmfPn2ytLTU09Oruv3ChQuSTmnizWcMw968eSOvJb4AaBzo\nvh/UWNjPrk10TfuYmJgZM2bk5ORQi7AcOnRo586dR44cadOmTc2D79y54+/v//LlS+qhi4vL\n+vXrRUcWFRUpL+6GQ0dHJzExEZblBqAauhNVlnc1iAx5odBQmCk7O3vixIlfv35FCBEEQbVz\nX79+PXHixJpdyTExMZ6enq9evRJtuXHjxoABA0SrosACgzX17NkTsjMAYtFN0BPOh5Kb+8/d\n8F/q+yxhU/ouHhYWVlRUVG2sNEEQycnJ9+7dq3bwb7/9Vm39bIIg8vLytm/fTj2cPHmyogNu\nWNq2bRseHg7ZGQCx6CZoqy5jP2Cle5ZOaG/WnMeuTqEhqtbLly8xTPzYwZSUlKoPc3NzExMT\na057IUnyxo0b1M/jx4+HVahFNDU17969q+ooAGAuurlVR0cHIR19Y4UGw0Q8nsQBKlULeiCE\npCzWUFhYSP2QlJSUmpoqr9gaNHV19eTkZFVHAQCj0U3QT58+VWgcjNW+fXtJ4yuq1Vdq3rw5\nn88vKyurdhiO49RNwtLSUnd3d4YvtqQcrVu3vnXrlpQPPwAAknWqd97LqAOhN16+fPlFqNmh\nQ4denpMHdNSr/bSG7MGDB5J2PXv2rGoVaS6XO2LEiKNHj1ZL6ARBjBkzBiE0Z84cmOSNEGre\nvPnNmzchOwNQK/rlRol/F3madHRdtGLtX/+FnDl+cN2qxYM6tfRcfKjB3TLMzc319/c3Nzfn\ncDjdu3fftm1beXm5pIPj4+Ml7YqOjq62ZfXq1V26dEEI4TiOYRiO4wghHx+f0aNH37x58/z5\n83L6DRowdXX1uLi4ar1DAACx6LagXx8bOWnraR3rn3/zn+FkbamNF79Kjtu7PvD0lomjHHqd\n9BYzIpiZ0tPT+/fvL1qHOC0tbf369REREZGRkRoaGjJdqmbXh7a29sWLF0+cOHHt2rWPHz+2\nbdt28ODBGRkZQ4cOhbthCCFtbe0nT55oamqqOhAAGga6CXrLosu8Zj3iHl2yUv92SucevYaO\ndP+phf0Fv23Ie7csT0rePb4n7Nbj94Xsjrbdps+fasKrvh4gWfE18tD+i3HJn/MrTVu3dx83\n08VWPiUaVq5cmZubK8qt1A+JiYm7du1asmRJzePt7e1fvBA/ALxv3741N+I47uPj4+PjgxCK\njo6eNWsWzO2m8Hi8J0+ewGIFANBHt4sjJLvE2HmNKDtT2OpWa11MSr4cl+kpE4KXrT9+hd/a\nwb13p4wHl/zmbS4lqjdFQ5bP/zviQcuuvb2HueoUJe9YPvNcWqFMzyJWeXn5lStXao6EwzAs\nIiJC7CnLly+neiqqadasmfSlmLKyskQzXABC6Nq1a5CdAZAJ3QStgeOCfDEpsihPgLNk+MZK\nCLMCI5IMus0LXDx77FTfTavcSj7F7kzIrXqMoODO0eSvrQb9vmzW5JGjJ6z6c1dzDnlsw0X6\nzyJJbm5uRUVFze0kSX7+/FnsKSYmJnv37q1WeUNbWzsiIkLKAPAPHz74+PiUlJQofzUAZjp9\n+rS1dRMt5AJAndFN0PPaNMu6P/1oSl7VjfmpJ6fFfm7Weh795yv+eLykkrTxcaQe6tqMV2dh\nKSFvqh5TmhWNEOo4xOJbiBwDew1ORWk6/WeRRFdXV2xWxTBMSpXL4cOHJyYm+vr6du/evU+f\nPqtXr05KSrKxsZF0/KNHj3744Ydnz57VP+DGwcPDo1evXqqOAoCGh24f9OSQ5SvsF423NT3s\nM87Juo0mKnrz8v7hI5eLSM7mk5PoP1/Bi48IIUu9b0OsMJaGOY/97nMyQt/HFGu3XVG1vyE3\n8UJUXrmBc5/vFykomDNnjujh0KFDPTw8aAbg5uZ26dKlag1bkiQ9PT11dHQknaWjo7Nt2zaa\nT+Hn5ydlWEhTM2nSpL/++kvVUXyHYZimpibDaweyWCyEEJvNlvI3yQTUJFttbW1VB1IL6vXk\n8XjSi1CqhPQv2XQTtJ6t3/Nz/Mlzllw+vPfy/zZqmTvv2n1wTicZhkILCoQIIR3298nTumz8\nTYX4jlqysjDmzOGgI5dxLZsA3y6i7ZWVlVUnofXq1Yv+dPOtW7dGRUVVS6BGRkbLli2Ty5z1\nFy9eJCUl1f86jcOFCxeotR8ZhXq7Mh+GYQ2ijkKDCBIhhOO42PtJqiV9cSUZXtm2g2bHvBr3\n9P7DlJSUPPR/7d15XBTl/wDwz8yesNz3jUooiKioIJ4oHsjPQM1U1BTvKLO8Na20w9LU8kqp\nr2Z5S5q3kqVSKp55i6CiIAqIAgKyLOwxvz9G1o1jmIVld9j9vP/wxe48+zyfmVk+Ds888zw2\nrVq16tQ5yErLNa4ElnwAKFK8vn4pVKhIoVX1knk3j69YuTGtsNyv+9APpo72FL/+pRKLxbGx\nseqXbdq0KSsrYxnA9u3bq1/e5uXlJSYmRkdHs9+R2uCCsGpr1qzp1asX+1OjH2ZmZuXl5Ry/\nNyAUCnk8nkql4vifYgRBiMVirp3i6kQiEUmSCoWCg8/xqlQqhgG+bBN037iV29fMcBZaBnXp\nHdTl9fAyWf61+eNGrTrE9prR2s8NIPVhcQXYiwGAomSPyxVi76rzBz1Oip/6/TEL79CFn8aF\n+FRdw8XMzGzatGnql1KplGEeDE0qlWrDhg3VF5ciSXLt2rV9+vRhU4lCoSgoKKixz3rTpk01\njtUzQV9++eXIkSNZnhd9oh/H5+AvqiaSJOmFKzl4ADXxeDyxWCyVSjneZSQQCEiSlMvl3Dye\nDAma7QX/iR9nv+Ebvv1CjuabyVsWtfYKWX1YiylvJK4xZiSRknCbflnyMKFESbX673MuStnD\n2auPOXefsHHV/OrZuSHy8/PVj6hoUqlUt2/frvPjaWlpI0aM8PT0DAgIaNGixaeffqr56Pbq\n1asxO9MCAgLi4uIMHQVCTR7bK+izWxaNift6TNfmB+fHb/oiVlB0Y37syFWH74hsA7/d8iv7\n9kihy4Jov88OLF25aaSfrfJEwgFzly7T29sDwOa50y+WlK/8YX3Rhf9JlZQ7L3Nz/Ab1BwXm\n/pPH1fBgiFYYOsvq7Ee7dOnS4MGDFQoF/ddxSUlJfHz8iRMnfvnllwkTJty/fx8X6qZFRkZu\n2bLF0FEgZAzYJuiuYxanDBzy8fjR3389/s8Du2wfnnwglfcc98XmHz5uYa7dLYJ2E5bOFa3Z\neybxfAnfPyjiu+kTzUgCAApzsp8UySgKCi7lA8C9pL805+UU2yobnqBtbW1btGiRkZFRpQuS\nJMnQ0FDmz86fP1+dndXu3bvXvXt3jv99p0+fffaZZu8TQqghqvbGMqsoShnqH3I4pxQAPPov\nuXvsYzq3GpZUKpVKpSwL//777++++y5JkupUS9/bTUxMbNeuXW2fysvLYxj4jGjz58+fNWuW\noaOog729fXFxMcf7oC0tLUUikVwu5/j0hzwez9bWtsZuQ06xsbHh8/llZWXc7INmWAlPi0En\nF3cuaesZdCRX2jdu2fv9fR4fX+jZYeihWwV1f5JL3nrrrdWrV1tZvR434uHhsWPHDobsDAAF\nBU1sN/XP09NzyZIlho4CIaPCtndi1uDA7w7cMnMKXr93Z1w/H4DZg36Y9c6MtYPbe436ZMPW\nxWMaNUrdGjVq1ODBg+/cufPo0SMXF5eOHTvWuSaei4uL5kU3qsLLyyszMxOPD0K6xbaLgyCI\n7rFfbo//2EtjPPLLjFNTho/ceempYf/A0aqLgyYQCOjHnwoKCmpMKxUVFXl5ea6uruqHGkaM\nGJGUlIQ5qDoPD4/U1FSJRKJSqbj/pwZ2cegQdnHohA66OL47cOv0L59oZmcAsGjWe8eFzJ1f\nxtb2qaYoJSVlyJAhnp6eQUFB3t7es2bNoucL/fbbbx0dHQ0dHecQBHH16lVDR4GQcWKboGdE\nBwBQybvXfzB2xMAB/WZdzJMVHP4t6ZaKEMV88ktjRqhXly5d6tevX3JyMn2lXF5evnXr1v79\n+7948cLb2/vcuXNRUVGGjpFDxGJxXl6eoaNAyGixvklIKZa9FdAtZuoPWxOO/vHXzcJyWcGx\n4b0DO477XmZEf/QvWLCgylg6iqIePXq0bt06ALC0tNRchNDELVmyJCsry9BRIGTM2Cbo+9sG\nz993x/fNGYmnXy3vLXGNmzuizbVfZ8bsSm+08PSqoKDg2rVrNU7nf/z4cfpnfBqF5uTkNGXK\nFENHgZCRYzuKY8mcJDO7gf/uX2lZOTuSQBK4bNe14jO2W2cshlFbGy1C/alt9ROKotTLVuFo\naAAgCILNk/EIoQZiewX92/Mylx4zLavOXccb2cO5LP+AzsMyCBcXl9oe+DYzM6OvrCMjIz08\nPDg4aaHeCAQC7HdGSD/YJhovEa8kvYYOx9xHpTyhq05DMhiJRBIREVFj8s3MzPzoo48AQCQS\n7dy509fXV+/RcQJJktnZ2YaOAiFTwTZBf9zRseDOB9tu/GeU6/NrWyZeeGrXdm4jBGYYX3/9\ntatrzf/f7Nq169y5cwDg5+d34sQJ/cbFCQ4ODk+fPjV0FAiZELYJemjCWhd+WWxHn8hhEwDg\n3qbP3xnaz7PjuDLSbmXCiMaMUK/c3Nw++eST2rbu2LFjy5Ythw8fTk5O1mdUXODh4aG5ig1C\nSA/Y3iQ0d3nr1uUDU6e8u3PPZgDI+G1TBoB76PC169YN8dJiVW/uKympYfFy2q5du3bt2gWV\nS7GZDjMzM3waBSH902KmUNs2A3ckP1qRfuf27dslfAc/Pz+/Fm7Gd7PMxcWlzjIcf7BVt0Qi\n0aNHOlhSHSGkLW1XeyTdfALcfIx2qJlMJps0aZKho+AQgiAeP35s6CgQMlHGdwXcIL6+vhUV\nFYaOgisIgmA/ou7MmTPdu3cPDAwMDw/ftm1bowaGkInABP3ali1bZDKZoaPgCq2yc58+fXr0\n6HHu3Lnc3NybN2/OmDGjdevW+F8dQg2ECfq1zZs3GzoErtAqO0+cOPHChQtV3nz27Fm3bt10\nHRdCpgUT9Gu1PeptgrR6VvDgwYM1vp+RkYEX0Qg1BCbo12xtbQ0dAido9azgnDlzGLb+9ddf\nDQ4HIdNlogm6vLz88uXL+/fvv3Pnjnr6uqFDhxo2Ki5ISUkRCATsyzPfD2QYVI4QqpMpJugj\nR44EBQUFBwcPGTKke/fu4eHh//77LwBMnjzZ29vb0NEZ0tOnT7VaNSY5OVmhUDAUiIyMbHBQ\nCJkuk0vQf/zxx7hx4zT/ik9JSRk0aNDdu3dFItHJkydHjDCeJ9e1QpLkd999p9VHjh49yrCV\nx+Nprp6OENKWySXoBQsWVHmHoqjy8vIvv/wSAKysrMaMaUorlOuQSqVatmzZqVOntPoIw9aE\nhIQGB4WQSTOtBF1UVFTbU8t///03/YNJPcZdBUmSWj1jEh4eXtsmDw+Pnj176iIohEyXaSXo\n58+f17ZJ/YiKn58fj8errZhxU6lUd+/eZV++b9++9vb21d8nCIKeVQoh1BCmlaDNzMzqLGNj\nY2PKl34WFtrNTZiYmFglR/P5/BUrVrRq1UqncSFkikwrQZubm9e2iV5I5dChQ+7u7lr1wxoZ\nbR//Kysr69u3r6urq7W1tYuLy6hRo9LS0saOHdtI4SFkUkwrQZeXl9e2iSCIAwcOTJgwwUQe\nfqttSusTJ04UFRWxqUGlUo0dO7Znz567d+/OyckpKip6+vTp/v37r1+/rtNIETJdppWgJRJJ\nbZtIkpw8ebI+gzGg6Ojo2u6F3r59mx7QwuzBgwe+vr7Hjh3TfJOiKJlMFhcXh3NOIaQTppWg\nGR5sk8vlJjJ+o1OnTv/73//Gjx9f41aKohISEpgfP6moqIiIiCguLq6+SaVS5eXlnT59Wjex\nImTaTCtB29nZ8fk1r1FgItnZw8Pj2LFjJEneunWrtjJlZWXPnj1jqOSdd9558eIFQ4HMzMz6\nh4gQqmRaCVokEvXv35++H2iCHB0d6aUFZTLZ5cuXaytGEISlpWWVNwsKCrZs2ZKRkbF69eo6\nb6La2Ng0PFqEkLZLXjV5X3755aVLl6pcIQoEArlcbqiQ9MPR0TElJYX++caNGwx/MXTo0EFz\nsF1iYuKUKVPKyspYNkSSZI8ePRoSKkKIZnLXkl5eXrNnzxaJROp3bG1tW7ZsacCQ9MDKykqd\nnQFALBYzFNa8SXjixIkxY8awz84A8N577zk7O9cjSIRQFSZ3Bb1///558+Zp9nIUFRXVeL/L\naJibmycnJw8bNiw1NbWiosLMzKxZs2YEQdR4ES0Sic6ePevv709fRMfFxWnVVqtWrRYtWqSb\nuBEyeTX/ljYtUqlUKpWyLBwaGvrw4cMqs/yQJMnj8Yyyl0MoFK5cufKjjz5inthIjSRJlUrl\n6Oi4ffv2oKAgrWYf7dSp05EjRzjexW9vb19cXMzxc21paSkSieRyOcsx6YbC4/FsbW3z8/M5\nnkZsbGz4fH5ZWVlpaamhY6mBg4NDbZtM6wq6sLAwPT29+vsqlYrj37D6EQgEaWlpb7zxBsvs\nDJUT1OXn50+YMKH6SoMM2rVrd+HChYKCgvoEihCqCacvdnSO4brJ+BI0n8/Pzs7+8ccflUql\ntp9VqVSPHz/+559/WF4Ot27d+sqVK9rHiBBiYloJ2sHBwc7OrvpTzhz/q7weeDxeTk4OADCM\nd65Tenq6r68vcxmSJGNjY8+fP298xxAhgzOtXyqSJCdOnFj9Ypl9D0BTkZubS/+g7ex0miQS\nyeHDh2ucfJUgiNGjRy9fvjw5OXnFihWYnRFqDCb3ezVz5syRI0fSF9F0WuHz+Ua8zFW9d40g\niHv37v3222+HDh3y8vLS3GRnZ/fzzz+vWrVq3LhxPj4+uggTIVQDkxvFQbt27VpSUlJWVpan\np2dUVJSDg0Pr1q2NZjoOPp9P92/QevTokZqayv7j1UfgDRw4cM6cOVevXs3JyfH19Q0PD6+y\n2KCZmZlEIlGpVNy/SYijOHQIR3HoBI7iqCo4ONjX1zcrK8vR0ZGeb37MmDGbNm0ydFy68dFH\nH2m+/PPPP2NjY0+ePFlbeYlEEhoampSUpL6dWCVHHz16tKKiYseOHY0UMEKoRqaYoJOTk+fN\nm6e+qOzRo8fSpUvT09Nre3ajaREIBPPnz9d8RywW7969u6Cg4K+//pJKpd27dy8pKSkvLy8q\nKrp//35wcHBISAgAFBUVpaWlHT169IcffqhSJ0VRf/75Z2pqqp+fn/72BCGTZ3IJ+syZM8OG\nDdO8K3j27NnIyMjS0lIjyM4EQezfv7/GTXZ2dsOHD6/yZkREhPpna2vrkJCQrVu30s+qVK/h\n+vXrmKAR0ieTS9CLFi1SKpWauVilUpWUlBhHdt64cSN9OdwYjOAQIdS0mNYojpKSkhoncjOC\n1EMQRHh4eHR0dAPradu2bW2DDtu1a9fAyhFCWjGtBG3EkyJRFPX22283vJ7hw4fb29tXGddM\nEES/fv38/f0bXj9CiD0j6eKobQnUKoz4eYqQkJChQ4eyPA4MbGxs9u7d++6776alpanfjIiI\nWLduHUPl6k0ND0A/ME6doMPjeJCamlCoNGMYtyCXywUCAZuSRUVFRrnYB0EQT548cXV11VWF\nCoXi2LFjN27cMDc379q1a+fOnXVVM0JIk1KprPFhXZoxJOiysjL2y0h369btzp07jRqPnhEE\nERERYdhBymKx2MzMjKIo5rUKucDGxubly5fMq+IanEQiEQqFcrn85cuXho6FCUmS1tbWL168\n4HgasbKy4vF4MplMq6Un9IOiKDs7u9q2GkMXB0VR7CdsW7Ro0ciRI8EobgzS/Pz8Vq5cSR+B\n8vLyP/74IzU11crKqnPnzkFBQfqJgb6vqNWJMCCVSsXxONVBFcAaAAAfJElEQVRfTo7HSasy\nLIqD6PCayvdTkzEkaK306dNn+/bt8+fPf/TokaFj0QEPD4+TJ0/SS5WfO3du6tSpWVlZ6q2D\nBw9etWqVRCIxXIAIofoz2ptmDPr16/fvv/8axw1DS0tLOjvn5OTExMQ8efJEc+v+/ftnz55t\noNAQQg1lDEmqHnbu3GkcU4wGBATQP/z6669SqbT6Tu3du1dz4iSEUBNiogn666+/NnQIOkAQ\nxMcff0z/fPPmzRr/JqAo6ubNm/qNCyGkG6aYoKOjo9Xz2XMfQRCZmZmak2ao3//mm2/UMzUT\nBFHbGM8mN/YTIUQzuQT99OnTc+fOGToKtgiCKC8vt7Oz27Zt28mTJ8PDw52cnFxcXPr373/1\n6tWJEyeqS7Zr167GO9QEQbRt21aPISOEdMbkRnGsX7/e0CGwRRBEXl6e+mVgYODu3btrKxwb\nG7thw4bS0tIq3dAxMTHOzs6NGCVCqNGY3BX0gwcPDB0CK1Wyc52cnJwSEhKaN2+uWUNMTMy3\n337bCNEhhPTB5K6gGVaX4Q5tszOtU6dOp0+fPnHiRGpqqo2NTefOnXF6I4SaNJNL0GPHjt22\nbZuho6hDPbIzTSAQDBgwYMCAAbqNByFkECbXxREUFNS+fXtDR8Hk2bNnhg4BIcQJJpegAeDG\njRuGDqFW//77r6FDQAhxhcl1cUDlzD4ctGjRIvW4ZpYUCsX58+dPnDghl8tDQkKsra3Pnz8v\nlUoDAwN79OiB4zcQatJMMUFz08KFCz/44AOtPnL58uXJkyc/fvyYfvnjjz9qbuXxeB9++OG8\nefMYZptFCHEZJmhOCAsLmz59ulYfyczMHDJkCMNE2Eql8vvvv5fL5YsWLWpwgAghAzDFPmiu\nad269Z49e7T91Pr169ksUxAfH8/9SfQRQjUyxQTt7e1t6BBe8/b2/vvvv+vxwYsXL7IpplAo\nrl27Vo/6EUIGZ4oJmjvLCLm6ul6+fLl+n62oqGBZUi6X168JhJBhmVyCTkpKys/PN3QUAACO\njo4NGfDXunVrliX9/Pzq3QpCyIBMLkHPnz/f0CEAAFhbW6ekpDSkBs2p7BgMGDDA09OzIQ0h\nhAzFtBL0kydP0tPTDR0FWFlZ3b9/v4GVhIaGrlixgl7vqjZdunRZvXp1AxtCCBmKaQ2z48LU\nbhKJRFf/ScTGxvbr1y8+Pv7s2bMKhcLX19fa2vrWrVsVFRUtW7YcOnRo//79ddIQQsggTCtB\nHzp0yLABmJubZ2Rk6LBCNze3L774QocVIoS4w7QSNEVRBmzdzMwsMzNTJ1Xdu3cvMTHxyZMn\nHh4ekZGRPj4+OqkWIcQpppWge/TocezYMYM0LRKJHj161PB6KIr6+uuv165dq1QqCYKgKGrJ\nkiUzZsyYO3duwytHCHGKad0kHDFihEHaFYlE6hkzGmjLli2rVq2ilx+k/yBQKBTLly/ftWuX\nTupHCHGHaSXo5ORk/TcqFAp1lZ0BID4+vvoq3SRJxsfH66oJhBBHmFaC1v8jKgKB4MmTJ7qq\nTS6Xp6enV+9JV6lUqampNa7qjRBqukwrQZ88eVKfzQkEguzsbB1WSJJk9ctnGkEQJGlaZxMh\no2dav9LFxcX6bE7nN+54PF67du2qJ2KSJDt27Fhb7kYINVGmlaD13AmwZMkSnV+zz549m6Ko\nKjlapVJNmDBBtw0hhAzOtBK0/n322We6rbB///7r1q2r8oQ3QRAzZsy4cuWKbttCCBkWJujG\nlZWVpfM6JRJJlblGKYqSyWTTpk0z7JM4CCHdwgTduBojYx44cKB6N7RKpbp7925aWprOm0MI\nGQom6MbVGFN95uTk1Jb3dTtoBCFkWJigG9cnn3yi8zrt7e1rG7Dh4OCg8+YQQoaCCbqxkCQ5\ne/bsyMhIHdYpl8uLi4sHDBigUqmqN+fm5tamTRsdNocQMizTmixJbyZMmDBz5kxnZ2ddVXjp\n0qXFixdfuXJFoVA4Ozs3a9YsIyODniwJKh9gWblyJT6rgpAxwQStezweb9myZTqscN++fbGx\nsQRB0BfOeXl5FEX5+PiUlJTk5eUJBILQ0NDFixe3bdtWh40ihAwOE7Tuubi40D88evQoMTEx\nIyPj2bNnubm5L1++bNmy5ezZs319fdnXVlFRMWPGDHV2hsqRIenp6Rs3buzdu7e5uTnzwlcI\noSYKf7F175tvvgGA1atXf/vtt1UGLN+6dWvfvn3jxo1jv/jWxYsXa5zjiSTJP/74Y9CgQQ0P\nGCHETdhlqWM+Pj6RkZH79u376quv5HJ59QIURW3evJn99M3Pnz+vxyaEkBHABK1j6enpp0+f\n3rBhA0mSDE+prFixgmWF7u7uNb5PUZSbm1t9QkQINRGYoHXvn3/+SUlJqT4STlNubi7L2jp2\n7Ojt7V19eAZFUYMHD65niAihpgATtO69fPmyzuFu7KcGJUnyp59+EgqF6o/QlY8bN65Xr14N\nCBMhxHWYoHVPpVJJJBLmMl5eXuwr7Nq16/nz52NiYry9vW1sbLp06bJ169bly5c3LEyEENfh\nKA5dIgiCx+P9/PPPdV5Bf/XVV1rV7O7uvmbNmgaEhhBqevAKWpf4fL5CoQAAhg5oHo8nFotj\nYmJcXFz8/f3PnDmjxwARQk0JJmid6dmzZ1hYWG3XzsOHDx84cKCHh4dSqZTJZCqVSqlUPn/+\nfMiQIbNmzdJzqAihJsEgXRzUuZ0/7Dl9JauE37pNxynTJ7mJePUqwwlCobBr165z584NDg4O\nDw+v7dp57NixOTk5R44cqb5py5YtM2fOrG04HULIZBngCvrm5gXf7Dwubt4pOizw8aXEmR8u\nL1NVHS/MpgxHnD59+rfffgsODgYAGxsbhivoKVOm1FbJp59+2ljxIYSaLH1fQavkeUsOpjh0\n/HDJnL4A8H8h/HGfHFtzs2BeO3utynBEYGBgixYt1C979+59+vTpGktKpVKGeu7fv6/jyBBC\nTZ++r6BLs3dKlVTAqGD6pW3AWHMekZrwUNsyHHH48GHNlxMmTHBycqpHPdbW1jqKCCFkPPR9\nBV2clg0APnYi+iXBk3iL+Jm5dwA6sS9TVla2ceNGdfmOHTsGBQXpJ35NLi4ujo6Omu9IJBJv\nb+9nz55puxThjBkzahs6zePxhEIhj8fRLngaPZ0eQRB1DgA3OIIgxGKxUCg0dCBM6OPJ4/E4\nfjzph6fMzc0NHUgd6I5HgUDAwePJ/MixvhN0RbEcAGz4r5+js+WTDxWFWpWRyWS//vqr+qVI\nJOratWvjxVybnJyc6m8+fvxY2+zs7u4+fPhwhgIczyZqBEGYmZkZOoq6iUQiQ4fACkmSTeJ4\nNokgAYDP53NwYl6lUsmwVd/hCiz5AFCkeJ3CChUqUmilVRkej6c55sHS0pJ5J3WOIIiUlJQa\nG7WxscnNzWWfo7t06ZKUlMQQPz3pUmOsDq5DBEHQFyl6PhH1wOPxVCoVx48nvUQORVHMl1cG\nR5937p90Lh9PlUrF8PexvhO0tZ8bQOrD4gqwFwMARckelyvE3q21KmNlZXXgwAH1S6lUWlj4\nn2vw2qjXiGLWvXt3hudHgoODjx49CgA1NhoWFpaamsomGABYsmTJlClTiouLGcpYW1uXl5fL\nZDKWdRqEmZmZRCJRqVQsT4QB2dvbl5SU1DgTLHdYWlqKRCKFQlFUVGToWJjweDxbW9sXL15w\n/D88GxsbPp8vk8lKS0sNHUsNGNZ61vdNQolrjBlJpCTcpl+WPEwoUVKtRrTQtkz95OXl1VmG\nz+evW7euttVXLS0taxzLrPbhhx86OTmxmQvJxsaGYeAdQgjpO0GTQpcF0X5Pk5eu3LTryO/b\nFy84YO7SZXp7ewDYPHf6e++9J1VRDGUaztramiF7uru7nzp1iv63+hJ/rq6uKSkpzMnX0dHx\n+PHjgwYNYi7Wpk2bGzduaBU5QsjUsPqTX9eos9vX7D1zNauE7x8YEjd9Iv2U4HdjhicVyXbu\nOyDhEbWVqZFUKmUeZVxdWFhYWloaAAQGBj558sTFxWXy5MlvvvmmpaVllZJbtmz54YcfRo8e\n/f7772t1h6GsrOzhw4eurq4VFRWXL19u27ats7PzwYMHAWDIkCHsR2U0rS6OgoICQ8dSB3t7\n++Li4ibRxSGXy5tEF0d+fn6T6OIoKytrcl0cBknQOlaPBC0QCOihxwUFBRy8b6AJE7RuYYLW\nIUzQOsGhPmiEEEIsYYJGCCGOwgSNEEIchQkaIYQ4ChM0QghxFCZohBDiKEzQCCHEUZigEUKI\nozBBI4QQR2GCRgghjsIEjRBCHIUJGiGEOMpEJ0t69uzZ2bNnASAyMpLjCyCJxWKFQqFQKAwd\nCJP09PSbN2+KRKLIyEhDx1IHc3Pz8vJyji8CcvXq1czMTCcnJ4Os5cYeQRDm5uZSqZTjaeTU\nqVNFRUW+vr4BAQGGjqUGDJMlcW6FrnowNzfXdtnK+/fvx8fHA8CQIUPs7XUz07QpO3bsWHx8\nvJ2d3ZgxYwwdS924v8jpmTNnjhw5EhISEh0dbehY6sb9NQn37dt39+7dd955JywszNCxaAe7\nOBBCiKMwQSOEEEdhgkYIIY4yhpuE9aBQKOj7ihYWFiSJ/0s1VEVFhUwmIwii+pphqB7Kysrk\ncjmfz+d+d3mTUFpaqlQqRSIRx0cEVGeiCRohhLgPLx4RQoijMEEjhBBHGcM46LpQ53b+sOf0\nlawSfus2HadMn+Qm4tWrDKLVfawoReGRX388dv5ObpHSs3nL6DFx4W2cDBJrU6DFd0+Wf27S\npGVeM+O/7uGizxCbDlYH8/7fu3cknrtzL8fG03/YpI/CA2z1HyhLxn8FfXPzgm92Hhc37xQd\nFvj4UuLMD5eXqap2u7Mpg2hsjlXCwun/O3jJvUPYiCH9bF7eWb0w7tCDEoNEy33sv3uUSrpu\n3upipUrPETYhbA5m3rkfZ67c/kToPiA63LI0dc2n0/8tqTBItKxQRk1Z8XTE4Ojxi/+kX+Zf\nXx8VFbX02nNtyyAam2NVXnQmKirq/fiblR95Nnno4JjJu/Uda1Og1XfvQvy0qKioqKioj//J\n0WOMTQarg6lSzB3x1uhpOxQqiqIoeWnK4OjoD35K03uwbBn5FXRp9k6pkgoYFUy/tA0Ya84j\nUhMealsG0dgcq7K8UwDQ+s1m9EtS4NBOIlCUPdJvpE0D++9e4e2dS45m9np3mH4DbEpYfTnz\nD9yRygOmRPAIAAC+uf/u3buXj/XRf7QsGXmCLk7LBgAfu1eDHwmexFvEL8u9o20ZRGNzrKzf\n+OTgwYNT3S3olwW3j554UW7btpd+I20aWH73lLIHiz//zbnr1Pe7YtdzrdgczJIHVwDA7tnJ\nT2dPi3k7ZurMhX/eyhdz+IaTkSfoimI5ANjwCfU7tnxSpSjUtgyiaXWsKGVJ0p71cQt/JC0D\n5k0L0lOITQq740klfP75E2HA17P66je6JobNwSx7UgYAR1cniLwCB70V6UY9+unL9/fcL9Jz\nqOwZ+SgOgSUfAIoUr28UFCpUpNBK2zKIxv5Y5d08vmLlxrTCcr/uQz+YOtpTzN2LFANiczwz\njyzZlVIat26eA5+U6TvApoTNwSR4BAC0mvjtJ1HNAIAaMeTDkbF7lh5+e+NovcbKmpEnaGs/\nN4DUh8UVYC8GAIqSPS5XiL1ba1sG0Vgeq8dJ8VO/P2bhHbrw07gQH+6OYTI4Nsfz9sEUiqrY\nMHXUhsp3bi2fEr0c9hw4KCSq1WjC2BxMCx9rAPAPeTXok+BZBVuIDpTe1n+0LBl5F4fENcaM\nJFISXp2AkocJJUqq1YgW2pZBNDbHSil7OHv1MefuEzaumo/ZmRmb49ly0LDRlWKGhgKAU/dB\no0ePxj9JqmBzMK2av0USxI2kx/RLlfxZcnG5mUMXfcfKGm/x4sWGjqERETyLltJrh//any0l\nCjOub16zX2bfeemk/gKC2Dx3+k+/Hwj/v4FCvmVtZQwdPuewOZ4Fycv3nc1zbWGfee3KpUrX\nbr/s2L65ocPnHDbH07mlf5tKvt6y3/ZfbD5i1kf/F0ri1/O/2BxMkdDJNefM0cMH7uW/fPrg\n1p6f1t4rIj9YMbOZRGDo8Gtm5F0cANBuwtK5ojV7zySeL+H7B0V8N32iGUkAQGFO9pMiGT1V\nVG1lUHV1Hs+CS/kAcC/pr3sanxLbKieP622gkDmNzfcTscTmYIbNWKNyWHPo3NmEArmXT9u5\nM9/v5sTdFWFwNjuEEOIoI++DRgihpgsTNEIIcRQmaIQQ4ihM0AghxFGYoBFCiKMwQSOEEEdh\ngkYIIY7CBI2QkTv3XmtL92mGjgLVByZopBu5FwYSBNFrd7qhA2kUFcXJBEE4+O/UVYX6OVwP\n/vrpnaiwIb/cK8390a9T7xnfHZDr/bm0tx0lfJGbvls1FpigUZNBEARBEOuySwHg34/bEwQx\n60FR9U2IdmfTKN/+cYlZ4jA/a6FFx14+0lWzBgcM32zouJAWMEEjZIRUFdn9piZI3MdlXkmc\nHuoksOgUv/vCtkHN7u2ZsPbJS0NHh9gy/smSkNEoLCwEAHNriVabTJM0b8eTcmXrQVMkGtN+\nRX4d3Sr1j0s3C6FyQbLGU154/8LV7DY9ezR2Q8YNr6CbhvL8q3PeGdy+padYYtuqfe9FGw7S\nC0dkn5xHEIRH35Xqkg/3TCQIwn9iAv3yRerR2MH9/Lwcze09u4S/uWzvdXXJSa4WZrZ9ZM8u\nvv92Tycbc4m9e9eo9048Ynt5defg8oG9gl2szJu3CR0zbXmWTMkm4DrbVcnzfv7q/Q6tfazF\nYmevlgPGLbz07NVCIrP9PezsXYUEvGlv3mnpdQD4zsdGbN1Nc1OdrTPUz8aFbZ9H9Q5xklj4\nBHYeM2vDC431OxgarfNw1Xaa5C+vEgThO/KfqwnL3o7o5mRh2TKox+RPt5bXtTuk0BUA8i9f\n02zFrvXq1NTULQM8mRud7W1NEMTqx5rfBGUPGzHJE18oqWDe00muFub2bz48+LmXi19Yn7BT\nL6oe2yt71w3uFuThaC2U2Lbwax+3aH2BQsVmT02UoZcVR3UrTN38hkRA8iR9BsW8O2FU5+aW\nAOA3dJmKoiiK+mVECwCYsj+DoqiK4stvmPHNHAfkVigpispKXGjDJ0mBXd/okZNHv9XGXgwA\nvT78na52oouEJ3Ttbm8z/qv1f/yTtGPtHAseaWbf+1mFss6QkhZHAwBf5NF/yDsjo8IdBDyx\noxgAwnbdrzNgxnZVn/f3AACPgNBRE8eHd2oFABbug5/LlfQHCVJMUdQfWzZ/MsADAPp9s+bn\nXw9qbqqrdab6GZQXnQUAvrg5AHTqEzU+dliHFpYAYBcY80yurHOXmQ8Xw2mqKLkCACLbAJJv\nO/WzFbt2bvp4Sh8A8AyfWa5i3h3lMDcJQRBvzlj5w1vNLNw+qLJHDI3e/aUXAAR8kKwu/OL+\nFwDg2H4Vm5NL8q0cBDzX4P4T3p+ZUiof6mDOE7rS9WTsm04QhMimZUzspEmjBrVrYQMAzYf8\nxGJPTRQmaO5TvuNqQfKtfrnxnH6tUhR/2dsNACYnZVMUpZA97GYlEpi3vv6yYnm4O0EKf7hd\nQFGUSvGig4WQ5El2Xs+nP1jxMu0tJ3MA+DGrhKKoiS4SAOiy7Iq6pX19PQBgYUYRc0Bl+UeF\nJCGQBJzMekm/k39jtw2frMw4dQTM0K40bxcAWLdYrP6V/F9nFwB4924B9d8sfHl+OwCYmf6C\nfqmxial15voZ0AkaAKI3nH9VrTz/s95uAND1u5vMjTIfLubTRKctAHjveJY6mD/ntAWAoQcy\nmHenJPNIZIdXF8skzyJ8+Hu/Hj4nVdLhMTb68oqIJMS2Eepqj4/wAYAJZ3NYntygOQmKys9q\nJuhp7pYEQR59Kq38bEkXKxHJs6QqE3Rte8p8dowYJmiuK83dCADOnbZqvlnyeDUAePQ+Qr98\neu4LgiDsO3YCgJD5J+g3C+/OBACPPjs1P5hxcCAAtJlxgar8XbpSUqHeemleWwB4/14hc0i3\n14QCQKdvrmm+uX9wMzrj1BkwQ7vSZ78BgJn9oDN38uhNKnl5WVkZfU3PJkEzt85cPwM6QQst\nO0qVry/nSnN3AYClxwzmRpkPF/NpenVdad1dobG1ouQyANi1XMNmdx5ePDK5vT3Js+ATBADY\nBw2/X6ao87uxzN8OADZkv6QoSqWUBkgEAklAiULF8uT+VShTb9VM0HdvXL92PUW9SaUo6W9L\nrx9IMe9pjSfFFGAfNNeVZh8CgKeXxxAaLD0+AoDSrId0GafQTxPG+eb/e1niOvLEV68WLim4\ndgUAvIa116zNITgKAHKPZ9IvCULQ3uL1Yj8Eu1W+so9mA0C34c0032w/0Yd9wLW1a+bw9o6p\n/SoKDnb3d2rRvsfY9+b8eji5jC8UsP6eMrfewPolzuM1l9oxdx7hJeLLChKZG2U+XGxOk5nD\n25orEAosOvqY8Utzd7HZnWbB/zc+1MnceVzho6sLh7XMv5rwf+//XWejMcu7A8D6lSkAUJCy\n4Hap3Gfkegsewe7kkqFWwhoPoG9gW2fq+Y5N6xZMfzeqb6izpe3xwv90Ute2pzXWZgpwFAfn\nkSQAuPdfuHhYsypbhBYdKn9U3UorBoDyF5dSpfJOljX/egAAQfABgFLJK1/z6rG0F8knodr9\nZVKdFdgEXHu7I9cdD598cuf+w3+eSDq46fut8Ssk7qE7L/4V5cZuhEZdrTes/qpRUwAEKWZu\nlNzOeLhqbKbKaaIqqhSQqShKJWPYna7Zx/ddy+8QE9uh8j9CC492X+w4u+p35yfHdkBkHY26\n911jwz98/5clsGL/37P3AsCCrzoBsDy5QkktK8Zt+yh87NokUmjXtV9Ez94jxs1cumf8wF15\n0tclat9TE2XoS3hUh9LcnwHAtct2zTcrSm4cO3bsTOqrzuLb/xsKAO3fHQ0Abr2W0m8WpE0H\nAM9+CZoffHQ0GgD845Kp//YY0Oh+gzq7OO782BUAgpff0HwzcdQb8KqLo46AGdp9mfXvqVOn\nrlb2USpkT/euHgoAzsE7KLZdHEytM9fP4FUXh0WH0v90cewEAJsW3zA3yny4mE8T/Ye/QNKm\nWKHRbs52ALDy/oxhd25+FwIAI248oygqOc6/8iah0lHAs3CNq/O7QVHUxhBnAPg5K8tTxLf0\nmFq5y1qfXHUXR2nuZgCw9ZtTqNEFE+ssAY0ujtr2tJbTYvwwQXOfcpiTOUGKf7n+vPIdxaqB\nXgAw7VIeRVHSp0edhDyr5mOlStWWES0AYPxvDyiKUskL21kISZ7l3juvEq689N5bzuYAsPZh\nHYmSOSBZ4Z8ikhBK2p7OKaXfKbp/2EnIg8qbhMwBM7SbdXwgAHhG/KrelH/rcwBwCtpO1ZSg\nZ9R0k5Chdeb6GahvEkau/od+RyUv/KynKwC8tfchc6PMh4v5NKlvnfX//u9X7SpeLOrlBgD9\nNqUx7E7hvYUA4DVwg1IjQWf/PRsA/N89Xed3g6KonLOxAOASHggA/Xfcr2xB65OrTtAFdycD\nQLPoRPWmp+fXCUkCAJQaNwlr3FPms2PEMEE3Ac+vrfMQ8UmexYCh70z74N3ITh4A4Dd6rZKi\nKJXsPX9bkifZllFMUZS87F4HCyHfzOdicTlFUZmH51nxSZ7QaeCw2PfGDW/naAYAPae+um6q\nd4KmKOrvzwcRBME3a/bm8PHjhv2fm4hn7u5QmaAZA2ZsVyF72NveDAA6RcZ88NG0YZE93cV8\nkm+7/np+lQ/eWB4MAG/EfPjF4nVVNjG0zlw/AzpBi+xcAKB976jJE0d19rUGgOYRH1eo6jpH\ndR0uhtNEpy2eyJ0giHa9oyZPGh3ayhoAnDpPKVGoGHdHuaCHCwD4vzlxek9XsW3fr2ZPsBOQ\nYruQ9DJFnd8NiqKU8nxPER8ASIHdg7LXt+60PbnqBK2QZYRYiUie5ZsjJn66eMGkmD4SvqCl\nlRAABoyNO/XkPMOe1vmFNFaYoJuG0uzkqcMiA5q7CM1t/dp3+2TDoTIlRVHUuSW9AKDLZ6fV\nJbMSpwGAW9g39MuCWwdHv9nb18NObOsWHDZg6Z7XYwkakqApirpzYHlkWLCLldjDr9PIuK/u\nPN4bERExJymbOeA6232ZdXre2P6+Xi5ivtDJwyfsrff2Xcit/kFZ4Zk3OzQT8YUubwyuXidD\n6wz1M5CX3o6IiJiyIeXomjkR3drZiC1adegZ91WCXCN1MDRa5+Gq7TTRCdqu5c+3fvuib5e2\ndmbmLQJDY+dtfFnZ08KwOypFya+LxrV7w4W+B2vh8sbAdxbeKXo9eIbhu0H7PcILqg32YN5T\nhgRNUVTR3aMTI0Pc7MwtnZv3jBy29czjonu/x0Z3b98l/FjWOeY9NU0ERZn4kzroNUpZ9vhB\nutzTv4WYV3fppt8u98lfXhVadrBr+XN+2vh6V3Jmsn/kkbCS7HgdBqZzOtlT44OjONBrBM/M\n07eN6bRrInh8AghB3eUQ9+A4aFTVxRmBBCOSZ2boGHXMuHe5y4aUkidrDR0Fqg+8gkZVdVp6\noeRLlaGj0CsT3GXUJGAfNEIIcRR2cSCEEEdhgkYIIY7CBI0QQhyFCRohhDgKEzRCCHEUJmiE\nEOIoTNAIIcRRmKARQoijMEEjhBBH/T/e/mav6/DtrQAAAABJRU5ErkJggg==",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "options(repr.plot.width=4, repr.plot.height=4)\n",
    "qplot(x=exon_densities_bedops$overlap, y=exon_densities$overlap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "\tPearson's product-moment correlation\n",
       "\n",
       "data:  exon_densities_bedops$overlap and exon_densities$overlap\n",
       "t = 2763.2, df = 484010, p-value < 2.2e-16\n",
       "alternative hypothesis: true correlation is not equal to 0\n",
       "95 percent confidence interval:\n",
       " 0.9695676 0.9699034\n",
       "sample estimates:\n",
       "     cor \n",
       "0.969736 \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cor.test(exon_densities_bedops$overlap, exon_densities$overlap)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting coordinates of regulatory features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It took forever to find out, how to get the regulatory features. In the end I found it totally by accident in some bugreport mail :( https://support.bioconductor.org/p/39545/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*TODO*: Figure out how to access older snapshot of the database.\n",
    "\n",
    "http://grch37.ensembl.org"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>biomart</th><th scope=col>version</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>ENSEMBL_MART_ENSEMBL </td><td>Ensembl Genes 89     </td></tr>\n",
       "\t<tr><td>ENSEMBL_MART_MOUSE   </td><td>Mouse strains 89     </td></tr>\n",
       "\t<tr><td>ENSEMBL_MART_SNP     </td><td>Ensembl Variation 89 </td></tr>\n",
       "\t<tr><td>ENSEMBL_MART_FUNCGEN </td><td>Ensembl Regulation 89</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ll}\n",
       " biomart & version\\\\\n",
       "\\hline\n",
       "\t ENSEMBL\\_MART\\_ENSEMBL & Ensembl Genes 89        \\\\\n",
       "\t ENSEMBL\\_MART\\_MOUSE   & Mouse strains 89        \\\\\n",
       "\t ENSEMBL\\_MART\\_SNP     & Ensembl Variation 89    \\\\\n",
       "\t ENSEMBL\\_MART\\_FUNCGEN & Ensembl Regulation 89   \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "biomart | version | \n",
       "|---|---|---|---|\n",
       "| ENSEMBL_MART_ENSEMBL  | Ensembl Genes 89      | \n",
       "| ENSEMBL_MART_MOUSE    | Mouse strains 89      | \n",
       "| ENSEMBL_MART_SNP      | Ensembl Variation 89  | \n",
       "| ENSEMBL_MART_FUNCGEN  | Ensembl Regulation 89 | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  biomart              version              \n",
       "1 ENSEMBL_MART_ENSEMBL Ensembl Genes 89     \n",
       "2 ENSEMBL_MART_MOUSE   Mouse strains 89     \n",
       "3 ENSEMBL_MART_SNP     Ensembl Variation 89 \n",
       "4 ENSEMBL_MART_FUNCGEN Ensembl Regulation 89"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "listMarts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>dataset</th><th scope=col>description</th><th scope=col>version</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>hsapiens_regulatory_feature                </td><td>Human Regulatory Features (GRCh38.p10)     </td><td>GRCh38.p10                                 </td></tr>\n",
       "\t<tr><td>hsapiens_motif_feature                     </td><td>Human Binding Motifs (GRCh38.p10)          </td><td>GRCh38.p10                                 </td></tr>\n",
       "\t<tr><td>hsapiens_external_feature                  </td><td>Human Other Regulatory Regions (GRCh38.p10)</td><td>GRCh38.p10                                 </td></tr>\n",
       "\t<tr><td>hsapiens_annotated_feature                 </td><td>Human Regulatory Evidence (GRCh38.p10)     </td><td>GRCh38.p10                                 </td></tr>\n",
       "\t<tr><td>hsapiens_mirna_target_feature              </td><td>Human miRNA Target Regions (GRCh38.p10)    </td><td>GRCh38.p10                                 </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lll}\n",
       " dataset & description & version\\\\\n",
       "\\hline\n",
       "\t hsapiens\\_regulatory\\_feature             & Human Regulatory Features (GRCh38.p10)      & GRCh38.p10                                 \\\\\n",
       "\t hsapiens\\_motif\\_feature                  & Human Binding Motifs (GRCh38.p10)           & GRCh38.p10                                 \\\\\n",
       "\t hsapiens\\_external\\_feature               & Human Other Regulatory Regions (GRCh38.p10) & GRCh38.p10                                 \\\\\n",
       "\t hsapiens\\_annotated\\_feature              & Human Regulatory Evidence (GRCh38.p10)      & GRCh38.p10                                 \\\\\n",
       "\t hsapiens\\_mirna\\_target\\_feature         & Human miRNA Target Regions (GRCh38.p10)     & GRCh38.p10                                 \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "dataset | description | version | \n",
       "|---|---|---|---|---|\n",
       "| hsapiens_regulatory_feature                 | Human Regulatory Features (GRCh38.p10)      | GRCh38.p10                                  | \n",
       "| hsapiens_motif_feature                      | Human Binding Motifs (GRCh38.p10)           | GRCh38.p10                                  | \n",
       "| hsapiens_external_feature                   | Human Other Regulatory Regions (GRCh38.p10) | GRCh38.p10                                  | \n",
       "| hsapiens_annotated_feature                  | Human Regulatory Evidence (GRCh38.p10)      | GRCh38.p10                                  | \n",
       "| hsapiens_mirna_target_feature               | Human miRNA Target Regions (GRCh38.p10)     | GRCh38.p10                                  | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  dataset                       description                                \n",
       "1 hsapiens_regulatory_feature   Human Regulatory Features (GRCh38.p10)     \n",
       "2 hsapiens_motif_feature        Human Binding Motifs (GRCh38.p10)          \n",
       "3 hsapiens_external_feature     Human Other Regulatory Regions (GRCh38.p10)\n",
       "4 hsapiens_annotated_feature    Human Regulatory Evidence (GRCh38.p10)     \n",
       "5 hsapiens_mirna_target_feature Human miRNA Target Regions (GRCh38.p10)    \n",
       "  version   \n",
       "1 GRCh38.p10\n",
       "2 GRCh38.p10\n",
       "3 GRCh38.p10\n",
       "4 GRCh38.p10\n",
       "5 GRCh38.p10"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ensembl_mart_funcgen <- useMart(\"ENSEMBL_MART_FUNCGEN\")\n",
    "listDatasets(ensembl_mart_funcgen) %>% filter(str_detect(dataset, \"sapiens\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "reg_data <- useDataset(dataset=\"hsapiens_regulatory_feature\", mart=ensembl_mart_funcgen)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>name</th><th scope=col>description</th><th scope=col>page</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>activity                </td><td>Activity                </td><td>regulatory_feature      </td></tr>\n",
       "\t<tr><td>regulatory_stable_id    </td><td>Regulatory stable ID    </td><td>regulatory_feature      </td></tr>\n",
       "\t<tr><td>bound_seq_region_start  </td><td>Bound start (bp)        </td><td>regulatory_feature      </td></tr>\n",
       "\t<tr><td>bound_seq_region_end    </td><td>Bound end (bp)          </td><td>regulatory_feature      </td></tr>\n",
       "\t<tr><td>chromosome_name         </td><td>Chromosome/scaffold name</td><td>regulatory_feature      </td></tr>\n",
       "\t<tr><td>chromosome_start        </td><td>Start (bp)              </td><td>regulatory_feature      </td></tr>\n",
       "\t<tr><td>chromosome_end          </td><td>End (bp)                </td><td>regulatory_feature      </td></tr>\n",
       "\t<tr><td>feature_type_name       </td><td>Feature type            </td><td>regulatory_feature      </td></tr>\n",
       "\t<tr><td>feature_type_description</td><td>Feature type description</td><td>regulatory_feature      </td></tr>\n",
       "\t<tr><td>epigenome_name          </td><td>Epigenome name          </td><td>regulatory_feature      </td></tr>\n",
       "\t<tr><td>epigenome_description   </td><td>Epigenome description   </td><td>regulatory_feature      </td></tr>\n",
       "\t<tr><td>so_accession            </td><td>SO term accession       </td><td>regulatory_feature      </td></tr>\n",
       "\t<tr><td>so_name                 </td><td>SO term name            </td><td>regulatory_feature      </td></tr>\n",
       "\t<tr><td>efo_id                  </td><td>EFO term accession      </td><td>regulatory_feature      </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lll}\n",
       " name & description & page\\\\\n",
       "\\hline\n",
       "\t activity                 & Activity                 & regulatory\\_feature    \\\\\n",
       "\t regulatory\\_stable\\_id     & Regulatory stable ID         & regulatory\\_feature        \\\\\n",
       "\t bound\\_seq\\_region\\_start   & Bound start (bp)               & regulatory\\_feature          \\\\\n",
       "\t bound\\_seq\\_region\\_end     & Bound end (bp)                 & regulatory\\_feature          \\\\\n",
       "\t chromosome\\_name          & Chromosome/scaffold name   & regulatory\\_feature      \\\\\n",
       "\t chromosome\\_start         & Start (bp)                 & regulatory\\_feature      \\\\\n",
       "\t chromosome\\_end           & End (bp)                   & regulatory\\_feature      \\\\\n",
       "\t feature\\_type\\_name        & Feature type                 & regulatory\\_feature        \\\\\n",
       "\t feature\\_type\\_description & Feature type description     & regulatory\\_feature        \\\\\n",
       "\t epigenome\\_name           & Epigenome name             & regulatory\\_feature      \\\\\n",
       "\t epigenome\\_description    & Epigenome description      & regulatory\\_feature      \\\\\n",
       "\t so\\_accession             & SO term accession          & regulatory\\_feature      \\\\\n",
       "\t so\\_name                  & SO term name               & regulatory\\_feature      \\\\\n",
       "\t efo\\_id                   & EFO term accession         & regulatory\\_feature      \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "name | description | page | \n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| activity                 | Activity                 | regulatory_feature       | \n",
       "| regulatory_stable_id     | Regulatory stable ID     | regulatory_feature       | \n",
       "| bound_seq_region_start   | Bound start (bp)         | regulatory_feature       | \n",
       "| bound_seq_region_end     | Bound end (bp)           | regulatory_feature       | \n",
       "| chromosome_name          | Chromosome/scaffold name | regulatory_feature       | \n",
       "| chromosome_start         | Start (bp)               | regulatory_feature       | \n",
       "| chromosome_end           | End (bp)                 | regulatory_feature       | \n",
       "| feature_type_name        | Feature type             | regulatory_feature       | \n",
       "| feature_type_description | Feature type description | regulatory_feature       | \n",
       "| epigenome_name           | Epigenome name           | regulatory_feature       | \n",
       "| epigenome_description    | Epigenome description    | regulatory_feature       | \n",
       "| so_accession             | SO term accession        | regulatory_feature       | \n",
       "| so_name                  | SO term name             | regulatory_feature       | \n",
       "| efo_id                   | EFO term accession       | regulatory_feature       | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "   name                     description              page              \n",
       "1  activity                 Activity                 regulatory_feature\n",
       "2  regulatory_stable_id     Regulatory stable ID     regulatory_feature\n",
       "3  bound_seq_region_start   Bound start (bp)         regulatory_feature\n",
       "4  bound_seq_region_end     Bound end (bp)           regulatory_feature\n",
       "5  chromosome_name          Chromosome/scaffold name regulatory_feature\n",
       "6  chromosome_start         Start (bp)               regulatory_feature\n",
       "7  chromosome_end           End (bp)                 regulatory_feature\n",
       "8  feature_type_name        Feature type             regulatory_feature\n",
       "9  feature_type_description Feature type description regulatory_feature\n",
       "10 epigenome_name           Epigenome name           regulatory_feature\n",
       "11 epigenome_description    Epigenome description    regulatory_feature\n",
       "12 so_accession             SO term accession        regulatory_feature\n",
       "13 so_name                  SO term name             regulatory_feature\n",
       "14 efo_id                   EFO term accession       regulatory_feature"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>name</th><th scope=col>description</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>chromosome_name                                      </td><td>Chromosome Name                                      </td></tr>\n",
       "\t<tr><td>start                                                </td><td>Start (bp)                                           </td></tr>\n",
       "\t<tr><td>end                                                  </td><td>End (bp)                                             </td></tr>\n",
       "\t<tr><td>chromosomal_region                                   </td><td>e.g 1:100:10000000, 1:100000:200000                  </td></tr>\n",
       "\t<tr><td>band_start                                           </td><td>Band Start                                           </td></tr>\n",
       "\t<tr><td>band_end                                             </td><td>Band End                                             </td></tr>\n",
       "\t<tr><td>marker_start                                         </td><td>Marker Start                                         </td></tr>\n",
       "\t<tr><td>marker_end                                           </td><td>Marker End                                           </td></tr>\n",
       "\t<tr><td>encode_region                                        </td><td>Encode region                                        </td></tr>\n",
       "\t<tr><td>regulatory_stable_id                                 </td><td>Regulatory stable ID (e.g. ENSR00000060894) [Max 500]</td></tr>\n",
       "\t<tr><td>regulatory_feature_type_name                         </td><td>Feature type name                                    </td></tr>\n",
       "\t<tr><td>epigenome_name                                       </td><td>Epigenome name                                       </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ll}\n",
       " name & description\\\\\n",
       "\\hline\n",
       "\t chromosome\\_name                                     & Chromosome Name                                      \\\\\n",
       "\t start                                                 & Start (bp)                                           \\\\\n",
       "\t end                                                   & End (bp)                                             \\\\\n",
       "\t chromosomal\\_region                                  & e.g 1:100:10000000, 1:100000:200000                  \\\\\n",
       "\t band\\_start                                          & Band Start                                           \\\\\n",
       "\t band\\_end                                            & Band End                                             \\\\\n",
       "\t marker\\_start                                        & Marker Start                                         \\\\\n",
       "\t marker\\_end                                          & Marker End                                           \\\\\n",
       "\t encode\\_region                                       & Encode region                                        \\\\\n",
       "\t regulatory\\_stable\\_id                                  & Regulatory stable ID (e.g. ENSR00000060894) {[}Max 500{]}\\\\\n",
       "\t regulatory\\_feature\\_type\\_name                    & Feature type name                                    \\\\\n",
       "\t epigenome\\_name                                      & Epigenome name                                       \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "name | description | \n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| chromosome_name                                       | Chromosome Name                                       | \n",
       "| start                                                 | Start (bp)                                            | \n",
       "| end                                                   | End (bp)                                              | \n",
       "| chromosomal_region                                    | e.g 1:100:10000000, 1:100000:200000                   | \n",
       "| band_start                                            | Band Start                                            | \n",
       "| band_end                                              | Band End                                              | \n",
       "| marker_start                                          | Marker Start                                          | \n",
       "| marker_end                                            | Marker End                                            | \n",
       "| encode_region                                         | Encode region                                         | \n",
       "| regulatory_stable_id                                  | Regulatory stable ID (e.g. ENSR00000060894) [Max 500] | \n",
       "| regulatory_feature_type_name                          | Feature type name                                     | \n",
       "| epigenome_name                                        | Epigenome name                                        | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "   name                        \n",
       "1  chromosome_name             \n",
       "2  start                       \n",
       "3  end                         \n",
       "4  chromosomal_region          \n",
       "5  band_start                  \n",
       "6  band_end                    \n",
       "7  marker_start                \n",
       "8  marker_end                  \n",
       "9  encode_region               \n",
       "10 regulatory_stable_id        \n",
       "11 regulatory_feature_type_name\n",
       "12 epigenome_name              \n",
       "   description                                          \n",
       "1  Chromosome Name                                      \n",
       "2  Start (bp)                                           \n",
       "3  End (bp)                                             \n",
       "4  e.g 1:100:10000000, 1:100000:200000                  \n",
       "5  Band Start                                           \n",
       "6  Band End                                             \n",
       "7  Marker Start                                         \n",
       "8  Marker End                                           \n",
       "9  Encode region                                        \n",
       "10 Regulatory stable ID (e.g. ENSR00000060894) [Max 500]\n",
       "11 Feature type name                                    \n",
       "12 Epigenome name                                       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "listAttributes(reg_data)\n",
    "listFilters(reg_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download the dataframe from Biomart:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "regulatory_features_df <-\n",
    "    getBM(attributes=c(\"chromosome_name\", \"chromosome_start\", \"chromosome_end\", \"feature_type_name\"),\n",
    "          filters=\"chromosome_name\",\n",
    "          values=1:22,\n",
    "          mart=reg_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "329554"
      ],
      "text/latex": [
       "329554"
      ],
      "text/markdown": [
       "329554"
      ],
      "text/plain": [
       "[1] 329554"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(regulatory_features_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert it into a GRanges object:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in valid.GenomicRanges.seqinfo(x, suggest.trim = TRUE):\n",
      "“GRanges object contains 1139 out-of-bound ranges located on sequences\n",
      "  chr3, chr5, chr7, chr11, chr17, chr18, and chr20. Note that only ranges\n",
      "  located on a non-circular sequence whose length is not NA can be\n",
      "  considered out-of-bound (use seqlengths() and isCircular() to get the\n",
      "  lengths and circularity flags of the underlying sequences). You can use\n",
      "  trim() to trim these ranges. See ?`trim,GenomicRanges-method` for more\n",
      "  information.”"
     ]
    }
   ],
   "source": [
    "regulatory_features_gr <-\n",
    "    dplyr::select(regulatory_features_df,\n",
    "                  chrom=chromosome_name, start=chromosome_start, end=chromosome_end,\n",
    "                  feature=feature_type_name) %>%\n",
    "    arrange(chrom, start) %>% \n",
    "    mutate(chrom=paste0(\"chr\", chrom)) %>%\n",
    "    makeGRangesFromDataFrame(keep.extra.columns=TRUE)\n",
    "seqinfo(regulatory_features_gr) <- seqinfo(BSgenome.Hsapiens.UCSC.hg19)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "329554"
      ],
      "text/latex": [
       "329554"
      ],
      "text/markdown": [
       "329554"
      ],
      "text/plain": [
       "[1] 329554"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(regulatory_features_gr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 6 ranges and 1 metadata column:\n",
       "      seqnames         ranges strand |                  feature\n",
       "         <Rle>      <IRanges>  <Rle> |              <character>\n",
       "  [1]     chr1 [13201, 13800]      * |                 Enhancer\n",
       "  [2]     chr1 [13915, 14583]      * |           Open chromatin\n",
       "  [3]     chr1 [15460, 15817]      * |          TF binding site\n",
       "  [4]     chr1 [19755, 20635]      * |          TF binding site\n",
       "  [5]     chr1 [21062, 22000]      * | Promoter Flanking Region\n",
       "  [6]     chr1 [28200, 30001]      * |                 Promoter\n",
       "  -------\n",
       "  seqinfo: 93 sequences (1 circular) from hg19 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(regulatory_features_gr)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
