{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating SLiM exome recombination maps\n",
    "\n",
    "This notebook generates three sets of recombination map files for SLiM.\n",
    "\n",
    "1. Recombination rates between individual sites from the archaic admixture array.\n",
    "\n",
    "2. Recombination rates between exonic regions.\n",
    "\n",
    "3. Recombination rates between exonic regions and sites from the archaic admixture array."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from pybedtools import BedTool\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "plt.style.use('ggplot')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "RECOMB_RATE = 1e-8 # crossovers per bp per generation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exon coordinates processing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First download the GTF annotations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gtf = pd.read_table('ftp://ftp.ensembl.org/pub/release-75/gtf/homo_sapiens/Homo_sapiens.GRCh37.75.gtf.gz',\n",
    "                    header=None, sep='\\t', skipinitialspace=True, skiprows=5, compression='gzip',\n",
    "                    names=['chrom', 'source', 'feature', 'start', 'end',\n",
    "                           'score', 'strand', 'frame', 'attribute'], low_memory=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How does the data look like?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14412</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972\"; transcript_id \"ENST...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972\"; transcript_id \"ENST...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972\"; transcript_id \"ENST...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972\"; transcript_id \"ENST...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom                source     feature  start    end score strand frame  \\\n",
       "0     1            pseudogene        gene  11869  14412     .      +     .   \n",
       "1     1  processed_transcript  transcript  11869  14409     .      +     .   \n",
       "2     1  processed_transcript        exon  11869  12227     .      +     .   \n",
       "3     1  processed_transcript        exon  12613  12721     .      +     .   \n",
       "4     1  processed_transcript        exon  13221  14409     .      +     .   \n",
       "\n",
       "                                           attribute  \n",
       "0  gene_id \"ENSG00000223972\"; gene_name \"DDX11L1\"...  \n",
       "1  gene_id \"ENSG00000223972\"; transcript_id \"ENST...  \n",
       "2  gene_id \"ENSG00000223972\"; transcript_id \"ENST...  \n",
       "3  gene_id \"ENSG00000223972\"; transcript_id \"ENST...  \n",
       "4  gene_id \"ENSG00000223972\"; transcript_id \"ENST...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "protein_coding                        2001054\n",
       "nonsense_mediated_decay                293471\n",
       "processed_transcript                   173401\n",
       "retained_intron                        150034\n",
       "lincRNA                                 55928\n",
       "antisense                               45811\n",
       "processed_pseudogene                    24517\n",
       "pseudogene                              17830\n",
       "unprocessed_pseudogene                  13763\n",
       "miRNA                                   10209\n",
       "transcribed_unprocessed_pseudogene       7931\n",
       "misc_RNA                                 6554\n",
       "snRNA                                    6215\n",
       "snoRNA                                   4791\n",
       "sense_intronic                           3241\n",
       "polymorphic_pseudogene                   1906\n",
       "rRNA                                     1706\n",
       "unitary_pseudogene                       1467\n",
       "sense_overlapping                        1463\n",
       "IG_V_gene                                1277\n",
       "transcribed_processed_pseudogene         1182\n",
       "non_stop_decay                           1132\n",
       "TR_V_gene                                1019\n",
       "IG_V_pseudogene                           912\n",
       "TR_J_gene                                 328\n",
       "IG_C_gene                                 313\n",
       "IG_D_gene                                 260\n",
       "TR_V_pseudogene                           147\n",
       "3prime_overlapping_ncrna                  116\n",
       "IG_J_gene                                 102\n",
       "Mt_tRNA                                    66\n",
       "TR_C_gene                                  64\n",
       "IG_C_pseudogene                            51\n",
       "IG_J_pseudogene                            18\n",
       "TR_J_pseudogene                            12\n",
       "TR_D_gene                                  12\n",
       "Mt_rRNA                                     6\n",
       "translated_processed_pseudogene             3\n",
       "Name: source, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.source.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "exon              1306656\n",
       "CDS                791856\n",
       "UTR                304070\n",
       "transcript         215170\n",
       "stop_codon          73411\n",
       "start_codon         73358\n",
       "gene                63677\n",
       "Selenocysteine        114\n",
       "Name: feature, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.feature.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Take only autosomal protein-coding regions, since we won't be simulating sex chromosomes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "AUTOSOMES = [str(i + 1) for i in range(22)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "exons = gtf[gtf.chrom.isin(AUTOSOMES) &\n",
    "           (gtf.source == \"protein_coding\") &\n",
    "           (gtf.feature == \"exon\")]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge the overlapping exons:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "exons = BedTool.from_dataframe(exons).sort().merge().to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "exons['type'] = 'exon'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert chromosome IDs to integers and sort exons by chromosome:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "exons.chrom = exons.chrom.astype(int)\n",
    "exons.sort_values(by=['chrom', 'start'], inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What is the distribution of exon lengths?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    214193.000000\n",
       "mean        337.708039\n",
       "std         776.506070\n",
       "min           1.000000\n",
       "25%          96.000000\n",
       "50%         138.000000\n",
       "75%         217.000000\n",
       "max       24409.000000\n",
       "dtype: float64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "((exons.end - exons.start)).describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZsAAAEECAYAAAD+qJluAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3X9sW+d97/E3SUVNKZ6IOZTZSfE8t6Y931KRnYXaHG+R\nf6TAkCgDjK4QlgZZZdjIddPOHrG5qa77AxdN7bvGiUz9zFAF/TENC9Q4NFbsRwtEkp3WwiDVUarQ\nzTpWcRAtVmWRlSpWdG2KvH9oOrQiJ44lHcuSPy/AEM9DnuPn+YLyh89zjnkcuVwuh4iIiI2cS90B\nERFZ+RQ2IiJiO4WNiIjYTmEjIiK2U9iIiIjtFDYiImK7gmu9oLW1lTNnzlBcXMzRo0cBiMfjPP/8\n80xNTeFyudi7dy/r1q0DIBqN0tXVhcvlora2lk2bNgEwODhIS0sLly9f5p577qG2thaATCZDU1MT\ng4ODGIZBOBympKQEgO7ubqLRKACf/OQn2bZt26IXQERE7HfNmc2OHTs4dOjQrLZ//Md/5C/+4i/4\nxje+QU1NDe3t7QAMDQ3R09NDfX09dXV1tLW1MfPfeNra2ti3bx+RSITz58/T398PQGdnJx6Ph4aG\nBqqrq61jpVIpjh8/zpEjRzh8+DAvvvgik5OTH2hQsVjsg1dghVMt8lSLPNUiT7XIs7MW1wybjRs3\nUlRUNKvN6/Va//D/5je/4c477wSgr6+PrVu34nK58Pv9lJaWEo/HGRsbI51OEwgEAKiqqqK3txeA\n3t5ea8ayZcsWXn/9dQBee+01KioqcLvdFBUVUVFRYQXUtejNk6da5KkWeapFnmqRZ2ctrrmMdjWP\nPvooX/7yl/nud78LwNe+9jUAkskkGzZssF5nmibJZBKXy4XP57PafT4fyWTS2mfmOafTidvtJpVK\nzWq/8lgiIrL8zOsCgdbWVnbv3k1rayuf+cxnaG1tXbQO6dtzRERWnnnNbOLxOF/+8peB6aWv5557\nDpiefYyOjlqvSyQSmKaJaZokEok57TP7zGxns1nS6TQejwfTNGdN6RKJBOXl5VftTywWm/Xampqa\n+QxrRVIt8lSLPNUiT7XIq6mpoaOjw9oOBoMEg8FFOfYHCptcLjdrxvE7v/M7nD17lo9//OMMDAxQ\nWloKQCgUoqGhgYcffphkMsnw8DCBQACHw4Hb7SYej7Nu3TpOnTrFgw8+aO1z8uRJ1q9fT09PjxUo\nmzZt4oUXXmBycpJsNsvAwACPPvroVft3tYK8884711+NFcgwDCYmJpa6GzcF1SJPtchTLfLKysps\nC1/Htb71ORKJcPbsWSYmJiguLqampobf+73fo62tjUwmw2233cbevXv56Ec/Ckxf+tzZ2UlBQcGc\nS5+bm5utS593794NwOXLl2lsbOTcuXMYhsGBAwfw+/3A9KXPL730Eg6H47ovfVbYTNMvUp5qkada\n5KkWeWVlZbYd+5phs1wpbKbpFylPtchTLfJUizw7w0bfICAiIrZT2IiIiO0UNiIiYjuFjYiI2E5h\nIyIitlPYiIiI7RQ2IiJiO4WNiIjYTmEjIiK2U9iIiIjt5vWtz8uB6+fTN2Fz+FaR8X1kiXsjInJr\nW7Fhc+np/wNA4eMHQWEjIrKktIwmIiK2U9iIiIjtFDYiImI7hY2IiNjumhcItLa2cubMGYqLizl6\n9KjV/m//9m/88Ic/xOl08gd/8AfWLZuj0ShdXV24XK45d+psaWmx7tRZW1sLQCaToampicHBQQzD\nIBwOU1JSAkzfqTMajQJc9506RUTk5nHNmc2OHTs4dOjQrLZYLMZPfvITjh49yjPPPMOf/dmfATA0\nNERPTw/19fXU1dXR1tbGzI1A29ra2LdvH5FIhPPnz9Pf3w9AZ2cnHo+HhoYGqquraW9vByCVSnH8\n+HGOHDnC4cOHefHFF5mcnFzUwYuIyI1xzbDZuHEjRUVFs9p++MMfsmvXLlwuFwB33HEHAH19fWzd\nuhWXy4Xf76e0tJR4PM7Y2BjpdJpAIABAVVUVvb29APT29lozli1btvD669P/P+a1116joqICt9tN\nUVERFRUVVkCJiMjyMq//Z3P+/HnOnj3LP/3TP1FYWMhjjz3Gxz72MZLJJBs2bLBeZ5omyWQSl8uF\nz+ez2n0+H8lkEoBkMmk953Q6cbvdpFKpWe1XHktERJafeYXN1NQUv/nNb/j6179OPB7n2Wefpamp\naVE6NLPsdj1isRixWMzarqmpsR47nU7chrEofVuOCgsLMW7h8V9JtchTLfJUi9k6Ojqsx8FgkGAw\nuCjHnVfYlJSU8Ed/9EcABAIBnE4nExMTmKbJ6Oio9bpEIoFpmpimSSKRmNMOWM+Zpkk2myWdTuPx\neDBNc1aAJBIJysvLr9qf9ytINptlYmJiPsNcEQzDuKXHfyXVIk+1yFMt8gzDmPVhfTF9oEufc7nc\nrBlHZWWldW7lnXfeIZPJYBgGoVCI06dPk8lkGBkZYXh4mEAggNfrxe12E4/HyeVynDp1isrKSgBC\noRAnT54EoKenxwqUTZs2MTAwwOTkJKlUioGBAevKNhERWV4cuWusW0UiEc6ePcvExATFxcXU1NRQ\nVVVFS0sL586d47bbbuMv//Iv+fjHPw5MX/rc2dlJQUHBnEufm5ubrUufd+/eDcDly5dpbGzk3Llz\nGIbBgQMH8Pv9wPSlzy+99BIOh+O6L31+uzoETH832lTl/ddfmRVCn9ryVIs81SJPtcgrKyuz7djX\nDJvlSmEzTb9IeapFnmqRp1rk2Rk2+gYBERGxncJGRERsp7ARERHbKWxERMR2ChsREbGdwkZERGyn\nsBEREdspbERExHYKGxERsZ3CRkREbKewERER2ylsRETEdgobERGxncJGRERsp7ARERHbKWxERMR2\nBdd6QWtrK2fOnKG4uJijR4/Oeu773/8+7e3tPP/883g8HmD6Tp1dXV24XK45d+psaWmx7tRZW1sL\nQCaToampicHBQQzDIBwOU1JSAkzfqTMajQJc9506RUTk5nHNmc2OHTs4dOjQnPZEIsFPf/pTKxgA\nhoaG6Onpob6+nrq6Otra2pi5EWhbWxv79u0jEolw/vx5+vv7Aejs7MTj8dDQ0EB1dTXt7e0ApFIp\njh8/zpEjRzh8+DAvvvgik5OTizJoERG5sa4ZNhs3bqSoqGhO+3e+8x0ee+yxWW19fX1s3boVl8uF\n3++ntLSUeDzO2NgY6XSaQCAAQFVVFb29vQD09vZaM5YtW7bw+uuvA/Daa69RUVGB2+2mqKiIiooK\nK6BERGR5mdc5m76+Pnw+H2vWrJnVnkwmZ810TNMkmUySTCbx+XxWu8/nI5lMWvvMPOd0OnG73aRS\nqTn7zBxLRESWn2ues3m3S5cuEY1G+dKXvmRHf6xlt+sRi8WIxWLWdk1NjfXY6XTiNoxF6dtyVFhY\niHELj/9KqkWeapGnWszW0dFhPQ4GgwSDwUU57nWHzfDwMCMjIxw8eJBcLkcymeTJJ5/k8OHDmKbJ\n6Oio9dpEIoFpmpimSSKRmNMOWM+Zpkk2myWdTuPxeDBNc1aAJBIJysvLr9qn9ytINptlYmLieoe5\nYhiGcUuP/0qqRZ5qkada5BmGMevD+mL6QMtouVzOmnGsWbOGb37zmzQ1NdHc3Ixpmvzd3/0dxcXF\nhEIhTp8+TSaTYWRkhOHhYQKBAF6vF7fbTTweJ5fLcerUKSorKwEIhUKcPHkSgJ6eHitQNm3axMDA\nAJOTk6RSKQYGBqwr20REZHm55swmEolw9uxZJiYm+OxnP0tNTQ07duywnnc4HNbj1atXc9999xEO\nhykoKGDv3r3W83v27KG5udm69Hnz5s0A7Ny5k8bGRvbv349hGBw4cAAAj8fDn//5n/PFL34Rh8PB\npz71qateqCAiIjc/R24+J0mWgberQwAUPn6Qqcr7l7g3S0dLBHmqRZ5qkada5JWVldl2bH2DgIiI\n2E5hIyIitlPYiIiI7RQ2IiJiO4WNiIjYTmEjIiK2U9iIiIjtFDYiImI7hY2IiNhOYSMiIrZT2IiI\niO0UNiIiYjuFjYiI2E5hIyIitlPYiIiI7RQ2IiJiu2veqbO1tZUzZ85QXFzM0aNHAWhvb+cnP/kJ\nBQUFfOQjH+GJJ57A7XYDEI1G6erqwuVyUVtba93KeXBwkJaWFutOnbW1tQBkMhmampoYHBzEMAzC\n4TAlJSUAdHd3E41GAfjkJz/Jtm3bFr0AIiJiv2vObHbs2MGhQ4dmtVVUVPDMM8/w9NNPU1payokT\nJwAYGhqip6eH+vp66urqaGtrY+ZGoG1tbezbt49IJML58+fp7+8HoLOzE4/HQ0NDA9XV1bS3twOQ\nSqU4fvw4R44c4fDhw7z44otMTk4u6uBFROTGuGbYbNy4kaKiolltFRUVOJ3Tu65fv55EIgFAX18f\nW7duxeVy4ff7KS0tJR6PMzY2RjqdJhAIAFBVVUVvby8Avb291oxly5YtvP766wC89tprVFRU4Ha7\nKSoqoqKiwgooERFZXhZ8zqarq4t77rkHgGQyaS2BAZimSTKZJJlM4vP5rHafz0cymbT2mXnO6XTi\ndrtJpVJz9pk5loiILD/XPGfzfl566SVcLhd/8id/slj9sZbdrkcsFiMWi1nbNTU11mOn04nbMBal\nb8tRYWEhxi08/iupFnmqRZ5qMVtHR4f1OBgMEgwGF+W48w6b7u5uXn31Vb7yla9YbaZpMjo6am0n\nEglM08Q0TWup7cr2mX1mtrPZLOl0Go/Hg2maswIkkUhQXl5+1b68X0Gy2SwTExPzHeayZxjGLT3+\nK6kWeapFnmqRZxjGrA/ri+kDLaPlcrlZM47+/n7++Z//mS984QvcdtttVnsoFOL06dNkMhlGRkYY\nHh4mEAjg9Xpxu93E43FyuRynTp2isrLS2ufkyZMA9PT0WIGyadMmBgYGmJycJJVKMTAwYF3ZJiIi\ny4sjd411q0gkwtmzZ5mYmKC4uJiamhqi0SiZTMaaeq5fv569e/cC05c+d3Z2UlBQMOfS5+bmZuvS\n5927dwNw+fJlGhsbOXfuHIZhcODAAfx+PzA9e3rppZdwOBzXfenz29UhAAofP8hU5f3XWZaVQ5/a\n8lSLPNUiT7XIKysrs+3Y1wyb5UphM02/SHmqRZ5qkada5NkZNvoGARERsZ3CRkREbKewERER2yls\nRETEdgobERGxncJGRERsp7ARERHbKWxERMR2ChsREbGdwkZERGynsBEREdspbERExHYKGxERsZ3C\nRkREbKewERER213zttCtra2cOXOG4uJijh49CkAqleLYsWNcuHABv99POBzG7XYD0zdP6+rqwuVy\nzbl5WktLi3XztNraWgAymQxNTU0MDg5iGAbhcJiSkhJg+uZp0WgU4LpvniYiIjePa85sduzYwaFD\nh2a1nThxgrvvvptIJEIwGLQCYWhoiJ6eHurr66mrq6Otrc26nXRbWxv79u0jEolw/vx5+vv7Aejs\n7MTj8dDQ0EB1dTXt7e3AdKAdP36cI0eOcPjwYV588UUmJycXdfAiInJjXDNsNm7cSFFR0ay2vr4+\na5axfft2ent7rfatW7ficrnw+/2UlpYSj8cZGxsjnU4TCAQAqKqqsvbp7e21jrVlyxZef/11AF57\n7TUqKipwu90UFRVRUVFhBZSIiCwv8zpnMz4+jtfrBcDr9TI+Pg5AMpm0lsAATNMkmUySTCbx+XxW\nu8/nI5lMWvvMPOd0OnG73aRSqTn7zBxLRESWn0W5QMDhcCzGYQCsZTcREVk5rnmBwNV4vV7Gxsas\nn8XFxcD07GN0dNR6XSKRwDRNTNMkkUjMaZ/ZZ2Y7m82STqfxeDyYpkksFpu1T3l5+VX7E4vFZr22\npqbGeux0OnEbxnyGuSIUFhZi3MLjv5Jqkada5KkWs3V0dFiPg8EgwWBwUY77gcIml8vNmnHce++9\ndHd3s2vXLrq7uwmFQgCEQiEaGhp4+OGHSSaTDA8PEwgEcDgcuN1u4vE469at49SpUzz44IPWPidP\nnmT9+vX09PRYgbJp0yZeeOEFJicnyWazDAwM8Oijj161f+9XkGw2y8TExAevyApjGMYtPf4rqRZ5\nqkWeapFnGMasD+uLyZG7xrpVJBLh7NmzTExMUFxcTE1NDZWVldTX1zM6OsqqVasIh8PWRQTRaJTO\nzk4KCgrmXPrc3NxsXfq8e/duAC5fvkxjYyPnzp3DMAwOHDiA3+8Hpi99fumll3A4HNd96fPb1dMB\nWPj4QaYq77/+yqwQ+kXKUy3yVIs81SKvrKzMtmNfM2yWK4XNNP0i5akWeapFnmqRZ2fY6BsERETE\ndgobERGxncJGRERsp7ARERHbKWxERMR2ChsREbGdwkZERGynsBEREdspbERExHYKGxERsZ3CRkRE\nbKewERER2ylsRETEdgobERGxncJGRERsp7ARERHbfaDbQr+XaDTKK6+8gtPpZM2aNTzxxBNcvHiR\nY8eOceHCBfx+P+FwGLfbbb2+q6sLl8s15y6eLS0t1l08a2trAchkMjQ1NTE4OIhhGITDYUpKShY2\nYhERueHmPbO5cOECL7/8Mt/4xjc4evQoU1NT/OhHP+LEiRPcfffdRCIRgsEg0WgUgKGhIXp6eqiv\nr6euro62tjZmbhLa1tbGvn37iEQinD9/nv7+fgA6OzvxeDw0NDRQXV1Ne3v7IgxZRERutHmHzYc/\n/GEKCgq4ePEiU1NTXLp0CdM06evrY9u2bQBs376d3t5eAPr6+ti6dSsulwu/309paSnxeJyxsTHS\n6TSBQACAqqoqa5/e3l7rWFu2bGFgYGBBgxURkaUx72U0j8fDww8/zBNPPMGHPvQhKioqqKioYHx8\nHK/XC4DX62V8fByAZDLJhg0brP1N0ySZTOJyufD5fFa7z+cjmUxa+8w853Q6KSoqIpVK4fF45ttt\nERFZAvOe2fzyl7/kX/7lX2hpaeHv//7v+e1vf8srr7wy53UOh2NBHbzSzLKbiIgsL/Oe2fziF7/g\n93//961Zxh/+4R/yn//5n3i9XsbGxqyfxcXFwPRMZnR01No/kUhgmiamaZJIJOa0z+wzs53NZkmn\n01ed1cRiMWKxmLVdU1NjPXY6nbgNY77DXPYKCwsxbuHxX0m1yFMt8lSL2To6OqzHwWCQYDC4KMed\nd9iUlZVx/PhxLl26xG233cbAwADr1q3j9ttvp7u7m127dtHd3U0oFAIgFArR0NDAww8/TDKZZHh4\nmEAggMPhwO12E4/HWbduHadOneLBBx+09jl58iTr16+np6eH8vLyq/bl/QqSzWaZmJiY7zCXPcMw\nbunxX0m1yFMt8lSLPMMwZn1YX0zzDpu1a9eybds2vvjFL+J0Olm7di2f+MQnuHjxIvX19XR1dbFq\n1SrC4TAAq1ev5r777iMcDlNQUMDevXutJbY9e/bQ3NxsXfq8efNmAHbu3EljYyP79+/HMAwOHDiw\nCEMWEZEbzZFboSdC3q6enlEVPn6Qqcr7l7g3S0ef2vJUizzVIk+1yCsrK7Pt2PoGARERsZ3CRkRE\nbKewERER2ylsRETEdgobERGxncJGRERsp7ARERHbKWxERMR2ChsREbGdwkZERGynsBEREdspbERE\nxHYKGxERsZ3CRkREbKewERER2ylsRETEdvO+UyfA5OQkzz33HG+//TYOh4PPfvazlJaWcuzYMS5c\nuIDf7yccDuN2uwGIRqN0dXXhcrmora1l06ZNAAwODtLS0mLdqbO2thaATCZDU1MTg4ODGIZBOBym\npKRkYSMWEZEbbkEzm29961vcc8891NfX8/TTT3PXXXdx4sQJ7r77biKRCMFgkGg0CsDQ0BA9PT3U\n19dTV1dHW1sbMzcJbWtrY9++fUQiEc6fP09/fz8AnZ2deDweGhoaqK6upr29fYHDFRGRpTDvsJmc\nnOSNN95gx44dALhcLtxuN319fWzbtg2A7du309vbC0BfXx9bt27F5XLh9/spLS0lHo8zNjZGOp0m\nEAgAUFVVZe3T29trHWvLli0MDAzMf6QiIrJk5r2MNjIygmEYtLS08NZbb/Gxj32M2tpaxsfH8Xq9\nAHi9XsbHxwFIJpNs2LDB2t80TZLJJC6XC5/PZ7X7fD6SyaS1z8xzTqeToqIiUqkUHo9nvt0WEZEl\nMO+wyWazvPnmm+zZs4d169bx7W9/mxMnTsx5ncPhWFAHrzSz7PZusViMWCxmbdfU1FiPnU4nbsNY\ntD4sN4WFhRi38PivpFrkqRZ5qsVsHR0d1uNgMEgwGFyU4847bEzTxOfzsW7dOmB6mevEiRN4vV7G\nxsasn8XFxdbrR0dHrf0TiQSmaWKaJolEYk77zD4z29lslnQ6fdVZzfsVJJvNMjExMd9hLnuGYdzS\n47+SapGnWuSpFnmGYcz6sL6Y5n3Oxuv14vP5eOeddwAYGBhg9erV3HvvvXR3dwPQ3d1NKBQCIBQK\ncfr0aTKZDCMjIwwPDxMIBPB6vbjdbuLxOLlcjlOnTlFZWWntc/LkSQB6enooLy9fyFhFRGSJLOjS\n5927d9PY2Egmk+EjH/kITzzxBNlslvr6erq6uli1ahXhcBiA1atXc9999xEOhykoKGDv3r3WEtue\nPXtobm62Ln3evHkzADt37qSxsZH9+/djGAYHDhxY4HBFRGQpOHLvdSJkmXu7enpGVfj4QaYq71/i\n3iwdLRHkqRZ5qkWeapFXVlZm27H1DQIiImI7hY2IiNhOYSMiIrZT2IiIiO0UNiIiYjuFjYiI2E5h\nIyIitlPYiIiI7RQ2IiJiO4WNiIjYTmEjIiK2U9iIiIjtFDYiImI7hY2IiNhOYSMiIrZT2IiIiO0W\ndKdOgGw2S11dHaZp8uSTT5JKpTh27BgXLlzA7/cTDodxu90ARKNRurq6cLlc1NbWsmnTJgAGBwdp\naWmx7tRZW1sLQCaToampicHBQQzDIBwOU1JSstAui4jIDbbgmc2//uu/ctddd1nbJ06c4O677yYS\niRAMBolGowAMDQ3R09NDfX09dXV1tLW1MXOT0La2Nvbt20ckEuH8+fP09/cD0NnZicfjoaGhgerq\natrb2xfaXRERWQILCptEIsGrr77KAw88YLX19fWxbds2ALZv305vb6/VvnXrVlwuF36/n9LSUuLx\nOGNjY6TTaQKBAABVVVXWPr29vdaxtmzZwsDAwEK6KyIiS2RBYfOd73yHxx57DIfDYbWNj4/j9XoB\n8Hq9jI+PA5BMJmctgZmmSTKZJJlM4vP5rHafz0cymbT2mXnO6XRSVFREKpVaSJdFRGQJzPuczZkz\nZyguLmbt2rXEYrH3fN2VQbRQM8tu7xaLxWb1oaamxnrsdDpxG8ai9WG5KSwsxLiFx38l1SJPtchT\nLWbr6OiwHgeDQYLB4KIcd95h88Ybb9DX18err77KpUuXSKfTNDY24vV6GRsbs34WFxcD0zOZ0dFR\na/9EIoFpmpimSSKRmNM+s8/MdjabJZ1O4/F45vTl/QqSzWaZmJiY7zCXPcMwbunxX0m1yFMt8lSL\nPMMwZn1YX0zzXkb79Kc/TWtrK01NTfz1X/815eXl/NVf/RX33nsv3d3dAHR3dxMKhQAIhUKcPn2a\nTCbDyMgIw8PDBAIBvF4vbrebeDxOLpfj1KlTVFZWWvucPHkSgJ6eHsrLyxc4XBERWQoLvvT53Xbt\n2kV9fT1dXV2sWrWKcDgMwOrVq7nvvvsIh8MUFBSwd+9ea4ltz549NDc3W5c+b968GYCdO3fS2NjI\n/v37MQyDAwcOLHZ3RUTkBnDk3utEyDL3dvX0jKrw8YNMVd6/xL1ZOloiyFMt8lSLPNUir6yszLZj\n6xsERETEdgobERGxncJGRERsp7ARERHbKWxERMR2ChsREbGdwkZERGynsBEREdspbERExHYKGxER\nsZ3CRkREbKewERER2ylsRETEdgobERGxncJGRERsp7ARERHbzftOnYlEgqamJsbHx3E4HDzwwAM8\n9NBDpFIpjh07xoULF/D7/YTDYdxuNwDRaJSuri5cLhe1tbVs2rQJgMHBQVpaWqw7ddbW1gKQyWRo\nampicHAQwzAIh8OUlJQsfNQiInJDzXtm43K5+MxnPsOzzz7L17/+dX7wgx/w3//935w4cYK7776b\nSCRCMBgkGo0CMDQ0RE9PD/X19dTV1dHW1sbMTULb2trYt28fkUiE8+fP09/fD0BnZycej4eGhgaq\nq6tpb29fhCGLiMiNNu+w8Xq9rF27FoDbb7+du+66i0QiQV9fH9u2bQNg+/bt9Pb2AtDX18fWrVtx\nuVz4/X5KS0uJx+OMjY2RTqcJBAIAVFVVWfv09vZax9qyZQsDAwPzHqiIiCydeS+jXWlkZIS33nqL\nDRs2MD4+jtfrBaYDaXx8HIBkMsmGDRusfUzTJJlM4nK58Pl8VrvP5yOZTFr7zDzndDopKioilUrh\n8Xiuq3+uX41C8sL//MWrmLpTS3EiIjfSgsPm4sWLPPvss9TW1nL77bfPed7hcCz0r7DMLLu9WywW\nIxaLWds1NTXWY6fTiWP8V6T/35MAfPjQM7jXfHTR+nSzKywsxDCMpe7GTUG1yFMt8lSL2To6OqzH\nwWCQYDC4KMddUNhMTU3xzDPPUFVVRWVlJTA9mxkbG7N+FhcXA9MzmdHRUWvfRCKBaZqYpkkikZjT\nPrPPzHY2myWdTl91VvN+BclmszCVuaLPGSYmJhYy7GXFMIxbarzvR7XIUy3yVIs8wzBmfVhfTAu6\n9Lm1tZXVq1fz0EMPWW333nsv3d3dAHR3dxMKhQAIhUKcPn2aTCbDyMgIw8PDBAIBvF4vbrebeDxO\nLpfj1KlTVnCFQiFOnjwJQE9PD+Xl5QvproiILJF5z2zeeOMNXnnlFdasWcMXvvAFHA4HjzzyCLt2\n7aK+vp6uri5WrVpFOBwGYPXq1dx3332Ew2EKCgrYu3evtcS2Z88empubrUufN2/eDMDOnTtpbGxk\n//79GIbBgQMHFmHIIiJyozly73UiZJl7u3p6RlX4+EEwS7j0P+dsCr/4d0yt+19L2bUbSksEeapF\nnmqRp1o+KxzEAAAHL0lEQVTklZWV2XZsfYOAiIjYTmEjIiK2U9iIiIjtFDYiImI7hY2IiNhOYSMi\nIrZT2IiIiO0UNiIiYjuFjYiI2E5hIyIitlPYiIiI7Rbl5mnLiaOgANcvfja9oRupiYjcELdc2DDx\nay5F/i8w/aWcKGxERGynZTQREbGdwkZERGy3LJbR+vv7+fa3v00ul2PHjh3s2rVrUY476/wN6ByO\niIhNbvqwyWazPP/883zlK1/hzjvvpK6ujsrKSu66666FH/yK8zcAH/rSM7iSF6Y3FDwiIovmpl9G\ni8fjlJaWsmrVKgoKCvjjP/5jent77fnLJn7Npf/35PRdPWdCR0REFuymn9kkk0l8Pp+1bZom8Xjc\n9r/3yiU2R5FB7jcTcx5Pd0gzIBGRa7npw2a+bvv0/wbAuTZA9tdj13+AK5bYPnTgq1d9DLOX3t4r\nlN4dUB/kde+3jwJORJYbRy6Xyy11J97Pz3/+c773ve9x6NAhAE6cOAEw6yKBWCxGLBaztmtqam5s\nJ0VEVoiOjg7rcTAYJBgMLspxb/pzNoFAgOHhYS5cuEAmk+HHP/4xoVBo1muCwSA1NTXWnyuLdatT\nLfJUizzVIk+1yOvo6Jj1b+liBQ0sg2U0p9PJnj17eOqpp8jlcuzcuZPVq1cvdbdEROQ63PRhA7B5\n82YikchSd0NERObppl9Gm4/FnPotd6pFnmqRp1rkqRZ5dtbipr9AQERElr8VObMREZGbi8JGRERs\ntywuELgedn1p583kc5/7HG63G4fDgcvl4siRI6RSKY4dO8aFCxfw+/2Ew2HcbjcA0WiUrq4uXC4X\ntbW1bNq0CYDBwUFaWlq4fPky99xzD7W1tUs4qg+utbWVM2fOUFxczNGjRwEWdfyZTIampiYGBwcx\nDINwOExJyc35n2ivVovvfe97vPzyyxQXFwPwyCOPsHnzZmDl1iKRSNDU1MT4+DgOh4MHHniAhx56\n6JZ8X7y7Fp/4xCd48MEHl/59kVtBpqamcp///OdzIyMjucuXL+f+9m//Njc0NLTU3Vp0n/vc53IT\nExOz2v7hH/4hd+LEiVwul8tFo9Fce3t7LpfL5d5+++3cwYMHc5lMJvfLX/4y9/nPfz6XzWZzuVwu\nV1dXl/uv//qvXC6Xyx0+fDj36quv3sBRzN/Pfvaz3Jtvvpn7m7/5G6ttMcf/gx/8IPfNb34zl8vl\ncj/+8Y9z9fX1N2xs1+tqtejo6Mh9//vfn/PalVyLX/3qV7k333wzl8vlcul0Ord///7c0NDQLfm+\neK9aLPX7YkUto93QL+1cQrlcjty7ruvo6+tj27ZtAGzfvt0ad19fH1u3bsXlcuH3+yktLSUejzM2\nNkY6nSYQCABQVVW1bGq1ceNGioqKZrUt5vh7e3utY23ZsoWBgYEbNbTrdrVaAHPeH7Cya+H1elm7\ndi0At99+O3fddReJROKWfF9crRbJZBJY2vfFigqbq31p50yRVxKHw8FTTz1FXV0dL7/8MgDj4+N4\nvV5g+s02Pj4OTNfkyuntTE3eXSufz7esa7WY47/yOafTSVFREalU6kYNZVH8+7//OwcPHuS5555j\ncnISuHVqMTIywltvvcWGDRtu+ffFTC3Wr18PLO37YsWds7kVfO1rX+POO+/k17/+NU899RRlZWVz\nXuNwOJagZzePxRz/1T4N3sz+9E//lE996lM4HA5eeOEFvvvd77Jv375FOfbNXouLFy/y7LPPUltb\ny+233z7n+VvpffHuWiz1+2JFzWxM02R0dNTaTiaTmKa5hD2yx5133gnAHXfcQWVlJfF4HK/Xy9jY\n9Ldbj42NWScB312TRCKBaZqYpkkikZjTvlwt5vivfC6bzZJOp/F4PDdqKAt2xx13WP+oPvDAA9Yt\nOVZ6LaampnjmmWeoqqqisrISuHXfF1erxVK/L1ZU2HyQL+1c7n77299y8eJFYPqTy09/+lPWrFnD\nvffeS3d3NwDd3d3WuEOhEKdPnyaTyTAyMsLw8DCBQACv14vb7SYej5PL5Th16pT1plwO3n3eajHH\nHwqFOHnyJAA9PT2Ul5ff2MFdp3fXYuYfV4D/+I//4Hd/93eBlV+L1tZWVq9ezUMPPWS13arvi6vV\nYqnfFyvuGwT6+/v51re+ZX1p50q79HlkZISnn34ah8PB1NQU999/P7t27SKVSlFfX8/o6CirVq0i\nHA5bJ46j0SidnZ0UFBTMuayxubnZuqxx9+7dSzm0DywSiXD27FkmJiYoLi6mpqaGysrKRRv/5cuX\naWxs5Ny5cxiGwYEDB/D7/Us23vdztVrEYjHOnTuHw+Fg1apVPP7449Z5i5VaizfeeIOvfvWrrFmz\nBofDgcPh4JFHHiEQCNxy74v3qsWPfvSjJX1frLiwERGRm8+KWkYTEZGbk8JGRERsp7ARERHbKWxE\nRMR2ChsREbGdwkZERGynsBEREdspbERExHb/H5OU7LCfXT0lAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f607b956a20>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure_ = plt.hist((exons.end - exons.start), bins=100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter out some very small exons:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "exons = exons[(exons.end - exons.start) > 5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Specify the recombination rate within exons:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "exons.ix[exons.type == 'exon', 'recomb_rate'] = RECOMB_RATE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>recomb_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69090</td>\n",
       "      <td>70008</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>134900</td>\n",
       "      <td>135802</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>137620</td>\n",
       "      <td>139379</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>367639</td>\n",
       "      <td>368634</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>621058</td>\n",
       "      <td>622053</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end  type   recomb_rate\n",
       "0      1   69090   70008  exon  1.000000e-08\n",
       "1      1  134900  135802  exon  1.000000e-08\n",
       "2      1  137620  139379  exon  1.000000e-08\n",
       "3      1  367639  368634  exon  1.000000e-08\n",
       "4      1  621058  622053  exon  1.000000e-08"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process positions from the archaic admixture array"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the coordinates of sites from the archaic admixture array:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "array_sites = pd.read_table('../raw_data/ice_age/filter.gz', names=['chrom', 'end', 'set_1', 'set_2'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Subset to sites in the so called \"Set 1\" (see `raw_data/ice_age/README.md` for more information):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "array_sites = array_sites[array_sites.set_1 == 1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find sites that don't fall within exonic regions. These will be simulated individually when building the recombination map. The ones within the exons will be recombining simply as part of the exons:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "array_sites['start'] = array_sites.end - 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "array_sites = array_sites[['chrom', 'start', 'end']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "array_sites['type'] = 'snp'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a union of exon regions with admixture array informative positions\n",
    "\n",
    "Informative sites within exons will be simulated simply as part of the exons. Inter-exon informative sites will be simulated individually, with an appropriate recombination rate set between them and both adjacent exons or other informative sites."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "exome = BedTool.from_dataframe(exons)\n",
    "\n",
    "# take subset of array sites that lie outside exons\n",
    "array = BedTool.from_dataframe(array_sites).intersect(exome, v=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "exons_and_sites = pd.concat([exome.to_dataframe(),\n",
    "                             array.to_dataframe()]).sort_values(by=['chrom', 'start']).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# rename and rearrange columns\n",
    "exons_and_sites = exons_and_sites.rename(columns={'name': 'type', 'score': 'recomb_rate'})[['chrom', 'start', 'end', 'type', 'recomb_rate']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generating the recombination map"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Recombination rate between exons and/or informative positions is implemented by inserting a 1 bp \"gap\" between each adjacent region/site and setting the recombination rate at these positions to $L \\times 1\\cdot10^{-8}$ crossovers per generation ($L$ is the distance between exons/sites).\n",
    "\n",
    "The recombination rate of the \"gap\" between the last feature on one chromosome and the first feature on another chromosome will be 0.5."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def add_recombination_gaps(regions):\n",
    "    \"\"\"Add 1 bp \"recombination gaps\" between each record in\n",
    "    a given DataFrame.\n",
    "    \"\"\"\n",
    "    # create a new DataFrame with coordinates of 1 bp gaps\n",
    "    gaps = pd.DataFrame({'chrom'       : regions.chrom.values,\n",
    "                         'start'       : regions.end.values,\n",
    "                         'end'         : regions.end.values + 1,\n",
    "                         'recomb_rate' : list(RECOMB_RATE * (regions.start[1:].values - regions.end[:-1].values)) + [0.5],\n",
    "                         'type'        : 'gap'},\n",
    "                        columns=['chrom', 'start', 'end', 'recomb_rate', 'type'])\n",
    "\n",
    "    # merge the dataframes of regions and gap coordinates\n",
    "    regions_and_gaps = pd.concat([regions.query('type != \"snp\"'), gaps]).sort_values(by=['chrom', 'start']).reset_index(drop=True)\n",
    "    \n",
    "    # calculate the end position of each simulated region and gap\n",
    "    # (this is how a recombination map is specified for SLiM)\n",
    "    regions_and_gaps['interval_end'] = (regions_and_gaps.end - regions_and_gaps.start).cumsum()\n",
    "    \n",
    "    # return recombination map between regions and admixture array sites\n",
    "    # (snp records can be ignored, since they were used only to place\n",
    "    # 1 bp \"gaps\" in between)\n",
    "    return regions_and_gaps[['chrom', 'start', 'end', 'type', 'interval_end', 'recomb_rate']]\n",
    "\n",
    "\n",
    "def create_genome_recomb_map(regions):\n",
    "    \"\"\"Create recombination map for all chromosomes.\n",
    "    \n",
    "    SLiM simulates the whole genome as one segment. Therefore, all\n",
    "    exons on all chromosomes have to be concatenated and all coordinates\n",
    "    have to be shifted appropriately (i.e. first coordinate on the chromosome 2\n",
    "    has to be incremented by a last coordinate of the chromosome 1).\n",
    "    \"\"\"\n",
    "    recomb_map = {}\n",
    "    previous_end = 0\n",
    "\n",
    "    for chrom in AUTOSOMES:\n",
    "        # add recombination gaps between regions on this chromosome\n",
    "        recomb_map[chrom] = add_recombination_gaps(regions.query('chrom == {}'.format(chrom)))\n",
    "\n",
    "        # increment the positions of the recombination map based on the\n",
    "        # last position on the previous chromosome\n",
    "        recomb_map[chrom].interval_end += previous_end\n",
    "\n",
    "        # get the currently last position of the\n",
    "        previous_end = max(recomb_map[chrom].interval_end)\n",
    "        \n",
    "    # concatenate exome recombination maps of all chromosomes\n",
    "    recomb_map = pd.concat(recomb_map, ignore_index=True).sort_values(by='interval_end')\n",
    "\n",
    "    # remove the very last base of the recombination map\n",
    "    # (it has a 0.5 recombination rate anyway and there's no other chromosome\n",
    "    # after it)\n",
    "    recomb_map = recomb_map[:-1]\n",
    "    \n",
    "    return recomb_map"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Recombination map of informative positions from the admixture array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sites_recomb_map = create_genome_recomb_map(array_sites)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>interval_end</th>\n",
       "      <th>recomb_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>436719</th>\n",
       "      <td>22</td>\n",
       "      <td>51147735</td>\n",
       "      <td>51147736</td>\n",
       "      <td>gap</td>\n",
       "      <td>783969</td>\n",
       "      <td>0.000051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436720</th>\n",
       "      <td>22</td>\n",
       "      <td>51152866</td>\n",
       "      <td>51152867</td>\n",
       "      <td>gap</td>\n",
       "      <td>783970</td>\n",
       "      <td>0.000020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436721</th>\n",
       "      <td>22</td>\n",
       "      <td>51154908</td>\n",
       "      <td>51154909</td>\n",
       "      <td>gap</td>\n",
       "      <td>783971</td>\n",
       "      <td>0.000187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436722</th>\n",
       "      <td>22</td>\n",
       "      <td>51173632</td>\n",
       "      <td>51173633</td>\n",
       "      <td>gap</td>\n",
       "      <td>783972</td>\n",
       "      <td>0.000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436723</th>\n",
       "      <td>22</td>\n",
       "      <td>51174091</td>\n",
       "      <td>51174092</td>\n",
       "      <td>gap</td>\n",
       "      <td>783973</td>\n",
       "      <td>0.000032</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom     start       end type  interval_end  recomb_rate\n",
       "436719     22  51147735  51147736  gap        783969     0.000051\n",
       "436720     22  51152866  51152867  gap        783970     0.000020\n",
       "436721     22  51154908  51154909  gap        783971     0.000187\n",
       "436722     22  51173632  51173633  gap        783972     0.000005\n",
       "436723     22  51174091  51174092  gap        783973     0.000032"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sites_recomb_map.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>954450</th>\n",
       "      <td>22</td>\n",
       "      <td>51152865</td>\n",
       "      <td>51152866</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>954451</th>\n",
       "      <td>22</td>\n",
       "      <td>51154907</td>\n",
       "      <td>51154908</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>954453</th>\n",
       "      <td>22</td>\n",
       "      <td>51173631</td>\n",
       "      <td>51173632</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>954454</th>\n",
       "      <td>22</td>\n",
       "      <td>51174090</td>\n",
       "      <td>51174091</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>954455</th>\n",
       "      <td>22</td>\n",
       "      <td>51177247</td>\n",
       "      <td>51177248</td>\n",
       "      <td>snp</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom     start       end type\n",
       "954450     22  51152865  51152866  snp\n",
       "954451     22  51154907  51154908  snp\n",
       "954453     22  51173631  51173632  snp\n",
       "954454     22  51174090  51174091  snp\n",
       "954455     22  51177247  51177248  snp"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "array_sites.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Recombination map with exons only\n",
    "\n",
    "This will be used in the mutation accumulation phase."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "exons_recomb_map = create_genome_recomb_map(exons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>interval_end</th>\n",
       "      <th>recomb_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>288404</th>\n",
       "      <td>22</td>\n",
       "      <td>51220615</td>\n",
       "      <td>51220779</td>\n",
       "      <td>exon</td>\n",
       "      <td>72547537</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288405</th>\n",
       "      <td>22</td>\n",
       "      <td>51220779</td>\n",
       "      <td>51220780</td>\n",
       "      <td>gap</td>\n",
       "      <td>72547538</td>\n",
       "      <td>4.170000e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288406</th>\n",
       "      <td>22</td>\n",
       "      <td>51221196</td>\n",
       "      <td>51221714</td>\n",
       "      <td>exon</td>\n",
       "      <td>72548056</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288407</th>\n",
       "      <td>22</td>\n",
       "      <td>51221714</td>\n",
       "      <td>51221715</td>\n",
       "      <td>gap</td>\n",
       "      <td>72548057</td>\n",
       "      <td>2.140000e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288408</th>\n",
       "      <td>22</td>\n",
       "      <td>51221928</td>\n",
       "      <td>51222091</td>\n",
       "      <td>exon</td>\n",
       "      <td>72548220</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom     start       end  type  interval_end   recomb_rate\n",
       "288404     22  51220615  51220779  exon      72547537  1.000000e-08\n",
       "288405     22  51220779  51220780   gap      72547538  4.170000e-06\n",
       "288406     22  51221196  51221714  exon      72548056  1.000000e-08\n",
       "288407     22  51221714  51221715   gap      72548057  2.140000e-06\n",
       "288408     22  51221928  51222091  exon      72548220  1.000000e-08"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons_recomb_map.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>recomb_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>144296</th>\n",
       "      <td>22</td>\n",
       "      <td>51215097</td>\n",
       "      <td>51215177</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144297</th>\n",
       "      <td>22</td>\n",
       "      <td>51216379</td>\n",
       "      <td>51216409</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144298</th>\n",
       "      <td>22</td>\n",
       "      <td>51220615</td>\n",
       "      <td>51220779</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144299</th>\n",
       "      <td>22</td>\n",
       "      <td>51221196</td>\n",
       "      <td>51221714</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144300</th>\n",
       "      <td>22</td>\n",
       "      <td>51221928</td>\n",
       "      <td>51222091</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom     start       end  type   recomb_rate\n",
       "144296     22  51215097  51215177  exon  1.000000e-08\n",
       "144297     22  51216379  51216409  exon  1.000000e-08\n",
       "144298     22  51220615  51220779  exon  1.000000e-08\n",
       "144299     22  51221196  51221714  exon  1.000000e-08\n",
       "144300     22  51221928  51222091  exon  1.000000e-08"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Recombination map of exons and informative positions from the admixture array\n",
    "\n",
    "This will be used to simulate the fate of introgressed sequence after Neanderthal admixture."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "exons_and_sites_recomb_map = create_genome_recomb_map(exons_and_sites)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>interval_end</th>\n",
       "      <th>recomb_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>712341</th>\n",
       "      <td>22</td>\n",
       "      <td>51220615</td>\n",
       "      <td>51220779</td>\n",
       "      <td>exon</td>\n",
       "      <td>73312246</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>712342</th>\n",
       "      <td>22</td>\n",
       "      <td>51220779</td>\n",
       "      <td>51220780</td>\n",
       "      <td>gap</td>\n",
       "      <td>73312247</td>\n",
       "      <td>4.170000e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>712343</th>\n",
       "      <td>22</td>\n",
       "      <td>51221196</td>\n",
       "      <td>51221714</td>\n",
       "      <td>exon</td>\n",
       "      <td>73312765</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>712344</th>\n",
       "      <td>22</td>\n",
       "      <td>51221714</td>\n",
       "      <td>51221715</td>\n",
       "      <td>gap</td>\n",
       "      <td>73312766</td>\n",
       "      <td>2.140000e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>712345</th>\n",
       "      <td>22</td>\n",
       "      <td>51221928</td>\n",
       "      <td>51222091</td>\n",
       "      <td>exon</td>\n",
       "      <td>73312929</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom     start       end  type  interval_end   recomb_rate\n",
       "712341     22  51220615  51220779  exon      73312246  1.000000e-08\n",
       "712342     22  51220779  51220780   gap      73312247  4.170000e-06\n",
       "712343     22  51221196  51221714  exon      73312765  1.000000e-08\n",
       "712344     22  51221714  51221715   gap      73312766  2.140000e-06\n",
       "712345     22  51221928  51222091  exon      73312929  1.000000e-08"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons_and_sites_recomb_map.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>recomb_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>978742</th>\n",
       "      <td>22</td>\n",
       "      <td>51215097</td>\n",
       "      <td>51215177</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>978743</th>\n",
       "      <td>22</td>\n",
       "      <td>51216379</td>\n",
       "      <td>51216409</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>978744</th>\n",
       "      <td>22</td>\n",
       "      <td>51220615</td>\n",
       "      <td>51220779</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>978745</th>\n",
       "      <td>22</td>\n",
       "      <td>51221196</td>\n",
       "      <td>51221714</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>978746</th>\n",
       "      <td>22</td>\n",
       "      <td>51221928</td>\n",
       "      <td>51222091</td>\n",
       "      <td>exon</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom     start       end  type   recomb_rate\n",
       "978742     22  51215097  51215177  exon  1.000000e-08\n",
       "978743     22  51216379  51216409  exon  1.000000e-08\n",
       "978744     22  51220615  51220779  exon  1.000000e-08\n",
       "978745     22  51221196  51221714  exon  1.000000e-08\n",
       "978746     22  51221928  51222091  exon  1.000000e-08"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons_and_sites.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output the results in a format suitable for import into SLiM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "!mkdir -p ../clean_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sites_recomb_map[['interval_end', 'recomb_rate']].to_csv('../clean_data/recomb_map__array_sites.txt', sep='\\t', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "exons_recomb_map[['interval_end', 'recomb_rate']].to_csv('../clean_data/recomb_map__exome.txt', sep='\\t', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "exons_and_sites_recomb_map[['interval_end', 'recomb_rate']].to_csv('../clean_data/recomb_map__exome_with_array_sites.txt', sep='\\t', index=False, header=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
